# r document van Steensel lab

# Gene repression in LADs
## Christ Leemans, 03-11-2016 - 22-09-2017

## Introduction
Generally speaking, genes inside lamina associated domains are not or very lowly expressed. These genes can either be actively repressed by their DNA context (e.g. heterochromatin, lamina association), or simply be inactive (because essential factors for expression are missing?). Yet another group of genes seem to evade gene repression in the context of lamina associated domains.




```{r}

library(reshape2)
library(rtracklayer)
library(ggplot2)
library(gridExtra)
library(plyr)
library(scales)
library(grid)
library(gtable)
library(affy)
library(limma)
library(biomaRt)
## FROM STACKOVERFLOW:
## https://stackoverflow.com/questions/12539348/ggplot-separate-legend-and-plot

pseudo_log10 <- function(val_vec){
    Pseud=min(val_vec[val_vec > 0], na.rm=TRUE)/2
    val_vec = val_vec + Pseud
    return(log10(val_vec))
}

## get a table with matching sets
## table = complete table to take matching sets from
## class_col = column name of class of interest
## class = name of class to match the set on
## order_on = column name to order on
matchSet <- function(table, class_col, class, order_on){
  o_vec = order(table[,order_on])
  o_table = table[o_vec, ]
  setA = which(o_table[,class_col]==class)
  setB = c(setA + 1, setA -1)
  ## check if setB is all within the possible indexes
  setB = setB[setB %in% 1:length(o_vec)]
  ## can also return o_table[unique(c(setA, setB)), ]
  ## but this way order is perserved.
  i_vec = o_vec[unique(c(setA, setB))]
  return(table[i_vec[order(i_vec)], ])
}


COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")



id_table = read.table('../raw_data/transcript.table', stringsAsFactors=F,
                      row.names=1, col.names=c('transcript_id', 'gene_id',
                                               'symbol'))
load("~joris/mydata/git/SuRE/Joris//analysis_postNBT/Gencode_DF_generation_170707/gencode.sure.170712.rda")


P<-gencode.sure.170712[,c('chr', 'strand', 'txStart', 'txEnd', 'name', 'name2',
                          'tss', 'distance.to.previous.tss',
                          'k562.combined.45.55.sense', 'HEPG2.sense',
                          'HT1080.sense', 'gro.cap.1kb.sense',
                          'encode.cage.reprocessed.1kb.sense',
                          'nr.of.tissues.in.which.expressed.max')]

names(P)[9:14]<-c("SuRE_K562", "SuRE_HepG2", "SuRE_HT1080", "GROcap_K562",
                  "CAGE_K562_encode", 'tissues_expressed')

rownames(P) = P$name
sd_jit = min(P$GROcap_K562[P$GROcap_K562>0])
jit = rnorm(nrow(P), sd = sd_jit / 20)

P$GROcap_K562_jitter = log10(P$GROcap_K562 + jit + sd_jit / 2)

for (col in c('SuRE_K562', 'SuRE_HepG2', 'SuRE_HT1080', 'GROcap_K562',
              'CAGE_K562_encode')){
    P[,col] = pseudo_log10(P[,col])
}

P$gene_id = id_table[P$name,'gene_id']

gene_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                            start=P$txStart,
                                            end=P$txEnd,
                                            strand=P$strand,
                                            name=P$name,
                                            tss=P$tss),
                                            keep.extra.columns=TRUE)
names(gene_gr) = P$name
tss_gr = gene_gr
ranges(tss_gr) = IRanges(gene_gr$tss,
                       gene_gr$tss)
names(tss_gr) = P$name

## get LAD data for K562
LAD_K562 = import.bed('~c.leemans/mydata/data/tracks/hg19/cl20161019_LAD_continuous_2state_K562.bed')
## to keep with Joris's previous analysis, let's assign a state to every promoter
o = findOverlaps(tss_gr, LAD_K562[LAD_K562$name=='LAD'])
P$LAD_K562 = 0
P$LAD_K562[queryHits(o)] = 1


create_RM <-function(data, x, y, lad){
    #then calculate running mean for iLAD promoters:
    #sort by SuRE and then random for ties
    o = order(data[,x],sample(c(1:nrow(data))))

    x_sorted = data[o,x]
    y_sorted = data[o,y]
    lad_sorted = data[o,lad]

    n<-60 #number of windows
    w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
    s<-round(seq(from=w/2+0.0001, to=nrow(data)-w/2, length.out=n))
    RM<-data.frame(x.low=rep(NA,n), x.mean=rep(NA,n), x.hi=rep(NA,n), y.lad=rep(NA,n), y.ilad=rep(NA,n))
    RM$x.low=x_sorted[s-floor(w/2)]
    for(i in 1:n){RM$x.mean[i]=mean(x_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
    RM$x.hi=x_sorted[s+floor(w/2)]
    for(i in 1:n)
      {t<-data.frame(LAD=lad_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))],
                     y=y_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))])
       RM$y.lad[i]<-mean(t$y[t$LAD==1], na.rm=TRUE)
       RM$y.ilad[i]<-mean(t$y[t$LAD==0], na.rm=TRUE)
      }
    #add first datapoint (SuRE equals pseudocount)
    RM1<-RM[0,] #empty df
    RM1[1,]<-c(rep(min(x_sorted),3), mean(y_sorted[x_sorted==min(x_sorted) & lad_sorted==1]), mean(y_sorted[x_sorted==min(x_sorted) & lad_sorted==0]))
    RM<-rbind(RM1, RM)
    rm(RM1)
    return(RM)
}

RM_GRO = create_RM(P, 'SuRE_K562', 'GROcap_K562', lad='LAD_K562')
P$LRS_GROcap <- P$GROcap_K562 - approx(x=RM_GRO$x.mean, y=RM_GRO$y.ilad, xout=P$SuRE_K562, rule=2)$y

classify <- function(sure, exp, lrs, lad, exp_cut){
    INACT<- sure< -0.3 & lad & exp< exp_cut #inactive
    NREP<- sure> 0 & lrs > -0.5 & lad & exp> exp_cut #not repressed
    REP<- sure> 0.3 & lrs < -1 & lad  & exp< exp_cut #repressed
    Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
    names(Pcnts)<-c("repressed", "escaper", "inactive")
    BND <- lad & !INACT & !NREP & !REP
    class = rep(NA, length(sure))
    class[lad==0] = 'iLAD'
    class[INACT]<-"inactive"
    class[NREP]<-"escaper"
    class[REP]<-"repressed"
    class[BND] <- "boundary"
    return(factor(class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary')))
}
P$class = classify(P$SuRE_K562, P$GROcap_K562, P$LRS_GROcap, P$LAD_K562, -2)

```


```{r}
trip_gr = import.gff3('../results/cl20170925_TRIP_promoters_evsr_hg19.gff3')
trip_data = as.data.frame(trip_gr)[c('ID', 'name', 'gene_id', 'gene_name',
                                     'gene_type', 'plasmid_id', 'class')]
trip_data$number=gsub('.*_', '', trip_data$name)
trip_data$transcript_id = gsub('_ext','', trip_data$ID)
trip_data$GROcap_K562=P[trip_data$transcript_id, 'GROcap_K562']
trip_data$GROcap_K562_jitter=P[trip_data$transcript_id, 'GROcap_K562_jitter']
trip_data$SuRE_K562=P[trip_data$transcript_id, 'SuRE_K562']
p_classes = P[which(P$class %in% c('inactive', 'escaper', 'repressed')),]

trip_sub = trip_data[trip_data$plasmid_id %in% c('pMvdZ06', 'pMvdZ08', 'pMvdZ011'), ]
ggplot(P[P$LAD_K562==0, ], aes(x=SuRE_K562, y=GROcap_K562_jitter,
                                            color=class)) +
               geom_line(data=RM_GRO[RM_GRO$y.ilad > -1.5,],
                         aes(x=x.mean, y=y.ilad - 0.5), color='black',
                         linetype='dotdash', size=0.5, show.legend=F) +
               geom_line(data=RM_GRO[RM_GRO$x.mean > 0.3 & RM_GRO$y.ilad < -1,],
                         aes(x=x.mean, y=y.ilad - 1), color='black',
                         linetype='dotdash', size=0.5, show.legend=F) +
               geom_segment(x=-0.3, xend=-0.3, y=-2, yend=min(P$GROcap_K562_jitter),
                            linetype='dotdash', color='black', size=0.5) +
               geom_hline(yintercept=-2, linetype='dotdash', size=0.5) +
               geom_point(size=0.1, alpha=0.1) +
           	   geom_point(data=trip_sub, size=1) +
               theme_bw() +
               geom_line(data=RM_GRO, aes(x=x.mean, y=y.ilad),
			             color=COL_lad['iLAD'], size=1) +
               labs(y='log10(GROcap)', x='log10(SuRE)') +
               theme(legend.justification =c(0,1),
			   		 legend.position=c(0.05,0.95),
					 legend.background = element_rect(fill="transparent"),
			   		 legend.title=element_blank(),
                     panel.grid.minor = element_blank()) +
               scale_color_manual(values=COL_class) +
               coord_equal(ratio=1)

```
