
for f in $(tail -n+2 raw_data/encode_jaspar.txt | awk '{print $1}')
do
    out=$(echo $f | sed 's/wgEncodeAwgTfbs/tssr_300_300_/' | \
          sed 's/.narrowPeak.gz/.signal.txt.gz/')
    bedtools intersect -wb -a raw_data/tssr_300_300.bed \
                       -b <(gunzip -c ~/data/tracks/hg19/EncodeTfbs_uniform/$f) | \
        awk '{arr[$4]+=$13}END{for(a in arr){print a"\t"arr[a]}}' | \
        gzip -c > $out
done

```{r, fig.width=10, fig.height=10}
library(reshape2)
library(rtracklayer)
library(DESeq2)
library(ggplot2)
library(gridExtra)
library(plyr)
library(preprocessCore)
library(scales)
library(gage)
library(fgsea)
library(CGtools)
library(gcrma)
library(limma)
library(biomaRt)
library(Matrix)
library(grid)
library(data.table)
load('../raw_data/biomart.rdata')

## There was one promoter that was wrongly annotated
bm_p[bm_p$ensembl_transcript_id=='ENST00000357491','ensembl_gene_id' ] = 'ENSG00000196350'

matchSet <- function(table, class_col, class, order_on){
  o_vec = order(table[,order_on])
  o_table = table[o_vec, ]
  setA = which(o_table[,class_col]==class)
  setB = c(setA + 1, setA -1)
  ## can also return o_table[unique(c(setA, setB)), ]
  ## but this way order is perserved.
  i_vec = o_vec[unique(c(setA, setB))]
  return(table[i_vec[order(i_vec)], ])
}
```


```{r}
load("../raw_data/gencode.sure.160617.rda")
Prom<-gencode.sure.160617; rm(gencode.sure.160617) #simpler name
#first re-calculate pseudocounts without jitter
P<-Prom[,c(1:8,23,20,26,28, 27)] #SuRE, GRO-cap, CAGE and LAD columns only
names(P)[9:13]<-c("SuRE", "GROcap", "CAGE", "LAD", 'tissues_expressed')


## for promoters and gene expression let's convert promoter transcript id's to gene id's
P$ensembl_transcript_id = do.call(rbind, strsplit(P$name, split='[.]'))[,1]

nrow(P) #orriginal number of rows
bm_match = match(P$ensembl_transcript_id, bm_p$ensembl_transcript_id)
P<-merge(P, bm_p, by="ensembl_transcript_id", all.x=TRUE)
nrow(P) #some double rows were introduced

P = P[match(Prom$name, P$name), ]

length(unique(P$ensembl_gene_id)) #number of unique genes

table(P[,c('strand.x','strand.y')]) #almost all strand listings are consistent

P<-P[, colnames(P)!='strand.y']
colnames(P)[colnames(P)=='strand.x'] = "strand"


Pseud<-min(P$SuRE[P$SuRE>0], na.rm=TRUE)/2
P$SuRE<-P$SuRE+Pseud
P$SuRE<-log10(P$SuRE)
PseudGro<-min(P$GROcap[P$GROcap>0], na.rm=TRUE)/2
P$GROcap<-P$GROcap+PseudGro
P$GROcap<-log10(P$GROcap)
PseudCage<-min(P$CAGE[P$CAGE>0], na.rm=TRUE)/2
P$CAGE<-P$CAGE+PseudCage
P$CAGE<-log10(P$CAGE)

#then calculate running mean for iLAD promoters:
P<-P[order(P$SuRE,sample(c(1:nrow(P)))),] #sort by SuRE and then random for ties
n<-60 #number of windows
w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
s<-round(seq(from=w/2+0.0001, to=nrow(P)-w/2, length.out=n))
RM<-data.frame(SuRE.low=rep(NA,n), SuRE.mean=rep(NA,n), SuRE.hi=rep(NA,n), GROcap.lad=rep(NA,n), GROcap.ilad=rep(NA,n))
RM$SuRE.low=P$SuRE[s-floor(w/2)]
for(i in 1:n){RM$SuRE.mean[i]=mean(P$SuRE[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
RM$SuRE.hi=P$SuRE[s+floor(w/2)]
for(i in 1:n)
  {t<-P[(s[i]-floor(w/2)):(s[i]+floor(w/2)),]
   RM$GROcap.lad[i]<-mean(t$GROcap[t$LAD==1], na.rm=TRUE)
   RM$GROcap.ilad[i]<-mean(t$GROcap[t$LAD==0], na.rm=TRUE)
  }


#add first datapoint (SuRE equals pseudocount)
RM1<-RM[0,] #empty df
RM1[1,]<-c(rep(log10(Pseud),3), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==1]), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==0]))
RM<-rbind(RM1, RM)
rm(RM1)

#finally calculate LRS for all promoters:
P$LRS<- P$GROcap - approx(x=RM$SuRE.mean, y=RM$GROcap.ilad, xout=P$SuRE, rule=2)$y
#so the more negative the score, the more 'repressed' is the promoter by its chromatin/LAD context


#(arbitrary) cutoffs to define three groups of promoters:
INACT<- P$SuRE< -0.3 & P$LAD==1 & P$GROcap< -2 #inactive
NREP<- P$SuRE> 0 & P$LRS> -0.5 & P$LAD==1 & P$GROcap> -2 #not repressed
REP<- P$SuRE> 0.3 & P$LRS< -1 & P$LAD==1  & P$GROcap< -2 #repressed
Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
names(Pcnts)<-c("repressed", "escaper", "inactive")
BND <- P$LAD==1 & !INACT & !NREP & !REP

#add class annotation column to P:
P$class<-NA
P$class[P$LAD==0]<-"iLAD"
P$class[INACT]<-"inactive"
P$class[NREP]<-"escaper"
P$class[REP]<-"repressed"
P$class[BND] <- "boundary"
P$class = factor(P$class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary'))

COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")

class_names = paste0(levels(P$class), '; n=',table(P$class))
names(class_names) = levels(P$class)
P$class_n_prom = P$class
levels(P$class_n_prom) = class_names
COL_class_n_prom = COL_class[names(class_names)]
names(COL_class_n_prom) = class_names

lad_names = c(LAD=paste0('LAD; n=', table(P$LAD)['1']),
              iLAD=paste0('LAD; n=', table(P$LAD)['0']))
P$lad_n_prom = factor(ifelse(P$LAD==1, lad_names['LAD'], lad_names['iLAD']))
COL_lad_n = COL_lad
names(COL_lad_n) = lad_names


RM_melt = melt(RM, measure.vars=c('GROcap.ilad', 'GROcap.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='GROcap.lad', lad_names['LAD'], lad_names['iLAD'])

```

```{r}
tss_gr = makeGRangesFromDataFrame(data.frame(seqnames = P$chr,
                                             start = P$tss,
                                             end = P$tss,
                                             row.names = P$name,
                                             strand=P$strand))
export.bed(tss_gr, '../raw_data/tss.bed')


```

```{r}
count_rep1 = read.table('../raw_data/expression/K562_rep1ReadsPerGene.out.tab')
count_rep2 = read.table('../raw_data/expression/K562_rep2ReadsPerGene.out.tab')

count_table = cbind(count_rep1[,4], rep2=count_rep2[,4])
rownames(count_table) = count_rep1[,1]
colnames(count_table) = c('rep1', 'rep2')
exp = SummarizedExperiment(assays = list(counts=count_table[-(1:4), ]))
dds = DESeqDataSet(exp, design= ~ 1)
fpm = fpm(dds)
fpm = rowMeans(fpm)
fpm = log10(fpm + min(fpm[fpm!=0])/2)

g_match = match(P$ensembl_gene_id, names(fpm))
P$K562_fpm = NaN
P$K562_fpm[!is.na(g_match)] = fpm[g_match[!is.na(g_match)]]


```


```
liftOver raw_data/tss.bed ~/data/hg19ToHg38.over.chain.gz /dev/stdout raw_data/tss_unmapped.bed | awk '{sub($2,$2-300,$0);sub($3,$3+300,$0);print $0}' > raw_data/tssr_300_300_hg38.bed
cp raw_data/tssr_300_300.bed raw_data/tssr_300_300_hg19.bed

# min_count="/home/NFS/users/c.leemans/data/ChIP/TF/min_count.txt"
# rm min_count
# for file in $(tail -n+2 ~/data/ChIP/TF/META_INFO.tsv | awk -F'\t' '{print $1}')
# do
#     bigWigToWig ~/data/ChIP/TF/$file /dev/stdout | \
#         awk -v F=$file '
#             BEGIN{
#                 min=3
#             }{
#                 if ($4 < min && $4 !=0 && $4 != "") {
#                     min=$4
#                 }
#             }
#             END {
#                 print F"\t"min
#             }' >> $min_count
# done


for line in $(tail -n+2 ~/data/ChIP/TF/META_INFO.tsv | awk -F'\t' '{print $1">"$4">"$7}')
do
    IFS='>' read -r -a array <<< "$line"
    file=${array[0]}
    name=${array[1]}
    genome=raw_data/tssr_300_300_${array[2]}.bed

    bwtool summary -decimal=4 -skip-median -keep-bed -header $genome /home/NFS/users/c.leemans/data/ChIP/TF/$file /dev/stdout | gzip -c > raw_data/tssr_ChIP_300_300/$name.txt.gz
done

# for line in $(tail -n+2 ~/data/ChIP/TF/META_INFO.tsv | awk -F'\t' '{print $1">"$4">"$7}')
# do
#     IFS='>' read -r -a array <<< "$line"
#     file=${array[0]}
#     name=${array[1]}
#     genome=raw_data/tssr_300_300_${array[2]}.bed
#     pseudo=$(grep $file ~/data/ChIP/TF/min_count.txt | awk '{print $2/2}')
#     bwtool extract -decimals=5 -tabs bed $genome /home/NFS/research_projects/van_steensel_lab/c.leemans/ChIP/TF/$file /dev/stdout | \
#         awk -v pseudo=$pseudo '{
#             logSum=0
#             for(i = 8; i <= NF; i++) {
#                 logSum+=log($i + pseudo)
#             }
#             print $8
#             print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"logSum/$7
#         }' | gzip -c > raw_data/tssr_ChIP_300_300/$name.txt.gz
# done



```


```{r}
tissues_expressed = read.table('../../../data/fantom/max_tissues_expressed.txt.gz',
                               sep='\t', header=T, stringsAsFactors=T)

aff_table_jaspar = read.table('../raw_data/jaspar_affinity_pseudo/seq_psam.dat',
                              stringsAsFactors=F)
colnames(aff_table_jaspar) = gsub('.xml','',colnames(aff_table_jaspar))
tf_table_jaspar = read.table('../raw_data/tf_table.txt', sep='\t', row.names=1, stringsAsFactors=F)
colnames(tf_table_jaspar) = c('name', 'species', 'class', 'family')
tf_translation = read.table('../raw_data/tf_translation.txt', stringsAsFactors=F)

tf_translation$K562_fpm = fpm[tf_translation$gene_id]


KBM7_essential = read.table('../raw_data/KBM7_essentialome_aac7557_SM_Table_S1.csv',
                            stringsAsFactors=F, skip=1, header=T, sep='\t',
                            row.names=2)
HAP1_essential = read.table('../raw_data/HAP1_essentialome_aac7557_SM_Table_S2.csv',
                            stringsAsFactors=F, skip=1, header=T, sep='\t',
                            row.names=2)
tf_translation$KBM7_essential = KBM7_essential[tf_translation$gene_id, 'selected']=='YES'
tf_translation$HAP1_essential = HAP1_essential[tf_translation$gene_id, 'selected']=='YES'

tf_expression = ddply(tf_translation,.(name),
                      function(x, te){
                          name=x$name[1]
                          ens = which(te$ensembl_gene_id%in%x$gene_id)
                          if (length(ens) > 0){
                            tissues= min(te[ens,'tissues_expressed'])
                          } else {
                            tissues = NA
                          }
                          c(expression=min(x$K562_fpm),
                            KBM7_essential=any(x$KBM7_essential),
                            HAP1_essential=any(x$HAP1_essential),
                            tissues_expressed=tissues)
                      }, te=tissues_expressed)
tf_table_jaspar$expression = NaN
tf_match = match(tf_expression$name, tf_table_jaspar$name)
tf_table_jaspar$expression[tf_match] = tf_expression$expression
tf_table_jaspar$tissues_expressed = NaN
tf_table_jaspar$tissues_expressed[tf_match] = tf_expression$tissues_expressed
tf_table_jaspar$KBM7_essential = NA
tf_table_jaspar$KBM7_essential[tf_match] = tf_expression$KBM7_essential
tf_table_jaspar$HAP1_essential = NA
tf_table_jaspar$HAP1_essential[tf_match] = tf_expression$HAP1_essential


encode_table = read.table('../../../data/ChIP/TF/META_INFO.tsv', sep='\t',
                          stringsAsFactors=F, header=T)
signal_list = lapply(paste0('../raw_data/tssr_ChIP_300_300/', encode_table$Target.gene,
                            '.txt.gz'),
                     read.table, stringsAsFactors=F, row.names=4, 
                     col.names=c('chrom', 'start', 'end', 'name','score', 'strand', 'size',
                                 'num_data', 'min', 'max', 'mean'))

pdf('cl20170530_motif_affinity_ChIP_correlation.pdf', width = 100, height=100)
plot_list = list()
for (i in 1:nrow(encode_table)){
    symbol = encode_table$Target.gene[i]
    pattern = sprintf('^%s$|^%s::|::%s$', symbol, symbol, symbol)
    j_vec = grep(pattern, toupper(tf_table_jaspar$name))
    for (j in j_vec){
        jaspar_id = rownames(tf_table_jaspar)[j]
        data = data.frame(motif_affinity=rank(aff_table_jaspar[P$name,jaspar_id]),
                          chip_score=signal_list[[i]][P$name,'mean'],
                          class=P$class)
        cor_data = data.frame(cor=vector("numeric",length=4), 
                              class=c('iLAD', 'escaper', 'repressed', 'inactive'),
                              row.names=c('iLAD', 'escaper', 'repressed', 'inactive'))
        for (class in cor_data$class){
          cor_data[class, 'cor'] = cor(data$chip_score[which(data$class==class)],
                                       data$motif_affinity[which(data$class==class)])
        }
        if (!is.na(cor_data['iLAD', 'cor']) & cor_data['iLAD', 'cor'] > 0.1){
          title = sprintf('antibody: %s\njaspar name: %s\njaspar id: %s',
                          encode_table$Target.label[i],
                          tf_table_jaspar[j, 'name'],
                          jaspar_id)
          plot_list[[title]] = ggplot(cor_data, aes(x=class, y=cor, fill=class)) +
                      geom_bar(stat='identity') + 
                      ggtitle(title) +
                      scale_fill_manual(values=COL_class)
        }
      }
    }
do.call(grid.arrange, plot_list)
dev.off()
```