# knitr document van Steensel lab

# Gene repression in LADs
## Christ Leemans, 03-11-2016 - to date

## Introduction
Generally speaking, genes inside lamina associated domains are not or very lowly expressed. These genes can either be actively repressed by their DNA context (e.g. heterochromatin, lamina association), or simply be inactive (because essential factors for expression are missing?). Yet another group of genes seem to evade gene repression in the context of lamina associated domains. In this report I would like to investigate what defines these 3 groups of genes and how they compare to genes outside of lamina associated domains.

## Description of Data.

gencode.sure.160617.rda:
    file from Joris, received 17 June 2016. Promoter positions in this file are from Gencode. Contains SuRE, gro-cap and cage expression data as well as the number of tissues in which each promoter is expressed.


## libraries, paths and data prep

```{r, fig.width=10, fig.height=10}
library(reshape2)
library(rtracklayer)
library(ggplot2)
library(gridExtra)
library(plyr)
library(preprocessCore)
library(scales)
library(gage)
library(fgsea)
library(CGtools)
library(biomaRt)
library(Matrix)
library(grid)
load('../raw_data/biomart.rdata')

## There was one promoter that was wrongly annotated
bm_p[bm_p$ensembl_transcript_id=='ENST00000357491','ensembl_gene_id' ] = 'ENSG00000196350'

```


# MAIN

### SuRE data
Previously, Bas defined three different groups of promoters in LADs. The "escaper", "repressed" and "inactive" group. The "escaper" group contains promoters that are expressed in their endogenous LAD environment. The "repressed" group is inactive in the endogenous setting, but they do show autonomous promoter activity in SuRE data. The 'inactive' group shows neither endogenous expression nor autonomous promoter activity in SuRE data.

```{r, fig.width=10, fig.height=8}
load("../raw_data/gencode.sure.160617.rda")
Prom<-gencode.sure.160617; rm(gencode.sure.160617) #simpler name
#first re-calculate pseudocounts without jitter
P<-Prom[,c(1:8,23,20,26,28, 27)] #SuRE, GRO-cap, CAGE and LAD columns only
names(P)[9:13]<-c("SuRE", "GROcap", "CAGE", "LAD", 'tissues_expressed')


## for promoters and gene expression let's convert promoter transcript id's to gene id's
P$ensembl_transcript_id = do.call(rbind, strsplit(P$name, split='[.]'))[,1]

nrow(P) #orriginal number of rows
bm_match = match(P$ensembl_transcript_id, bm_p$ensembl_transcript_id)
P<-merge(P, bm_p, by="ensembl_transcript_id", all.x=TRUE)
nrow(P) #some double rows were introduced

P = P[match(Prom$name, P$name), ]

length(unique(P$ensembl_gene_id)) #number of unique genes

table(P[,c('strand.x','strand.y')]) #almost all strand listings are consistent

P<-P[, colnames(P)!='strand.y']
colnames(P)[colnames(P)=='strand.x'] = "strand"

## to be used by CGtools as the complete set of TSS's
peaks = data.frame(seqname=P$chr,
                   start=P$tss,
                   end=P$tss,
                   strand=P$strand)


Pseud<-min(P$SuRE[P$SuRE>0], na.rm=TRUE)/2
P$SuRE<-P$SuRE+Pseud
P$SuRE<-log10(P$SuRE)
PseudGro<-min(P$GROcap[P$GROcap>0], na.rm=TRUE)/2
P$GROcap<-P$GROcap+PseudGro
P$GROcap<-log10(P$GROcap)
PseudCage<-min(P$CAGE[P$CAGE>0], na.rm=TRUE)/2
P$CAGE<-P$CAGE+PseudCage
P$CAGE<-log10(P$CAGE)

#then calculate running mean for iLAD promoters:
P<-P[order(P$SuRE,sample(c(1:nrow(P)))),] #sort by SuRE and then random for ties
n<-60 #number of windows
w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
s<-round(seq(from=w/2+0.0001, to=nrow(P)-w/2, length.out=n))
RM<-data.frame(SuRE.low=rep(NA,n), SuRE.mean=rep(NA,n), SuRE.hi=rep(NA,n), GROcap.lad=rep(NA,n), GROcap.ilad=rep(NA,n))
RM$SuRE.low=P$SuRE[s-floor(w/2)]
for(i in 1:n){RM$SuRE.mean[i]=mean(P$SuRE[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
RM$SuRE.hi=P$SuRE[s+floor(w/2)]
for(i in 1:n)
  {t<-P[(s[i]-floor(w/2)):(s[i]+floor(w/2)),]
   RM$GROcap.lad[i]<-mean(t$GROcap[t$LAD==1], na.rm=TRUE)
   RM$GROcap.ilad[i]<-mean(t$GROcap[t$LAD==0], na.rm=TRUE)
  }


#add first datapoint (SuRE equals pseudocount)
RM1<-RM[0,] #empty df
RM1[1,]<-c(rep(log10(Pseud),3), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==1]), mean(P$GROcap[P$SuRE==log10(Pseud) & P$LAD==0]))
RM<-rbind(RM1, RM)
rm(RM1)

#finally calculate LRS for all promoters:
P$LRS<- P$GROcap - approx(x=RM$SuRE.mean, y=RM$GROcap.ilad, xout=P$SuRE, rule=2)$y
#so the more negative the score, the more 'repressed' is the promoter by its chromatin/LAD context


#(arbitrary) cutoffs to define three groups of promoters:
INACT<- P$SuRE< -0.3 & P$LAD==1 & P$GROcap< -2 #inactive
NREP<- P$SuRE> 0 & P$LRS> -0.5 & P$LAD==1 & P$GROcap> -2 #not repressed
REP<- P$SuRE> 0.3 & P$LRS< -1 & P$LAD==1  & P$GROcap< -2 #repressed
Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
names(Pcnts)<-c("repressed", "escaper", "inactive")
BND <- P$LAD==1 & !INACT & !NREP & !REP

#add class annotation column to P:
P$class<-NA
P$class[P$LAD==0]<-"iLAD"
P$class[INACT]<-"inactive"
P$class[NREP]<-"escaper"
P$class[REP]<-"repressed"
P$class[BND] <- "boundary"
P$class = factor(P$class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary'))

COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")

class_names = paste0(levels(P$class), '; n=',table(P$class))
names(class_names) = levels(P$class)
P$class_n_prom = P$class
levels(P$class_n_prom) = class_names
COL_class_n_prom = COL_class[names(class_names)]
names(COL_class_n_prom) = class_names

lad_names = c(LAD=paste0('LAD; n=', table(P$LAD)['1']),
              iLAD=paste0('LAD; n=', table(P$LAD)['0']))
P$lad_n_prom = factor(ifelse(P$LAD==1, lad_names['LAD'], lad_names['iLAD']))
COL_lad_n = COL_lad
names(COL_lad_n) = lad_names


RM_melt = melt(RM, measure.vars=c('GROcap.ilad', 'GROcap.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='GROcap.lad', lad_names['LAD'], lad_names['iLAD'])
ggplot(P, aes(x=SuRE, y=GROcap, color=lad_n_prom)) +
    geom_point(data=P[P$LAD==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD==1, ], size=0.5, alpha=0.2) +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=SuRE.mean, y=value, color=variable), size=1) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_n)
p_classes = P[which(P$class %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P, aes(x=SuRE, y=GROcap)) +
    geom_point(size=0.1,color=COLi) +
    geom_point(data=p_classes, aes(color=class_n_prom), size=0.6) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme_bw() +
    geom_line(data=RM, aes(x=SuRE.mean, y=GROcap.ilad), color=COL_lad['iLAD']) +
    theme(legend.title=element_blank()) +
    scale_colour_manual(values=COL_class_n_prom)

```



```{r}
#genes as Granges object
gene_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                              start=P$txStart,
                                              end=P$txEnd,
                                              strand=P$strand,
                                              tss=P$tss,
                                              name=P$name,
                                              gene_id = P$ensembl_gene_id),
                                              keep.extra.columns=TRUE)
tss_gr = gene_gr
ranges(tss_gr) = IRanges(gene_gr$tss,
                         gene_gr$tss)

TSSR_gr = gene_gr
P_start = P$tss - ifelse(P$strand=='+',50,300)
P_start = ifelse(P_start<1,1,P_start)
P_end = P$tss + ifelse(P$strand=='+',300,50)
ranges(TSSR_gr) = IRanges(P_start, P_end)

body_gr = gene_gr
P_start = ifelse(P$strand=='+',
                 P$tss + 300,
                 P$txStart - 3000)
P_start = ifelse(P_start<1,1,P_start)
P_end = ifelse(P$strand=='+',
               P$txEnd + 3000,
               P$tss - 300)
ranges(body_gr) = IRanges(P_start, P_end)

F_window=22000

window_gr = gene_gr[width(gene_gr)>F_window]
p_match = match(window_gr$name, P$name)
P_start = P[p_match, 'tss'] - F_window
P_start = ifelse(P_start<1,1,P_start)
P_end = P[p_match, 'tss'] + F_window
ranges(window_gr) = IRanges(P_start, P_end)


p_most_active = ddply(P, .(ensembl_gene_id), function(x){
    if (nrow(x)==1){
        result = x[1,]
    } else {
        result = x[order(x$GROcap, decreasing=T)[1],]
    }
    return(result)
  })

m_active = match(p_most_active$name, P$name)


p_most_down = ddply(P[!is.na(P$ensembl_gene_id),], .(ensembl_gene_id), function(x){
    if (nrow(x)==1){
        result = x[1,]
    } else if (all(x$strand=='+')){
        result = x[order(x$tss, decreasing=T)[1],]
    } else if (all(x$strand=='-')){
        result = x[order(x$tss)[1],]
    } else {
        print('huh???')
        print(x)
        result = x[1,]
    }
    return(result)
  })

o = findOverlaps(tss_gr, gene_gr, maxgap=20000, ignore.strand=T)
o = o[which(P[to(o), 'ensembl_gene_id']!=P[from(o), 'ensembl_gene_id'])]

ensg_too_close = P[from(o), 'ensembl_gene_id']

m_most_down = match(p_most_down$name, P$name)
m_most_down_window = match(p_most_down$name, window_gr$name)


p_most_up = ddply(P[!is.na(P$ensembl_gene_id),], .(ensembl_gene_id), function(x){
    if (nrow(x)==1){
        result = x[1,]
    } else if (all(x$strand=='+')){
        result = x[order(x$tss, decreasing=F)[1],]
    } else if (all(x$strand=='-')){
        result = x[order(x$tss)[1],]
    } else {
        print('huh???')
        print(x)
        result = x[1,]
    }
    return(result)
  })
m_most_up = match(p_most_up$name, P$name)
m_most_up_window = match(p_most_up$name, window_gr$name)

## choose -/+ 1000 based on doi: 10.4161/epi.19565
F_cpg = 1000
cpg_gr = gene_gr
p_match = match(cpg_gr$name, P$name)
P_start = P[p_match, 'tss'] - F_cpg
P_start = ifelse(P_start<1,1,P_start)
P_end = P[p_match, 'tss'] + F_cpg
ranges(cpg_gr) = IRanges(P_start, P_end)

window_gr_3prime = gene_gr[width(gene_gr)>F_window]
p_match = match(window_gr_3prime$name, P$name)
P_start = ifelse(P[p_match, 'strand'] =='+',
                 P[p_match, 'txEnd'] - F_window,
                 P[p_match, 'txStart'] - F_window)
P_start = ifelse(P_start<1,1,P_start)
P_end = ifelse(P[p_match, 'strand'] =='+',
               P[p_match, 'txEnd'] + F_window,
               P[p_match, 'txStart'] + F_window)
ranges(window_gr_3prime) = IRanges(P_start, P_end)
export.bed(window_gr_3prime, '../raw_data/3prime_window.bed')


F_cpg = 300
cpg_gr = gene_gr
p_match = match(cpg_gr$name, P$name)
P_start = P[p_match, 'tss'] - F_cpg
P_start = ifelse(P_start<1,1,P_start)
P_end = P[p_match, 'tss'] + F_cpg
ranges(cpg_gr) = IRanges(P_start, P_end)
export.bed(cpg_gr, '../raw_data/cpg_window_300.bed')

```



```
bedtools nuc -C -pattern CG -fi ~/data/hg19/genome.fa -bed raw_data/cpg_window.bed | \
awk '
{
    CGdinuc=$16
    CGsum=$10 + $11
    CGE=((CGsum/2)*(CGsum/2))/$15
    print $1"\t"$2"\t"$3"\t"$4"\t"CGdinuc"\t"CGsum"\t"CGE"\t"CGdinuc/CGE
}' > raw_data/cpg_density_1000.txt

bedtools nuc -C -pattern CG -fi ~/data/hg19/genome.fa -bed raw_data/cpg_window_300.bed | \
awk '
{
    CGdinuc=$16
    CGsum=$10 + $11
    CGE=((CGsum/2)*(CGsum/2))/$15
    print $1"\t"$2"\t"$3"\t"$4"\t"CGdinuc"\t"CGsum"\t"CGE"\t"CGdinuc/CGE
}' > raw_data/cpg_density_300.txt

bedtools getfasta -name -tab -fi ~/data/hg19/genome.fa -bed raw_data/prom_window.bed -fo /dev/stdout | awk '{
    line=tolower($2)
    i_line=""
    j=0
    for (i=1;i<44000;i=i+100){
        sub_line = substr(line, i, 100)
        i_line=i_line"\t"gsub(/cg/, "", sub_line)
    }
    print $1 i_line
}' | gzip -c > raw_data/cpg_window.txt.gz
```


```{r}
cpg = read.table('../raw_data/cpg_density_300.txt', stringsAsFactors=F)
colnames(cpg) = c('seqnames', 'start','stop','name','GC-dinuc', 'CGsum', 'CG_exp', 'CG_OE')
cpg_match = match(P$name, cpg$name)
P$CG_OE = NaN
P$CG_OE = cpg[cpg_match,'CG_OE']

P_class = P[P$class%in%names(COL), ]
ggplot(P[P$class!='boundary', ], aes(x=class, y=CG_OE, color=class)) + geom_violin(alpha=0.3) +
    geom_point(data=P_class, position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('CpG observed over expected') +
    ggtitle('all CpG in < 300b from TSS') +
    scale_color_manual(values=COL_class)

p_expressed = which(P$class %in% c('escaper', 'iLAD'))
s_escaper = which(P[p_expressed, 'class']=='escaper')
gro_order = order(P[p_expressed,'GROcap'])
gro_escaper = which(gro_order %in% s_escaper)

gro_ilad = gro_order[c(gro_escaper + 1, gro_escaper - 1)]
s_ilad = gro_ilad[P[p_expressed,'class'][gro_ilad] == 'iLAD']

norm_grocap_active = c(rownames(P[p_expressed,])[s_escaper],
                rownames(P[p_expressed,])[s_ilad])
ggplot(P[norm_grocap_active,] , aes(x=class, y=CG_OE, color=class)) +
    geom_violin(alpha=0.3) +
    ggtitle('all CpG in < 300b from TSS') +
    geom_point(position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('GC frequency') +
    scale_color_manual(values=COL_class)


cpg = read.table('../raw_data/cpg_density_1000.txt', stringsAsFactors=F)
colnames(cpg) = c('seqnames', 'start','stop','name','GC-dinuc', 'CGsum', 'CG_exp', 'CG_OE')
cpg_match = match(P$name, cpg$name)
P$CG_OE = NaN
P$CG_OE = cpg[cpg_match,'CG_OE']

P_class = P[P$class%in%names(COL), ]
ggplot(P[P$class!='boundary', ], aes(x=class, y=CG_OE, color=class)) + geom_violin(alpha=0.3) +
    geom_point(data=P_class, position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ggtitle('all CpG in < 1000b from TSS') +
    ylab('CpG observed over expected') +
    scale_color_manual(values=COL_class)

p_expressed = which(P$class %in% c('escaper', 'iLAD'))
s_escaper = which(P[p_expressed, 'class']=='escaper')
gro_order = order(P[p_expressed,'GROcap'])
gro_escaper = which(gro_order %in% s_escaper)

gro_ilad = gro_order[c(gro_escaper + 1, gro_escaper - 1)]
s_ilad = gro_ilad[P[p_expressed,'class'][gro_ilad] == 'iLAD']

norm_grocap_active = c(rownames(P[p_expressed,])[s_escaper],
                rownames(P[p_expressed,])[s_ilad])
ggplot(P[norm_grocap_active,] , aes(x=class, y=CG_OE, color=class)) +
    geom_violin(alpha=0.3) +
    ggtitle('all CpG in < 1000b from TSS') +
    geom_point(position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('GC frequency') +
    scale_color_manual(values=COL_class)
```

```{r}
cpg_table = read.table('../raw_data/cpg_window.txt.gz')
cpg_table = cpg_table[match(unique(cpg_table[,1]), cpg_table[,1]),]
rownames(cpg_table) = cpg_table[,1]
cpg_table = cpg_table[,-1]
this_p = P[which(P$name%in%rownames(cpg_table)), ]
sample_n = table(this_p[,'class'])['escaper']
plot_list = list()
for(i in c('iLAD', 'escaper', 'repressed', 'inactive')) #for the three gene classes
   {
    s = which(this_p[, 'class']==i)
    strand = this_p[,'strand'][s]
    name_vec = this_p[, 'name'][s]
    s = match(name_vec, rownames(cpg_table))
    mean_list = lapply(seq(1,440,4), function(x){
                           c(x*100-50, sum(cpg_table[s,x:(x+3)])/length(s)/4)
                       })
    mean = do.call(rbind.data.frame, mean_list)
    colnames(mean) = c('x', 'y')
    #plot:
    plot_list[[i]] = ggplot(mean, aes(x=x-F_window, y=y)) + geom_line(col=COL_class[i]) +
                        ylim(0,10) +
                        ggtitle(paste(i, '; n=', length(s))) +
                        geom_vline(xintercept=0, lty='dotted') +
                        theme_bw() +
                        theme(axis.title=element_blank()) +
                        scale_x_continuous(breaks=seq(-F_window, F_window, F_window))
}
do.call(grid.arrange, c(plot_list, nrow=1))

```

```{r}

cpg_table = read.table('../raw_data/cpgIslandExtUnmasked.bed.gz', sep='\t',
                       stringsAsFactors=F)
colnames(cpg_table) = c('seqnames', 'start', 'end', 'name', 'length', 'cpgNum', 'gcNum', 'perCpg', 'perGC', 'obsExp')
cpg_ranges = makeGRangesFromDataFrame(cpg_table, keep.extra.columns= T)

o = findOverlaps(tss_gr, cpg_ranges, maxgap=500, ignore.strand=T)
o_frame = data.frame(o)
o_frame$score = cpg_ranges[to(o)]$cpgNum
o_frame$pos = start(tss_gr[from(o)]) - start(cpg_ranges[to(o)])

maxCpG = ddply(o_frame, .(queryHits), function(x){
          if (nrow(x) > 1){
            result = x[which(x$score==max(x$score)),]
          } else {
            result = x[1,]
          }
          return(result)})


P$cpg_length = 0
P$cpg_length[maxCpG$queryHits] = cpg_ranges[maxCpG$subjectHits]$length
P$cpg_length = P$cpg_length + min(cpg_ranges[maxCpG$subjectHits]$length) / 2
P$cpg_num = 0
P$cpg_num[maxCpG$queryHits] = cpg_ranges[maxCpG$subjectHits]$cpgNum
P$cpg_num = P$cpg_num + min(cpg_ranges[maxCpG$subjectHits]$cpgNum) / 2

P$perGC = 0
P$perGC[maxCpG$queryHits] = cpg_ranges[maxCpG$subjectHits]$perGC
# P$perGC = P$perGC + min(cpg_ranges[maxCpG$subjectHits]$perGC) / 2
P$perCpg = 0
P$perCpg[maxCpG$queryHits] = cpg_ranges[maxCpG$subjectHits]$perCpg

P$obsExp = 0
P$obsExp[maxCpG$queryHits] = cpg_ranges[maxCpG$subjectHits]$obsExp

ggplot(P[P$class!='boundary', ] , aes(x=class, y=log2(cpg_num), color=class)) +
    geom_violin(alpha=0.3) +
    geom_point(data=P[P$class%in%names(COL), ],  position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('log2(number of CpG in island) (biggest island < 500b)') +
    scale_color_manual(values=COL_class)

ggplot(P[P$class!='boundary', ] , aes(x=class, y=perGC, color=class)) +
    geom_violin(alpha=0.3) +
    geom_point(data=P[P$class%in%names(COL), ],  position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('percentage of C and G in island (biggest island < 500b)') +
    scale_color_manual(values=COL_class)

ggplot(P[P$class!='boundary', ] , aes(x=class, y=log2(perCpg), color=class)) +
    geom_violin(alpha=0.3) +
    geom_point(data=P[P$class%in%names(COL), ],  position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('log2(percentage of CpG in island) (biggest island <500b)') +
    scale_color_manual(values=COL_class)

ggplot(P[P$class!='boundary', ] , aes(x=class, y=obsExp, color=class)) +
    geom_violin(alpha=0.3) +
    geom_point(data=P[P$class%in%names(COL), ],  position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('CpG island OE score (biggest island < 500b)') +
    scale_color_manual(values=COL_class)


ggplot(P[P$class!='boundary', ] , aes(x=class, y=log2(cpg_length), color=class)) +
    geom_violin(alpha=0.3) +
    geom_point(data=P[P$class%in%names(COL), ],  position=position_jitter(width=0.5), alpha=0.3, size=0.3) +
    ylab('log2(length CpG island) (biggest island < 500b)') +
    scale_color_manual(values=COL_class)
tss_gr$cpg_distance = mcols(distanceToNearest(tss_gr, cpg_ranges))$distance

P$cpg_class = NA
cpg_prom = tss_gr[which(tss_gr$cpg_distance < 1000)]$name

non_cpg_prom = tss_gr[which(tss_gr$cpg_distance >= 1000)]$name
P$cpg_class[P$name %in% cpg_prom] = 'CpG'
P$cpg_class[P$name %in% non_cpg_prom] = 'non_CpG'
plot_list = list()
for (class in c('iLAD', 'escaper', 'repressed', 'inactive')){
    subset = P[which(P$class==class),]
    n = paste('n =', nrow(subset))
    plot_list[[class]] = ggplot(subset, aes(x=cpg_class, fill=cpg_class)) +
                                theme_bw() +
                                guides(fill=FALSE) +
                                geom_bar(aes(y = (..count..)/sum(..count..))) +
                                scale_y_continuous(labels=percent, limits=c(0,1)) +
                                ggtitle(class) +

                                geom_text(aes(y = ((..count..)/sum(..count..)),
                                              label = paste0('n=',..count..)),
                                          stat = "count", vjust = -0.25) +
                                theme(axis.title=element_blank(),
                                      plot.title = element_text(color=COL_class[class], size=14, face="bold.italic"))
}
grid.arrange(plot_list[['iLAD']], plot_list[['escaper']], plot_list[['repressed']], plot_list[['inactive']], top=textGrob("ratio between cpg and\nnon-cpg promoter classes",gp=gpar(fontsize=20,font='bold')), nrow=1)
grid.arrange(plot_list[['iLAD']], plot_list[['escaper']], plot_list[['repressed']], plot_list[['inactive']], top=textGrob("ratio between cpg and\nnon-cpg promoter classes",gp=gpar(fontsize=20,font='bold')))

tss_gr$cpg_distance = mcols(distanceToNearest(tss_gr, cpg_ranges[cpg_ranges$score>50]))$distance

P$cpg_class_stringent = NA
cpg_prom = tss_gr[which(tss_gr$cpg_distance < 1000)]$name

non_cpg_prom = tss_gr[which(tss_gr$cpg_distance >= 1000)]$name
P$cpg_class_stringent[P$name %in% cpg_prom] = 'CpG'
P$cpg_class_stringent[P$name %in% non_cpg_prom] = 'non_CpG'
plot_list = list()
for (class in c('iLAD', 'escaper', 'repressed', 'inactive')){
    subset = P[which(P$class==class),]
    n = paste('n =', nrow(subset))
    plot_list[[class]] = ggplot(subset, aes(x=cpg_class_stringent, fill=cpg_class_stringent)) +
                                theme_bw() +
                                guides(fill=FALSE) +
                                geom_bar(aes(y = (..count..)/sum(..count..))) +
                                scale_y_continuous(labels=percent, limits=c(0,1)) +
                                ggtitle(class) +

                                geom_text(aes(y = ((..count..)/sum(..count..)),
                                              label = paste0('n=',..count..)),
                                          stat = "count", vjust = -0.25) +
                                theme(axis.title=element_blank(),
                                      plot.title = element_text(color=COL_class[class], size=14, face="bold.italic"))
}
grid.arrange(plot_list[['iLAD']], plot_list[['escaper']], plot_list[['repressed']], plot_list[['inactive']], top=textGrob("ratio between cpg and\nnon-cpg promoter classes",gp=gpar(fontsize=20,font='bold')), nrow=1)
grid.arrange(plot_list[['iLAD']], plot_list[['escaper']], plot_list[['repressed']], plot_list[['inactive']], top=textGrob("ratio between cpg and\nnon-cpg promoter classes",gp=gpar(fontsize=20,font='bold')))
```



```{r}

CpG_over_window = matrix(nrow=nrow(P), ncol = 440, dimnames=list(rownames(P)))
for(j in 1:ncol(CpG_over_window)){
  this_start = j * 100-50
  this_end = j *100+ 50
  y = lapply(1:nrow(o_frame), function(x){
      if (o_frame[x,'rel_start'] < this_end &
          o_frame[x,'rel_end'] > this_start){
        return(min(c(this_end - o_frame[x,'rel_start'],
                     o_frame[x,'rel_end'] - this_start,
                     100)))
      } else{
        return(NA)
      }
    })
  y = unlist(y)
  CpG_over_window[from(o)[!is.na(y)], j] = y[!is.na(y)]
}

plot_list = list()
for(i in c('iLAD', 'escaper', 'repressed', 'inactive')) #for the three gene classes
   {
    s = which(P[, 'class']==i)
    strand = P[,'strand'][s]
    this_window = rbind(CpG_over_window[s[strand=='+'], ],
                        CpG_over_window[s[strand=='-'],440:1])

    mean_list = lapply(seq(1,440,4), function(x){
                           c(x*100-50, sum(this_window[,x:(x+3)], na.rm=T)/length(s)/4/100)
                       })
    mean = do.call(rbind.data.frame, mean_list)
    colnames(mean) = c('x', 'y')
    #plot:
    plot_list[[i]] = ggplot(mean, aes(x=x-F_window, y=y)) + geom_line(col=COL_class[i]) +
                        ggtitle(paste(i, '; n=', length(s))) + ylim(0,1) +
                        geom_vline(xintercept=0, lty='dotted') +
                        theme_bw() +
                        theme(axis.title=element_blank()) +
                        scale_x_continuous(breaks=seq(-F_window, F_window, F_window))
}
do.call(grid.arrange, c(plot_list, nrow=1))

```
