
# knitr document van Steensel lab

# Thethered TRIP
## Christ Leemans, 31-05-2016 - to date 

## Introduction
Laura performed a thetered-TRIP experiment on a K562 cell pool. She transiently transfected the cell pool with GAL4 thetered to 3 different "proteins of interest"(POI), namely G9a, the KRAB domain and CBX5. She took measurements after day 2 to see the initial effect of the thetering and after 12, 11 and 9 days respectively to see if there was any memory of the silencing on day 2. Besides the thetered GAL4-POI, two seperate controlls were taken using unthetered GAL4 and unthethered POI.

## Experimental setup
At this moment Laura has data for 3 different tethering experiments using KRAB, G9a and CBX5. For each protein of interest (POI) there are 12 expression and 12 gDNA files: 3 conditions * 2 different days after induction * 2 replicates. One condition uses an unthethered POI, the second uses only GAL4 and the third condition uses the POI thethered to GAL4 (GAL4-POI). Expression and gDNA data was obtained on day 2 and day 9. With each sequencing run, spikeins were added to normalize across different experiments. There is a different config file to extract the expression values of the spikeins.

## Description of data
For this part of the analysis I used the data that was generated previously in the data preperation report: cl20160713_TRIP_K562_data_preperation. 
The .RData produced contains a large data frame with a row for each barcode and the following columns:

| column name | description |
| ----------- | ----------- |
| CBX_GAL4_exp_2 | expression after thetering GAL4 on day 2  for CBX5 experiment |
| CBX_GAL4_exp_9 | expression after thetering GAL4 on day 9 for CBX5 experiment |
| CBX_GAL4.POI_exp_2 | expression after thetering GAL4-CBX5 on day 2 |
| CBX_GAL4.POI_exp_9 | expression after thetering GAL4-CBX5 on day 9 |
| CBX_POI_exp_2 | expression after thetering CBX5 on day 2 |
| CBX_POI_exp_9 | expression after thetering CBX5 on day 9 |
| G9a_GAL4_exp_12 | expression after thetering GAL4 on day 12 for G9a experiment |
| G9a_GAL4_exp_2 | expression after thetering GAL4 on day 2 for G9a experiment |
| G9a_GAL4.POI_exp_12 | expression after thetering GAL4-G9a on day 12 |
| G9a_GAL4.POI_exp_2 | expression after thetering GAL4-G9a on day 2 |
| G9a_POI_exp_12 | expression after thetering G9a on day 12 |
| G9a_POI_exp_2 | expression after thetering G9a on day 2 |
| KRAB_GAL4_exp_11 | expression after thetering GAL4 on day 11 for KRAB experiment |
| KRAB_GAL4_exp_2 | expression after thetering GAL4 on day 2 for KRAB experiment |
| KRAB_GAL4.POI_exp_11 | expression after thetering GAL4-KRAB on day 11 |
| KRAB_GAL4.POI_exp_2 | expression after thetering GAL4-KRAB on day 2 |
| KRAB_POI_exp_11 | expression after thetering KRAB on day 11 |
| KRAB_POI_exp_2 | expression after thetering KRAB on day 2 |
| CBX_GPvsP_day2 | fold change GAL4-CBX5 vs CBX5 on day 2 |
| CBX_GPvsP_day9 | fold change GAL4-CBX5 vs CBX5 on day 9 |
| G9a_GPvsP_day12 | fold change GAL4-G9a vs G9a on day 12 |
| G9a_GPvsP_day2 | fold change GAL4-G9a vs G9a on day 2 |
| KRAB_GPvsP_day11 | fold change GAL4-KRAB vs KRAB on day 11 |
| KRAB_GPvsP_day2 | fold change GAL4-KRAB vs KRAB on day 2 |
| CBX_GPvsG_day2 | fold change GAL4-CBX5 vs GAL4 on day 2 |
| CBX_GPvsG_day9 | fold change GAL4-CBX5 vs GAL4 on day 9 |
| G9a_GPvsG_day12 | fold change GAL4-G9a vs GAL4 on day 12 |
| G9a_GPvsG_day2 | fold change GAL4-G9a vs GAL4 on day 2 |
| KRAB_GPvsG_day11 | fold change GAL4-KRAB vs GAL4 on day 11 |
| KRAB_GPvsG_day2 | fold change GAL4-KRAB vs GAL4 on day 2 |
| CBX_PvsG_day2 | fold change CBX5 vs GAL4 on day 2 |
| CBX_PvsG_day9 | fold change CBX5 vs GAL4 on day 9 |
| G9a_PvsG_day12 | fold change G9a vs GAL4 on day 12 |
| G9a_PvsG_day2 | fold change G9a vs GAL4 on day 2 |
| KRAB_PvsG_day11 | fold change KRAB vs GAL4 on day 11 |
| KRAB_PvsG_day2 | fold change KRAB vs GAL4 on day 2 |
| G9a_D2_above_norm_cut | all normalization read-counts for G9a > 50 on day 2 |
| G9a_D12_above_norm_cut | all normalization read-counts for G9a > 50 on day 12 |
| CBX_D2_above_norm_cut | all normalization read-counts for CBX5 > 50 on day 2 |
| CBX_D9_above_norm_cut | all normalization read-counts for CBX5 > 50 on day 9 |
| KRAB_D2_above_norm_cut | all normalization read-counts for G9a > 50 on day 2 |
| KRAB_D11_above_norm_cut | all normalization read-counts for G9a > 50 on day 11 |
| chr_f | chromosome of forward mapping read |
| ori_f | orientation of forward mapping read |
| pos_f | location of forward read on the chromosome |
| t_reads_f | total number of forward reads |
| mapq_f | average mapping quality of forward reads mapped to the same location |
| freq1_f | frequency of forward reads mapped to the primary location |
| freq2_f | frequency of forward reads mapped to a possible secondary location |
| chr_r | chromosome of reverse mapping read |
| ori_r | orientation of reverse mapping read |
| pos_r | location of reverse read on the chromosome |
| t_reads_r | total number of reverse reads |
| mapq_r | average mapping quality of reverse reads mapped to the same location |
| freq1_r | frequency of reverse reads mapped to the primary location |
| freq2_r | frequency of reverse reads mapped to a possible secondary location |
| unique_map | according to the thresholds set in the data preperation, is the barcode uniquely mapped |
| rep_family | with which repeat family (repeatmasker nomenclature) does the barcode overlap |
| rep_name | with which repeat name (repeatmasker nomenclature) does the barcode overlap |
| lad | with which state of lamina association does the barcode overlap (constitutive LAD, facultative LAD, faculatative interLAD or constitutive interLAD) |
| chrom | with which of the 12 state chromatin model does the barcode overlap |
| unique_rep_name | does the barcode map to a single repeat name according to the thresholds set |
| unique_rep_family | does the barcode map to a single repeat family according to the thresholds set |
| unique_lad | does the barcode map to a single LAD-state according to the thresholds set |
| unique_chrom | does the barcode map to a single chromatin state according to the thresholds set |



## Path, Libraries, Parameters and Useful Functions

```{r functions}
opts_chunk$set(out.width='750px', dpi=200)
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),3,8) 

# libraries:
library(stringr)
library(ggplot2)
library(reshape2)
library(knitr)
library(gridExtra)
library(plyr)
library(grid)


load('../results/TTRIP_K562_FC_exp.rData')

# make cpg distance numeric
fc_table$cpg_distance = as.numeric(fc_table$cpg_distance)

# just so I can sellect the right day for each protein of interest in the loops
days = list(KRAB=c(11,14), G9a=12 ,CBX=c(9,12))
conditions = c('GAL4.POI', 'POI', 'GAL4')



filter_data <- function(fc_table, condition_vec, poi, day, type_vec, state_vec=NA, unique_map=F){
	# Filter the data for a specific experiment (e.g. GAL4-KRAB vs GAL4 on day 2)
	# Check the data preperation script for a more in depth look on the different
	# selection criteria.
	#
	# Args:
	#   fc_table: the big data-frame with all data.
	#   conditionx: the column name with the classification/expression/fold change that should be on the x-axis
	#   conditiony: the column name with the classification/expression/fold change that should be on the y-axis
	#   poi: the protein of interest in this experiment
	#   day: the day of the experiment
	#   xtype: what is the type of data for the x column?
	#          Used to transform the data correctly [value, factor or log2]
	#   ytype: what is the type of data for the y column?
	#          Used to transform the data correctly [value, factor or log2]
	#   state: optional value to add a state definition as a third column.
	# Returns:
	#   A filtered data frame with an x and y column, the x column contains a factor with
	#   the classification, the y column contains log2 transformed values of the experiment.
	total = nrow(fc_table)
	# check if the normalization counts are above 50 for each sample
	above_norm = sprintf('%s_D%i_above_norm_cut', poi, day)
	if (length(above_norm)>1){
		fc_table = fc_table[apply(fc_table[,above_norm],1, all),]
	} else{
		fc_table = fc_table[fc_table[,above_norm], ]
	}
	col_names = condition_vec
	summary_table = data.frame(cutoff=rep(NA, 5),
							   removed=rep(0, 5),
							   left=rep(0, 5),
							   stringsAsFactors=F)
	summary_table[1,] = list('above norm-count of 50', 
		                     total - nrow(fc_table),
		                     nrow(fc_table))
	# is the barcode uniquely linked to a single classification for clasifications used
	
	cd_matrix = cbind(type_vec, condition_vec)
	if (any(!is.na(state_vec))){
		cd_matrix = rbind(cd_matrix, cbind('factor', state_vec))
		col_names = c(col_names, state_vec)
	}
	if (unique_map){
		cd_matrix = rbind(cd_matrix, c('map', 'map'))
		col_names = c(col_names, 'map')
	}
	for (i in 1:nrow(cd_matrix)){
		type = cd_matrix[i,1]
		if (type == 'factor'){
			state = cd_matrix[i,2]
			unique_name = paste0('unique_',state)
			fc_table = fc_table[fc_table[,unique_name] | fc_table[,state]=='-',]
		}
		else if(type == 'map'){
			fc_table = fc_table[fc_table[,'unique_map'],]
		}
	}
	summary_table[2, ] = list('unique barcode-state link',
				              summary_table[1,'left'] - nrow(fc_table),
				              nrow(fc_table))

	# check if the log2 expression for the barcode of the GAL4 control > 0
	base_exp = sprintf('%s_GAL4_exp_%s',poi, day)
	above_exp_cut = fc_table[,base_exp]>0
	if (length(base_exp)>1){
		fc_table = fc_table[apply(above_exp_cut,1, all),]
	} else{
		fc_table = fc_table[above_exp_cut, ]
	}
	summary_table[3, ] = list('log2 base expression > 0', 
		                      summary_table[2,'left'] - nrow(fc_table),
	                          nrow(fc_table))
	

	# create a factor column x with the classification
	# the order can be important for the order in which the data is later represented in a plot.
	xy_list = list()
	for (i in 1:nrow(cd_matrix)){
		type = cd_matrix[i,1]
		condition = cd_matrix[i,2]
		if (type == 'factor'){
			if (condition == 'lad'){
				new_values = factor(fc_table[,condition], levels=c('cLAD','fLAD', 'fiLAD', 'ciLAD', '-'))
			} else if(condition=='chrom'){
				chrom_levels = unique(fc_table[,condition])
				chrom_levels = chrom_levels[!is.na(chrom_levels)]
				# sort on the number in the state name
				chrom_levels = chrom_levels[order(sapply(chrom_levels,function(x){
					# if the state is unknown, return a high number so that it ends up at the end of the sort
					if (x!='-'){
						return(as.numeric(str_split(x,'_')[[1]][1]))
					} else{
						return(Inf)
					}}))]
				new_values = factor(fc_table[,condition], levels=chrom_levels)

			} else {
				new_values = factor(fc_table[,condition])
			}
			
		} else if (type=='log2'){
			new_values = log2(fc_table[,condition] + 0.01)
		} else if (type != 'map'){
	    	new_values = fc_table[,condition]
		}
		if (type != 'map'){
			xy_list[[i]] = new_values
		}
	}
	xy_table=data.frame(xy_list)

	colnames(xy_table) = col_names
	rownames(xy_table) = rownames(fc_table)
	xy_table = data.frame(xy_table)
	for (condition in condition_vec){
		xy_table = xy_table[!is.na(xy_table[,condition]), ]
	}
	summary_table[4, ] = list('not NA', 
		                      summary_table[3,'left'] - nrow(xy_table),
	                          nrow(xy_table))

	for (condition in condition_vec){
		xy_table = xy_table[!is.infinite(xy_table[,condition]), ]
	}
	summary_table[5, ] = list('not infinite', 
		                      summary_table[4,'left'] - nrow(xy_table),
	                          nrow(xy_table))
	return(list(xy_table, summary_table))
}

```


## density pattern of expression at memory days
In previous scatterplots of there appeared to be a specific subset of barcodes that had memory in repressive chromatin states, late replicating DNA and LADs. We wondered if this could also be seen in density plots. If there would be a particular subset having "memmory"  this should result in a bimodal distribution, whereas, when there is a general effect of "memmory", the unimodal distribution would be shifted as a whole (using "", since real memmory is not proven).

I put vertical lines at the 2,5% and 92,5% of the GAL4 controll.


```{r, fig.width=10, fig.height=60}
for (poi in names(days)){
	for (day in days[[poi]]){
		condition_vec = sprintf('%s_%s_exp_%i', poi, conditions, day)
		xy_table = filter_data(fc_table, condition_vec, poi, day, rep('log2',3), state_vec=c('lad','chrom', 'segment'), unique_map=F)
		xy_melt = melt(xy_table[[1]], measure.vars=condition_vec)
		cdat = ddply(xy_melt, "variable", summarise, confint=quantile(value, c(0.025, 0.975)))
		cdat_lad = ddply(xy_melt, c("variable",'lad'), summarise, confint=quantile(value, c(0.025, 0.975)))
		cdat_segment = ddply(xy_melt, c("variable",'segment'), summarise, confint=quantile(value, c(0.025, 0.975)))
		cdat_chrom = ddply(xy_melt, c("variable",'chrom'), summarise, confint=quantile(value, c(0.025, 0.975)))
		p1=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat[cdat$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable)))
		p2=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat_lad[cdat_lad$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable))) + facet_wrap(~ lad, ncol=2)
		p3=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat_segment[cdat_segment$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable))) + facet_wrap(~ segment, ncol=2)
		p4=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat_chrom[cdat_chrom$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable))) + facet_wrap(~ chrom, ncol=3)
		grid.arrange(p1, p2, p3, p4, heights=c(1,3,2,5))
	}
}
```
**conclusion:**
There are often already small bumbs in the density with a group of low expressing barcodes in the controll groups. But this bumb seems to get larger with the GAL4-POI construct. In the KRAB day 11 experiment, the main peak is shifted and barcode expression in general is lower, this does not a big effect and it can be either biological or technical, but this makes it a bit harder to detect a bimodal distribution.

## density pattern of fold change at memory days
There was already some slight bimodality in previous plots, so to distinguish between barcodes thate were previously low expressing and barcodes which became lowly expressing, 

I put vertical lines at the 2,5% and 92,5% of the POI vs GAL4 controll.

```{r, fig.width=10, fig.height=60}
vs = c('GPvsG', 'GPvsP', 'PvsG')
for (poi in names(days)){
	for (day in days[[poi]]){
		condition_vec = sprintf('%s_%s_day%i', poi, vs, day)
		xy_table = filter_data(fc_table, condition_vec, poi, day, rep('log2',3), state_vec=c('lad','chrom', 'segment'), unique_map=F)
		xy_melt = melt(cbind(xy_table[[1]], barcode=rownames(xy_table[[1]])), measure.vars=condition_vec)
		cdat = ddply(xy_melt, "variable", summarise, confint=quantile(value, c(0.025, 0.975)))
		cdat_lad = ddply(xy_melt, c("variable",'lad'), summarise, confint=quantile(value, c(0.025, 0.975)))
		cdat_segment = ddply(xy_melt, c("variable",'segment'), summarise, confint=quantile(value, c(0.025, 0.975)))
		cdat_chrom = ddply(xy_melt, c("variable",'chrom'), summarise, confint=quantile(value, c(0.025, 0.975)))
		p1=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat[cdat$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable)))
		p2=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat_lad[cdat_lad$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable))) + facet_wrap(~ lad, ncol=2)
		p3=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat_segment[cdat_segment$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable))) + facet_wrap(~ segment, ncol=2)
		p4=ggplot(xy_melt , aes(x=value, colour=variable)) + geom_density() + geom_vline(data = cdat_chrom[cdat_chrom$variable==condition_vec[3],], aes(xintercept=as.numeric(confint), color=factor(variable))) + facet_wrap(~ chrom, ncol=3)
		grid.arrange(p1, p2, p3, p4, heights=c(1,3,2,5))
	}
}
```
**conclusion:**
At this resolution it is dificult to have a clear distinction between the two peaks of a potential bimodal distribution, although it does seem like some plots are hinting more towards a bimodal distribution than towards a general shift in expression as seen on day 2.
Gal4-CBX vs GAL4 at day 12 shows a really nice bimodal distribution, unfortunately this is different from the GAL4-CBX vs CBX fold changes, although there still seems to be some bimodality there as well. This can also be seen in the hight of the central peak.


## violin plots

We can also show violin plots with the same data, this might look nicer.

```{r, fig.width=10, fig.height=15}
for (poi in names(days)){
	for (day in days[[poi]]){
		condition_vec = sprintf('%s_%s_day%i', poi, vs, day)
		xy_table = filter_data(fc_table, condition_vec, poi, day, rep('log2',3), state_vec=c('lad','chrom', 'segment'), unique_map=F)

		p1 = ggplot(xy_table[[1]], aes(x=factor(segment, levels=c('ERD','TZ','LRD','-')))) + labs(x='replication segment', colour='replication segment')
		for (condition in factor(condition_vec)){
			p1 = p1 + geom_violin(data=cbind(xy_table[[1]], colour=condition), aes_string(y=condition, colour='colour'), alpha=0.3)
		}
		p2 = ggplot(xy_table[[1]], aes(x=factor(lad, levels=c('cLAD','fLAD','fiLAD','ciLAD', '-')))) + labs(x='lad state', colour='lad state')
		for (condition in factor(condition_vec)){
			p2 = p2 + geom_violin(data=cbind(xy_table[[1]], colour=condition), aes_string(y=condition, colour='colour'), alpha=0.3)
		}
		p3 = ggplot(xy_table[[1]], aes(x=factor(chrom))) + labs(x='chrom state', colour='chrom state') +
            theme(axis.text.x = element_text(hjust = 1, angle = 90, size = 10))
		for (condition in factor(condition_vec)){
			p3 = p3 + geom_violin(data=cbind(xy_table[[1]], colour=condition), aes_string(y=condition, colour='colour'), alpha=0.3)
		}
		grid.arrange(p1, p2, p3)

	}
}

```
## violin plots with data points

Unfortulately with the previous violin plots we've lost the information on the sizes of each grouping. We could also overlay the data points, and plot the different conditions seperately.

```{r, fig.width=10, fig.height=15}
for (poi in names(days)){
	for (day in days[[poi]]){
		condition_vec = sprintf('%s_%s_day%i', poi, vs, day)
		xy_table = filter_data(fc_table, condition_vec, poi, day, rep('log2',3), state_vec=c('lad','chrom', 'segment'), unique_map=F)
		for (condition in condition_vec){
			p1 = ggplot(xy_table[[1]], aes(x=segment, colour=segment)) + labs(x='replication segment', colour='replication segment') + geom_point(aes_string(y=condition), position=position_jitter(width=0.9), size=0.6, alpha=0.4) + geom_violin(aes_string(y=condition), alpha=0.4)
			p2 = ggplot(xy_table[[1]], aes(x=lad, colour=lad)) + labs(x='lad state', colour='lad state') + geom_point(aes_string(y=condition), position=position_jitter(width=0.9), size=0.6, alpha=0.4) + geom_violin(aes_string(y=condition), alpha=0.4)
			p3 = ggplot(xy_table[[1]], aes(x=chrom, colour=chrom)) + labs(x='chromatin state', colour='chromatin state') +geom_point(aes_string(y=condition), position=position_jitter(width=0.9), size=0.6, alpha=0.4) + theme(axis.text.x = element_text(hjust = 1, angle = 90, size = 10))+ geom_violin(aes_string(y=condition), alpha=0.4)
			grid.arrange(p1, p2, p3, top=textGrob(condition, gp=gpar(fontsize=38)))
		}
	}
}

```


```{r}
sessionInfo()
getwd()
date()
paste("Run time: ",format(Sys.time()-StartTime))
```
