# r document van Steensel lab

# Gene repression in LADs
## Christ Leemans, 03-11-2016 - 22-09-2017

## Introduction
Generally speaking, genes inside lamina associated domains are not or very lowly expressed. These genes can either be actively repressed by their DNA context (e.g. heterochromatin, lamina association), or simply be inactive (because essential factors for expression are missing?). Yet another group of genes seem to evade gene repression in the context of lamina associated domains.




```{r}

library(reshape2)
library(rtracklayer)
library(ggplot2)
library(gridExtra)
library(plyr)
library(scales)
library(grid)
library(gtable)
library(affy)
library(limma)
library(biomaRt)
## FROM STACKOVERFLOW:
## https://stackoverflow.com/questions/12539348/ggplot-separate-legend-and-plot

pseudo_log10 <- function(val_vec){
    Pseud=min(val_vec[val_vec > 0], na.rm=TRUE)/2
    val_vec = val_vec + Pseud
    return(log10(val_vec))
}

## get a table with matching sets
## table = complete table to take matching sets from
## class_col = column name of class of interest
## class = name of class to match the set on
## order_on = column name to order on
matchSet <- function(table, class_col, class, order_on){
  o_vec = order(table[,order_on])
  o_table = table[o_vec, ]
  setA = which(o_table[,class_col]==class)
  setB = c(setA + 1, setA -1)
  ## check if setB is all within the possible indexes
  setB = setB[setB %in% 1:length(o_vec)]
  ## can also return o_table[unique(c(setA, setB)), ]
  ## but this way order is perserved.
  i_vec = o_vec[unique(c(setA, setB))]
  return(table[i_vec[order(i_vec)], ])
}


COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")



id_table = read.table('../raw_data/transcript.table', stringsAsFactors=F,
                      row.names=1, col.names=c('transcript_id', 'gene_id',
                                               'symbol'))
load("~joris/mydata/git/SuRE/Joris//analysis_postNBT/Gencode_DF_generation_170707/gencode.sure.170712.rda")


P<-gencode.sure.170712[,c('chr', 'strand', 'txStart', 'txEnd', 'name', 'name2',
                          'tss', 'distance.to.previous.tss',
                          'k562.combined.45.55.sense', 'HEPG2.sense',
                          'HT1080.sense', 'gro.cap.1kb.sense',
                          'encode.cage.reprocessed.1kb.sense',
                          'nr.of.tissues.in.which.expressed.max')]

names(P)[9:14]<-c("SuRE_K562", "SuRE_HepG2", "SuRE_HT1080", "GROcap_K562",
                  "CAGE_K562_encode", 'tissues_expressed')

rownames(P) = P$name

for (col in c('SuRE_K562', 'SuRE_HepG2', 'SuRE_HT1080', 'GROcap_K562',
              'CAGE_K562_encode')){
    P[,col] = pseudo_log10(P[,col])
}

P$gene_id = id_table[P$name,'gene_id']

gene_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                            start=P$txStart,
                                            end=P$txEnd,
                                            strand=P$strand,
                                            name=P$name,
                                            tss=P$tss),
                                            keep.extra.columns=TRUE)
names(gene_gr) = P$name
tss_gr = gene_gr
ranges(tss_gr) = IRanges(gene_gr$tss,
                       gene_gr$tss)
names(tss_gr) = P$name

## get LAD data for K562
LAD_K562 = import.bed('~c.leemans/mydata/data/tracks/hg19/cl20161019_LAD_continuous_2state_K562.bed')
## to keep with Joris's previous analysis, let's assign a state to every promoter
o = findOverlaps(tss_gr, LAD_K562[LAD_K562$name=='LAD'])
P$LAD_K562 = 0
P$LAD_K562[queryHits(o)] = 1


create_RM <-function(data, x, y, lad){
    #then calculate running mean for iLAD promoters:
    #sort by SuRE and then random for ties
    o = order(data[,x],sample(c(1:nrow(data))))

    x_sorted = data[o,x]
    y_sorted = data[o,y]
    lad_sorted = data[o,lad]

    n<-60 #number of windows
    w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
    s<-round(seq(from=w/2+0.0001, to=nrow(data)-w/2, length.out=n))
    RM<-data.frame(x.low=rep(NA,n), x.mean=rep(NA,n), x.hi=rep(NA,n), y.lad=rep(NA,n), y.ilad=rep(NA,n))
    RM$x.low=x_sorted[s-floor(w/2)]
    for(i in 1:n){RM$x.mean[i]=mean(x_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
    RM$x.hi=x_sorted[s+floor(w/2)]
    for(i in 1:n)
      {t<-data.frame(LAD=lad_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))],
                     y=y_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))])
       RM$y.lad[i]<-mean(t$y[t$LAD==1], na.rm=TRUE)
       RM$y.ilad[i]<-mean(t$y[t$LAD==0], na.rm=TRUE)
      }
    #add first datapoint (SuRE equals pseudocount)
    RM1<-RM[0,] #empty df
    RM1[1,]<-c(rep(min(x_sorted),3), mean(y_sorted[x_sorted==min(x_sorted) & lad_sorted==1]), mean(y_sorted[x_sorted==min(x_sorted) & lad_sorted==0]))
    RM<-rbind(RM1, RM)
    rm(RM1)
    return(RM)
}

RM_GRO = create_RM(P, 'SuRE_K562', 'GROcap_K562', lad='LAD_K562')
P$LRS_GROcap <- P$GROcap_K562 - approx(x=RM_GRO$x.mean, y=RM_GRO$y.ilad, xout=P$SuRE_K562, rule=2)$y

classify <- function(sure, exp, lrs, lad, exp_cut){
    INACT<- sure< -0.3 & lad & exp< exp_cut #inactive
    NREP<- sure> 0 & lrs > -0.5 & lad & exp> exp_cut #not repressed
    REP<- sure> 0.3 & lrs < -1 & lad  & exp< exp_cut #repressed
    Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
    names(Pcnts)<-c("repressed", "escaper", "inactive")
    BND <- lad & !INACT & !NREP & !REP
    class = rep(NA, length(sure))
    class[lad==0] = 'iLAD'
    class[INACT]<-"inactive"
    class[NREP]<-"escaper"
    class[REP]<-"repressed"
    class[BND] <- "boundary"
    return(factor(class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary')))
}

P$class = classify(P$SuRE_K562, P$GROcap_K562, P$LRS_GROcap, P$LAD_K562, -2)

```



```{r}
trip_gr = import.gff3('../results/cl20170925_TRIP_promoters_evsr_hg19.gff3')
trip_data = as.data.frame(trip_gr)[c('ID', 'name', 'gene_id', 'gene_name',
                                     'gene_type', 'plasmid_id', 'class')]
trip_data$number=gsub('.*_', '', trip_data$name)
trip_data$transcript_id = gsub('_ext','', trip_data$ID)
trip_data$GROcap_K562=P[trip_data$transcript_id, 'GROcap_K562']
trip_data$SuRE_K562=P[trip_data$transcript_id, 'SuRE_K562']
p_classes = P[which(P$class %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P, aes(x=SuRE_K562, y=GROcap_K562, color=class)) +
	geom_point(size=0.1, alpha=0.1) +
	geom_point(data=p_classes, size=0.3) +
	geom_point(data=trip_data, size=1, color='black') +
	geom_text(data=trip_data, aes(label=number), color='black', nudge_x=0.1,
			  fontface = "bold") +
    scale_colour_manual(values=COL_class)


K562_CAGE_rep1 = read.table('../results/tss_CAGE_K562_rep1.bed')
K562_CAGE_rep2 = read.table('../results/tss_CAGE_K562_rep2.bed')

K562_CAGE = rowSums(cbind(K562_CAGE_rep1[,7], K562_CAGE_rep2[,7]))
P$CAGE_K562 = pseudo_log10(K562_CAGE[p_complete%in%rownames(P)])


rnaseq_rep1 = read.table('../raw_data/K562_rna_rep1_ENCFF004LGY.tsv',
                         header=T, row.names=1, stringsAsFactors=F)
rnaseq_rep2 = read.table('../raw_data/K562_rna_rep2_ENCFF222NCB.tsv',
                         header=T, row.names=1, stringsAsFactors=F)

transcript_table = read.table('../raw_data/K562_rna_rep1_transcripts_ENCFF853QUL.tsv',
                               header=T, row.names=1, stringsAsFactors=F)

p_match = transcript_table[P$name, 'gene_id']

fpkm_rep1 = rnaseq_rep1[p_match, 'pme_FPKM']
fpkm_rep2 = rnaseq_rep2[p_match, 'pme_FPKM']

fpkm = rowMeans(cbind(fpkm_rep1, fpkm_rep2))
P$K562_fpkm = pseudo_log10(fpkm)


p_matched = matchSet(P[P$class%in%c('escaper', 'iLAD'), ],
                     'class', 'escaper', 'GROcap_K562')
p_subset = rbind(p_matched, P[P$class=='repressed', ])

trip_data$CAGE_K562_encode = P[trip_data$transcript_id, 'CAGE_K562_encode']
trip_data$CAGE_K562 = P[trip_data$transcript_id, 'CAGE_K562']
trip_data$K562_fpkm = P[trip_data$transcript_id, 'K562_fpkm']

ggplot(p_subset, aes(x=class, y=CAGE_K562_encode, color=class)) +
    geom_violin(alpha=0.5) +
    geom_point(size=0.5, alpha=0.5, position=position_jitter(width=0.5)) +
	geom_text(data=trip_data, aes(label=number), color='black',
			  position=position_jitter(width=0.5), fontface = "bold") +
    scale_color_manual(values = COL_class)



ggplot(p_subset, aes(x=class, y=CAGE_K562, color=class)) +
    geom_violin(alpha=0.5) +
    geom_point(size=0.5, alpha=0.5, position=position_jitter(width=0.5)) +
	geom_text(data=trip_data, aes(label=number), color='black',
			  position=position_jitter(width=0.5), fontface = "bold") +
    scale_color_manual(values = COL_class)


ggplot(p_subset, aes(x=class, y=K562_fpkm, color=class)) +
    geom_violin(alpha=0.5) +
    geom_point(size=0.5, alpha=0.5, position=position_jitter(width=0.5)) +
	geom_text(data=trip_data, aes(label=number), color='black',
			  position=position_jitter(width=0.5), fontface = "bold") +
    scale_color_manual(values = COL_class)

```
**conclusion:**
Escaper 7 is clearly on the lower end of the spectrum if we examine other
measures of expression. Escaper 3 which shows hardly any expression in the
TRIP experiment at this point, however seems completely fine.


```{r}
pol2_tssr = read.table('../raw_data/tssr_POL2A_K562.txt.gz', row.names=4,
                       col.names=c('chrom', 'start', 'end', 'name', 'score',
                                   'strand', 'size', 'num_data', 'min', 'max',
                                   'mean', 'sum'), stringsAsFactors=F)

pol2_body = read.table('../raw_data/body_POL2A_K562.txt.gz', row.names=4, skip=1,
                       col.names=c('chrom', 'start', 'end', 'name', 'score',
                                  'strand', 'size', 'num_data', 'min', 'max',
                                  'mean', 'sum'), stringsAsFactors=F)
pol2_tssr$mean = pol2_tssr$sum / pol2_tssr$num_data
pol2_body$mean = pol2_body$sum / pol2_body$num_data

P$tssr_POL2A_K562 = pseudo_log10(pol2_tssr[P$name, 'mean'])
P$body_POL2A_K562 = pseudo_log10(pol2_body[P$name, 'mean'])

trip_data$tssr_POL2A_K562 = P[trip_data$transcript_id, 'tssr_POL2A_K562']
trip_data$body_POL2A_K562 = P[trip_data$transcript_id, 'body_POL2A_K562']


p_matched = matchSet(P[P$class%in%c('escaper', 'iLAD'), ],
                     'class', 'escaper', 'GROcap_K562')
p_subset = rbind(p_matched, P[P$class=='repressed', ])

ggplot(p_subset, aes(x=class, y=tssr_POL2A_K562, color=class)) +
	geom_violin(alpha=0.5) +
	geom_point(size=0.5, alpha=0.5, position=position_jitter(width=0.5)) +
	geom_text(data=trip_data, aes(label=number), color='black',
			  position=position_jitter(width=0.5), fontface = "bold") +
	scale_color_manual(values = COL_class)

ggplot(p_subset, aes(x=class, y=body_POL2A_K562, color=class)) +
	geom_violin(alpha=0.5) +
	geom_point(size=0.5, alpha=0.5, position=position_jitter(width=0.5)) +
	geom_text(data=trip_data, aes(label=number), color='black',
			  position=position_jitter(width=0.5), fontface = "bold") +
	scale_color_manual(values = COL_class)

```

**conclusion:**

POL2 occupancy on gene body is really low for some escapers. On TSS however,
escapers are all at the higher end of the spectrum.




```{r}

wilcox_affinity <- function(x_affinity, y_affinity, groups, tf_table, id_vec=NULL){
  if (is.null(id_vec)){
    id_vec = colnames(x_affinity)
  }  
  fit = mclapply(id_vec, function(id){
    x = x_affinity[,id]
    y = y_affinity[,id]
    r = rank(c(x,y))
    r_x = r[1:length(x)]
    r_y = r[-(1:length(x))]
    mean_r = c(mean(r_x), mean(r_y))
    direction = groups[which(mean_r==max(mean_r))]
    if (length(direction) == 2){
      direction = 'unchanged'
    }
    median_fc = median(x) / median(y)
    mean_fc = mean(x) / mean(y)
    rank_sum = sum(r_x)
    w = wilcox.test(x, y)
    return(list(w,direction, median_fc, mean_fc, rank_fc))
  })
  p_vec = unlist(lapply(fit, function(x){ x[[1]]$p.value}))
  p_adjust = p.adjust(p_vec, method='fdr')
  direction = lapply(fit, function(x){ x[[2]]})
  median_fc =  lapply(fit, function(x){ x[[3]]})
  mean_fc =  lapply(fit, function(x){ x[[4]]})
  rank_fc =  lapply(fit, function(x){ x[[5]]})
  result_table = cbind(id=id_vec,
                       tf_table[id_vec, ],
                       direction=unlist(direction),
                       p_adjust = p_adjust,
                       median_fc = unlist(median_fc),
                       mean_fc = unlist(mean_fc),
                       rank_fc = unlist(rank_fc), stringsAsFactors=F)
  return(result_table)
}

apply_wilcox <- function(affinity, prom_table, group_vec, class_name,
                         tf_table, id_vec=NULL){
    if (is.null(id_vec)){
        id_vec = colnames(affinity)
    }
    id_vec = id_vec[id_vec%in%colnames(affinity)]
    fit = mclapply(id_vec, function(id){
        aff_vec = affinity[rownames(prom_table),id]
        group1 = which(prom_table[, class_name]==group_vec[1])
        rank_aff = rank(aff_vec)
        x = aff_vec[group1]
        y = aff_vec[-group1]
        rank_mean = c(mean(rank_aff[group1]), mean(rank_aff[-group1]))
        names(rank_mean) = paste0('mean_rank_', group_vec)
        rank_fc = mean(rank_aff[group1]) /mean(rank_aff[-group1])
        direction = ifelse(rank_fc > 1, group_vec[1], group_vec[2])
        median_fc = median(x) / median(y)
        mean_fc = mean(x) / mean(y)
        w = wilcox.test(x, y)
        return(list(w,direction, median_fc, mean_fc, rank_fc, rank_mean))
  })
  p_vec = unlist(lapply(fit, function(x){ x[[1]]$p.value}))
  p_adjust = p.adjust(p_vec, method='fdr')
  direction = lapply(fit, function(x){ x[[2]]})
  median_fc =  lapply(fit, function(x){ x[[3]]})
  mean_fc =  lapply(fit, function(x){ x[[4]]})
  rank_fc =  lapply(fit, function(x){ x[[5]]})
  rank_mean = do.call(rbind, lapply(fit, function(x){ x[[6]]}))
  result_table = data.frame(id=id_vec,
                            tf_table[id_vec, ],
                            direction=unlist(direction),
                            p_adjust = p_adjust,
                            mean_fc = unlist(mean_fc),
                            median_fc = unlist(median_fc),
                            rank_fc = unlist(rank_fc),
                            rank_mean,
                            stringsAsFactors=F)
  return(result_table)
}



aff_table_jaspar = read.table('../raw_data/jaspar_affinity_pseudo/seq_psam.dat',
                              stringsAsFactors=F)
tf_table = read.table('../raw_data/tf_table.txt', sep='\t', row.names=1,
                      stringsAsFactors=F)
colnames(tf_table) = c('name', 'species', 'class', 'family')

tf_translation = ddply(tf_table[,c('species','name')], .(name),
                       function(x){
                         symbol = gsub('[(]var.[0-9][)]','', x[,2])
                         cbind(x,symbol=unlist(strsplit(symbol, '::')))
                       })
tf_translation$symbol = as.character(tf_translation$symbol)

find_symbol <- function(symbol, id_table){
    result = id_table[which(id_table$symbol==toupper(symbol)),'gene_id']
    if (length(result) == 0){
        result = NA
    } else {
        table = table(result)
        result = names(which.max(table))
    }
    return(result)
}

id_table = read.table('../raw_data/transcript.table', stringsAsFactors=F,
                    row.names=1, col.names=c('transcript_id', 'gene_id',
                                             'symbol'))
id_table = unique(id_table)
rownames(id_table) = id_table$gene_id

tf_translation$gene_id = unlist(lapply(tf_translation$symbol, find_symbol,
                                       id_table))


fpkm_rep1 = rnaseq_rep1[tf_translation$gene_id, 'pme_FPKM']
fpkm_rep2 = rnaseq_rep2[tf_translation$gene_id, 'pme_FPKM']



tf_translation$K562_fpkm = pseudo_log10(rowMeans(cbind(fpkm_rep1, fpkm_rep2)))



KBM7_essential = read.table('../raw_data/KBM7_essentialome_aac7557_SM_Table_S1.csv',
                            stringsAsFactors=F, skip=1, header=T, sep='\t',
                            row.names=2)
HAP1_essential = read.table('../raw_data/HAP1_essentialome_aac7557_SM_Table_S2.csv',
                            stringsAsFactors=F, skip=1, header=T, sep='\t',
                            row.names=2)
ens_vec = gsub('[.].*','',tf_translation$gene_id)
tf_translation$KBM7_essential = KBM7_essential[ens_vec, 'selected']=='YES'
tf_translation$HAP1_essential = HAP1_essential[ens_vec, 'selected']=='YES'

tf_expression = ddply(tf_translation,.(name),
                      function(x){
                          c(K562_fpkm=min(x$K562_fpkm),
                            KBM7_essential=any(x$KBM7_essential),
                            HAP1_essential=any(x$HAP1_essential))
                      })


tf_table$K562_fpkm = NaN
tf_match = match(tf_expression$name, tf_table$name)
tf_table$K562_fpkm[tf_match] = tf_expression$K562_fpkm
tf_table$KBM7_essential = NA
tf_table$KBM7_essential[tf_match] = tf_expression$KBM7_essential
tf_table$HAP1_essential = NA
tf_table$HAP1_essential[tf_match] = tf_expression$HAP1_essential


tf_type = read.table('../raw_data/12859_2016_1349_MOESM2_ESM.csv',
                     sep='\t', header=T, stringsAsFactors=F)
match_vec = match(toupper(tf_table$name), tf_type$Symbol)
tf_table$tf_type = NA
not_na = !is.na(match_vec)
tf_table[not_na, 'tf_type'] = tf_type[match_vec[not_na],
                                      'Chromatin.Opening.Type']

colnames(aff_table_jaspar) = gsub('.xml','', colnames(aff_table_jaspar))
rownames(aff_table_jaspar) = gsub('::.*','', rownames(aff_table_jaspar))


load('../raw_data/cl20170814_jaspar_gencode_cage_cor.rda')
cor_data = t(cor_data)
colnames(cor_data) = rownames(tf_table)[match(colnames(cor_data), tf_table$name)]




id_vec = colnames(cor_data)

matched_evsr_K562 = matchSet(P[P$class%in%c('repressed', 'escaper'), ],
                             'class', 'escaper', 'SuRE_K562')

evsr_jaspar_K562 = apply_wilcox(aff_table_jaspar, matched_evsr_K562,
                                c('repressed', 'escaper'), 'class',
                                tf_table, id_vec)


evsr_jaspar_exp = apply_wilcox(cor_data, matched_evsr_K562,
                               c('escaper', 'repressed'), 'class',
                               tf_table)
evsr_jaspar_K562$p_fantom = evsr_jaspar_exp[evsr_jaspar_K562$id, 'p_adjust']
evsr_jaspar_K562$mean_rank_fantom_escaper = evsr_jaspar_exp[evsr_jaspar_K562$id,
                                                          'mean_rank_escaper']
evsr_jaspar_K562$mean_rank_fantom_repressed = evsr_jaspar_exp[evsr_jaspar_K562$id,
                                                            'mean_rank_repressed']


matched_enhancer = matchSet(P_enh[P_enh$class%in%c('repressed', 'escaper'), ],
                            'class', 'escaper', 'SuRE_K562')



## let's leave out enhancers from this analysis from now.
evsr_jaspar_K562$distance_evsr = sqrt(rank(evsr_jaspar_K562$mean_rank_escaper)^2 +
									  rank(evsr_jaspar_K562$mean_rank_fantom_escaper)^2)
```

```{r, fig.width=10, fig.height=60}
top = head(evsr_jaspar_K562[order(evsr_jaspar_K562$distance_evsr, decreasing=T), ])

plot_list = list()
for (id in rownames(top)){
	title = paste0(id, '; ', top[id, 'name'])
	data = data.frame(affinity = aff_table_jaspar[p_subset$name, id],
	                  class = p_subset$class)
	t = data.frame(number=trip_data$number,
		           affinity = aff_table_jaspar[trip_data$transcript_id, id],
				   class = trip_data$class)
	plot_list[[id]] = ggplot(data, aes(x=class, y=pseudo_log10(affinity),
	                                   color=class)) +
		ggtitle(title) +
		geom_violin(alpha=0.5) +
		geom_point(size=0.5, alpha=0.5, position=position_jitter(width=0.3)) +
		geom_text(data=t, aes(label=number), color='black',
				  position=position_jitter(width=0.4), fontface = "bold") +
		scale_color_manual(values = COL_class)
}
do.call(grid.arrange, c(plot_list, ncol=1))

```

**conclusions:**

If we would like to go for anecdotal evidence of transcription factors being
able to mediate escaper behavior, escaper 1 seems to be a good candidate.
