
# knitr document van Steensel lab

# Thethered TRIP
## Christ Leemans, 31-05-2016 - to date 

## Introduction
Laura performed a thetered-TRIP experiment on a K562 cell pool. She transiently transfected the cell pool with GAL4 thetered to 3 different "proteins of interest"(POI), namely G9a, the KRAB domain and CBX5. She took measurements after day 2 to see the initial effect of the thetering and after 12, 11 and 9 days respectively to see if there was any memory of the silencing on day 2. Besides the thetered GAL4-POI, two seperate controlls were taken using unthetered GAL4 and unthethered POI.

## Experimental setup
At this moment Laura has data for 3 different tethering experiments using KRAB, G9a and CBX5. For each protein of interest (POI) there are 12 expression and 12 gDNA files: 3 conditions * 2 different days after induction * 2 replicates. One condition uses an unthethered POI, the second uses only GAL4 and the third condition uses the POI thethered to GAL4 (GAL4-POI). Expression and gDNA data was obtained on day 2 and day 9. With each sequencing run, spikeins were added to normalize across different experiments. There is a different config file to extract the expression values of the spikeins.

## Description of data
For this part of the analysis I used the data that was generated previously in the data preperation report: cl20160713_TRIP_K562_data_preperation. 
The .RData produced contains a large data frame with a row for each barcode and the following columns:

| column name | description |
| ----------- | ----------- |
| CBX_GAL4_exp_2 | expression after thetering GAL4 on day 2  for CBX5 experiment |
| CBX_GAL4_exp_9 | expression after thetering GAL4 on day 9 for CBX5 experiment |
| CBX_GAL4.POI_exp_2 | expression after thetering GAL4-CBX5 on day 2 |
| CBX_GAL4.POI_exp_9 | expression after thetering GAL4-CBX5 on day 9 |
| CBX_POI_exp_2 | expression after thetering CBX5 on day 2 |
| CBX_POI_exp_9 | expression after thetering CBX5 on day 9 |
| G9a_GAL4_exp_12 | expression after thetering GAL4 on day 12 for G9a experiment |
| G9a_GAL4_exp_2 | expression after thetering GAL4 on day 2 for G9a experiment |
| G9a_GAL4.POI_exp_12 | expression after thetering GAL4-G9a on day 12 |
| G9a_GAL4.POI_exp_2 | expression after thetering GAL4-G9a on day 2 |
| G9a_POI_exp_12 | expression after thetering G9a on day 12 |
| G9a_POI_exp_2 | expression after thetering G9a on day 2 |
| KRAB_GAL4_exp_11 | expression after thetering GAL4 on day 11 for KRAB experiment |
| KRAB_GAL4_exp_2 | expression after thetering GAL4 on day 2 for KRAB experiment |
| KRAB_GAL4.POI_exp_11 | expression after thetering GAL4-KRAB on day 11 |
| KRAB_GAL4.POI_exp_2 | expression after thetering GAL4-KRAB on day 2 |
| KRAB_POI_exp_11 | expression after thetering KRAB on day 11 |
| KRAB_POI_exp_2 | expression after thetering KRAB on day 2 |
| CBX_GPvsP_day2 | fold change GAL4-CBX5 vs CBX5 on day 2 |
| CBX_GPvsP_day9 | fold change GAL4-CBX5 vs CBX5 on day 9 |
| G9a_GPvsP_day12 | fold change GAL4-G9a vs G9a on day 12 |
| G9a_GPvsP_day2 | fold change GAL4-G9a vs G9a on day 2 |
| KRAB_GPvsP_day11 | fold change GAL4-KRAB vs KRAB on day 11 |
| KRAB_GPvsP_day2 | fold change GAL4-KRAB vs KRAB on day 2 |
| CBX_GPvsG_day2 | fold change GAL4-CBX5 vs GAL4 on day 2 |
| CBX_GPvsG_day9 | fold change GAL4-CBX5 vs GAL4 on day 9 |
| G9a_GPvsG_day12 | fold change GAL4-G9a vs GAL4 on day 12 |
| G9a_GPvsG_day2 | fold change GAL4-G9a vs GAL4 on day 2 |
| KRAB_GPvsG_day11 | fold change GAL4-KRAB vs GAL4 on day 11 |
| KRAB_GPvsG_day2 | fold change GAL4-KRAB vs GAL4 on day 2 |
| CBX_PvsG_day2 | fold change CBX5 vs GAL4 on day 2 |
| CBX_PvsG_day9 | fold change CBX5 vs GAL4 on day 9 |
| G9a_PvsG_day12 | fold change G9a vs GAL4 on day 12 |
| G9a_PvsG_day2 | fold change G9a vs GAL4 on day 2 |
| KRAB_PvsG_day11 | fold change KRAB vs GAL4 on day 11 |
| KRAB_PvsG_day2 | fold change KRAB vs GAL4 on day 2 |
| G9a_D2_above_norm_cut | all normalization read-counts for G9a > 50 on day 2 |
| G9a_D12_above_norm_cut | all normalization read-counts for G9a > 50 on day 12 |
| CBX_D2_above_norm_cut | all normalization read-counts for CBX5 > 50 on day 2 |
| CBX_D9_above_norm_cut | all normalization read-counts for CBX5 > 50 on day 9 |
| KRAB_D2_above_norm_cut | all normalization read-counts for G9a > 50 on day 2 |
| KRAB_D11_above_norm_cut | all normalization read-counts for G9a > 50 on day 11 |
| chr_f | chromosome of forward mapping read |
| ori_f | orientation of forward mapping read |
| pos_f | location of forward read on the chromosome |
| t_reads_f | total number of forward reads |
| mapq_f | average mapping quality of forward reads mapped to the same location |
| freq1_f | frequency of forward reads mapped to the primary location |
| freq2_f | frequency of forward reads mapped to a possible secondary location |
| chr_r | chromosome of reverse mapping read |
| ori_r | orientation of reverse mapping read |
| pos_r | location of reverse read on the chromosome |
| t_reads_r | total number of reverse reads |
| mapq_r | average mapping quality of reverse reads mapped to the same location |
| freq1_r | frequency of reverse reads mapped to the primary location |
| freq2_r | frequency of reverse reads mapped to a possible secondary location |
| unique_map | according to the thresholds set in the data preperation, is the barcode uniquely mapped |
| rep_family | with which repeat family (repeatmasker nomenclature) does the barcode overlap |
| rep_name | with which repeat name (repeatmasker nomenclature) does the barcode overlap |
| lad | with which state of lamina association does the barcode overlap (constitutive LAD, facultative LAD, faculatative interLAD or constitutive interLAD) |
| chrom | with which of the 12 state chromatin model does the barcode overlap |
| unique_rep_name | does the barcode map to a single repeat name according to the thresholds set |
| unique_rep_family | does the barcode map to a single repeat family according to the thresholds set |
| unique_lad | does the barcode map to a single LAD-state according to the thresholds set |
| unique_chrom | does the barcode map to a single chromatin state according to the thresholds set |



## Path, Libraries, Parameters and Useful Functions

```{r functions}
opts_chunk$set(out.width='750px', dpi=200)
StartTime <-Sys.time()

# 6-digit Date tag:
Date <- substr(gsub("-","",Sys.time()),3,8) 

# libraries:
library(stringr)
library(ggplot2)
library(reshape2)
library(knitr)
library(gridExtra)
library(plyr)
library(grid)


load('../results/TTRIP_K562_FC_exp.rData')

# make cpg distance numeric
fc_table$cpg_distance = as.numeric(fc_table$cpg_distance)

# just so I can sellect the right day for each protein of interest in the loops
days = list(KRAB=c(11,14), G9a=12 ,CBX=c(9,12))


filter_data <- function(fc_table, conditionx, conditiony, poi, day, xtype, ytype, state=NA, unique_map=F){
	# Filter the data for a specific experiment (e.g. GAL4-KRAB vs GAL4 on day 2)
	# Check the data preperation script for a more in depth look on the different
	# selection criteria.
	#
	# Args:
	#   fc_table: the big data-frame with all data.
	#   conditionx: the column name with the classification/expression/fold change that should be on the x-axis
	#   conditiony: the column name with the classification/expression/fold change that should be on the y-axis
	#   poi: the protein of interest in this experiment
	#   day: the day of the experiment
	#   xtype: what is the type of data for the x column?
	#          Used to transform the data correctly [value, factor or log2]
	#   ytype: what is the type of data for the y column?
	#          Used to transform the data correctly [value, factor or log2]
	#   state: optional value to add a state definition as a third column.
	# Returns:
	#   A filtered data frame with an x and y column, the x column contains a factor with
	#   the classification, the y column contains log2 transformed values of the experiment.
	if (!is.na(state)){
		fc_table = fc_table[which(fc_table[,state]!='-'),]
	}
	if (xtype == 'factor'){
		fc_table = fc_table[which(fc_table[,conditionx]!='-'),]
	}
	if (ytype == 'factor'){
		fc_table = fc_table[which(fc_table[,conditiony]!='-'),]
	}
	total = nrow(fc_table)
	# check if the normalization counts are above 50 for each sample
	above_norm = sprintf('%s_D%i_above_norm_cut', poi, day)
	if (length(above_norm)>1){
		fc_table = fc_table[apply(fc_table[,above_norm],1, all),]

	} else{
		fc_table = fc_table[fc_table[,above_norm], ]
	}
	
	summary_table = data.frame(cutoff=rep(NA, 5),
							   removed=rep(0, 5),
							   left=rep(0, 5),
							   stringsAsFactors=F)
	summary_table[1,] = list('above norm-count of 50', 
		                     total - nrow(fc_table),
		                     nrow(fc_table))
	# is the barcode uniquely linked to a single classification for clasifications used
	
	if (!is.na(state)){
		col_names = c('x','y','state')
		cd_matrix = cbind(c(xtype,ytype, 'factor'),c(conditionx,conditiony, state))
	} else {
		col_names = c('x','y')
		cd_matrix = cbind(c(xtype,ytype),c(conditionx,conditiony))
	}
	if (unique_map){
		cd_matrix = rbind(cd_matrix, c('map', 'map'))
	}
	for (i in 1:nrow(cd_matrix)){
		type = cd_matrix[i,1]
		if (type == 'factor'){
			state = cd_matrix[i,2]
			unique_name = paste0('unique_',state)
			fc_table = fc_table[fc_table[,unique_name] | fc_table[,state]=='-',]
		}
		else if(type == 'map'){
			fc_table = fc_table[fc_table[,'unique_map'],]
		}
	}
	summary_table[2, ] = list('unique barcode-state link',
				              summary_table[1,'left'] - nrow(fc_table),
				              nrow(fc_table))

	# check if the log2 expression for the barcode of the GAL4 control > 0
	base_exp = sprintf('%s_GAL4_exp_%s',poi, day)
	above_exp_cut = fc_table[,base_exp]>0
	if (length(base_exp)>1){
		fc_table = fc_table[apply(above_exp_cut,1, all),]
	} else{
		fc_table = fc_table[above_exp_cut, ]
	}
	summary_table[3, ] = list('log2 base expression > 0', 
		                      summary_table[2,'left'] - nrow(fc_table),
	                          nrow(fc_table))
	

	# create a factor column x with the classification
	# the order can be important for the order in which the data is later represented in a plot.
	xy_list = list()
	for (i in 1:nrow(cd_matrix)){
		type = cd_matrix[i,1]
		condition = cd_matrix[i,2]
		if (type == 'factor'){
			if (condition == 'lad'){
				new_values = factor(fc_table[,condition], levels=c('cLAD','fLAD', 'fiLAD', 'ciLAD', '-'))
			} else if(condition=='chrom'){
				chrom_levels = unique(fc_table[,condition])
				chrom_levels = chrom_levels[!is.na(chrom_levels)]
				# sort on the number in the state name
				chrom_levels = chrom_levels[order(sapply(chrom_levels,function(x){
					# if the state is unknown, return a high number so that it ends up at the end of the sort
					if (x!='-'){
						return(as.numeric(str_split(x,'_')[[1]][1]))
					} else{
						return(Inf)
					}}))]
				new_values = factor(fc_table[,condition], levels=chrom_levels)

			} else {
				new_values = factor(fc_table[,condition])
			}
			
		} else if (type=='log2'){

			new_values = log2(fc_table[,condition] + 0.1)
		} else if (type != 'map'){
	    	new_values = fc_table[,condition]
		}
		if (type != 'map'){
			xy_list[[i]] = new_values
		}
	}
	xy_table=data.frame(xy_list)
	colnames(xy_table) = col_names
	rownames(xy_table) = rownames(fc_table)
	xy_table = data.frame(xy_table)
	xy_table = xy_table[!is.na(xy_table$x), ]
	xy_table = xy_table[!is.na(xy_table$y), ]
	summary_table[4, ] = list('not NA', 
		                      summary_table[3,'left'] - nrow(xy_table),
	                          nrow(xy_table))

	xy_table = xy_table[!is.infinite(xy_table$x), ]
	xy_table = xy_table[!is.infinite(xy_table$y), ]
	summary_table[5, ] = list('not infinite', 
		                      summary_table[4,'left'] - nrow(xy_table),
	                          nrow(xy_table))
	return(list(xy_table, summary_table))
}

plot_values <- function(xy_table, xlab, ylab, bc_vec = c(), which_state=NA, text_size=20){
	# function to plot log2 expression or fold changes in column y against classifications in column x
	# coloured by x
	# Args:
	#   xy_table: table with filtered x and y values
	#   xlab: label on x-axis
	#   ylab: label on y-axis
	#   bc_vec: vector for plotting a subset of barcodes as points (boxplot still uses complete range)
	#   which_state: optional value to plot the expression for a single state
	#   text_size: size of the text on the axis
	# Return:
	#   ggplot object with plot
	if (!is.na(which_state)){
		xy_table = xy_table[xy_table$x == which_state, ]
	}
	if (length(bc_vec)>0){
		subset = xy_table[bc_vec,]
		points = geom_point(data=subset,aes(y=y), shape=19, size =0.9, position=position_jitter(width=.2))
	} else {
		points = geom_point(shape=19, size =0.9, position=position_jitter(width=.2))
	}
	ggplot(xy_table,aes(x = x, y=y, colour=x)) +
		theme(panel.background = element_rect(fill = "lavender")) +
		theme(strip.text.x = element_text(size = 28)) + geom_boxplot( outlier.colour=NA) +
		points +
		stat_summary(fun.y=median, aes(ymin=..y.., ymax=..y..), geom='errorbar', width=0.3, color='black', size=1.25) +
		theme(legend.position="none") +
		theme(axis.title = element_text(size = 20, face='bold')) +
		theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
		theme(axis.text = element_text(size = text_size)) +
		geom_hline(yintercept=0, colour = "grey30") +
		theme(plot.title = element_text(size=38))+
		ylab(ylab) + xlab(xlab)
}

correlate <- function(xy_table, title, day1, day2){
	fit = lm(xy_table$y ~ xy_table$x)
	coef = summary(fit)$coefficients
	spearman = cor.test(xy_table$x, xy_table$y, method='spearman')
	pearson = cor.test(xy_table$x, xy_table$y, method='pearson')
	test_text = sprintf("    Spearman's rank correlation rho
 data:  day %.0f and day %.0f
 S = %.0f, p-value = %0.4f
 alternative hypothesis: 
 true rho is not equal to 0
 sample estimates:
       rho 
 %0.7f 
 
    Pearson's product-moment correlation
 data:  day %.0f and day %.0f
 t = %.04f, df = %.0f, p-value = %0.4f
 alternative hypothesis: 
 true correlation is not equal to 0
 95 percent confidence interval:
  %0.8f  %0.8f
 sample estimates:
       cor 
 %0.7f ", day1, day2, spearman$statistic, spearman$p.value, spearman$estimate,
 day1, day2, pearson$statistic, pearson$parameter, pearson$p.value, pearson$conf.int[1],
 pearson$conf.int[2], pearson$estimate)
	test_table = tableGrob(rbind(title,test_text),rows=NULL, theme=ttheme_default(core=list(fg_params=list(fontsize=c(12,10), fontface=c(2L,1L)))))
	spearman_p = spearman$p.value
	pearson_p = pearson$p.value

	label = sprintf('y = %0.3f + %0.3fx\nn=%i; P=%0.3f; S=%0.3f',  coef[1,'Estimate'], coef[2,'Estimate'], nrow(xy_table), pearson_p, spearman_p)
	return(list(test_table, label))
}

plot_2vs9 <- function(xy_table, xlab, ylab, day1, day2, y_lim = ylim(-7.5,2.5), x_lim=xlim(-10,2), nCol=NULL, label_size=20, text_size=10){
	# function to plot fold change of day 2 against fold change after day 9, 11 or12
	# seperated by state
	# Args:
	#   xy_table: table with filtered x and y values and a state linked to each barcode
	#   title: title of the plot
	#   xlab: label on x-axis
	#   ylab: label on y-axis
	#   mem_day: day of the memory experiment (e.g. 9)
	#   text_size: size of the text on the axis
	# Return:
	#   ggplot object with plot and list of ggplot objects with pearon and spearman results
	state_vec = names(which(table(xy_table$state)>1))
	grob_list = list()
	fit_table = data.frame(state=state_vec, x=Inf, y=Inf, label=NA)
	for (state in state_vec){
		this_xy = xy_table[which(xy_table$state==state), ]
		if (nrow(this_xy)>3){
			test = correlate(this_xy, state, day1, day2)
			grob_list[[state]] = test[[1]]
			fit_table[fit_table$state==state,'label'] = test[[2]]
		}
	}
	
	aes = aes(x = x, y=y, colour=state)
	cat('\n\n')
	g = ggplot(xy_table, aes) +
		theme(panel.background = element_rect(fill = "lavender"))+
		theme(strip.text.x = element_text(size = 10)) +
		geom_point(shape=19, size =0.9, position=position_jitter(width=.2))  +
		theme(legend.position="none") +
		theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
		theme(axis.title = element_text(size=label_size)) +
		theme(text = element_text(size = text_size)) +
		geom_hline(yintercept=0, colour = "grey30") +
		geom_text(aes(x=x,y=y,label=label), data=fit_table, vjust=1, hjust=1, size=5) +
		ggtitle('P=Pearson; S=Spearman') + y_lim + x_lim +
		ylab(ylab) + xlab(xlab) + stat_smooth(method = "lm") + facet_wrap(~ state, ncol=nCol)
	return(list(g, grob_list))
}

plot_wilcox <- function(xy_table, state, poi, day, title, xlab, ylab="log2 (fold change)", text_size=20, p_cut=0.05){
	# function to perform wilcoxon tests for a certain state (e.g. cLADs), and if significant,
	# display the results and plot the log2 fold change that state against the values of the
	# other states.
	# Args:
	#   xy_table: table with filtered x and y values and a state linked to each barcode
	#   title: title of the plot
	#   xlab: label on x-axis
	#   ylab: label on y-axis
	#   text_size: size of the text on the axis
	# Return:
	#   ggplot object with plot and ggplot object with outcome of Wilcoxon-test statistiscs
	if (length(which(xy_table$x==state)) > 0){
		wc = wilcox.test(xy_table$y[xy_table$x==state],xy_table$y[xy_table$x!=state], conf.int=T)
		wc$data.name = sprintf('%s day%i %s vs NOT %s', poi, day, state, state)
		nlevels = length(levels(xy_table$x))
		p.value = p.adjust(wc$p, n = nlevels)
		if (wc$p.value < p_cut){
			notx = paste('NOT', state)
			xy_table$this_x[xy_table$x==state] = state
			xy_table$this_x[xy_table$x!=state] = notx
			xy_table$this_x = factor(xy_table$this_x, levels=c(state,notx))
			t = textGrob(sprintf('        Wilcoxon rank sum test with continuity correction\nadjusted p-value threshold:\n%.2g\ndata:%s day%i %s vs NOT %s\nW = %i, p-value = %0.5f\n alternative hypothesis: true location shift is not equal to 0\n95 percent confidence interval:\n %0.8f %0.8f\nsample estimates:\ndifference in location\n%0.7f', p_cut, poi, day, state, state, wc$statistic, p.value, wc$conf.int[1], wc$conf.int[2], wc$estimate))
			p = ggplot(xy_table,aes(x = this_x, y=y, colour=state)) +
				theme(panel.background = element_rect(fill = "lavender")) +
				theme(strip.text.x = element_text(size = 28)) + geom_boxplot(aes(colour=factor(this_x)), outlier.colour=NA) +
				geom_point(shape=19, position=position_jitter(width=.2))  + ggtitle(title) +
				stat_summary(fun.y=median, aes(ymin=..y.., ymax=..y..), geom='errorbar', width=0.3, color='black', size=1.25) +
				theme(legend.position="none") +
				theme(axis.title = element_text(size = 28)) +
				theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
				theme(axis.text = element_text(size = text_size)) +
				geom_hline(yintercept=0, colour = "grey30") +
				theme(plot.title = element_text(size=24)) +
				ylab(ylab) + xlab(xlab)
			return(list(p,t))
		} else{
			return(NA)
		}
	} else {
		return(NA)
	}
	
}

```

## Create plots
We can make an easy loop function to first create all plots for each classification and POI.


```{r create_plots_main, cache=F}
categorical = c('lad', 'chrom', 'rep_name', 'rep_family', 'rep_class')
discrete = c('cpg_distance', 'timing')
empty_list = list()
for (class in c(categorical, discrete)){
	empty_list[[class]] = list()
}
# lists within lists to store the data
plots_2 = plots_9 = plots_exp = plots_2vs9 = plots_expVs9 = stats_2 = stats_9 = stats_exp = corr_2vs9 = stats_2vs9 = corr_expVs9 = stats_expVs9 = empty_list

empty_list = list()
for (class in categorical){
	empty_list[[class]] = list()
}
plots_wilcox_2 = plots_wilcox_9 = empty_list

class_vec = c('lad', 'chrom', 'rep_class')
for (poi in c('KRAB', 'G9a', 'CBX')){
	for (class in class_vec){
		# for different classes use descriptive x-axis labels
		# and if there is a lot of different classes, use a smaller font 
		if (class=='rep_family'){
			text_size=10
			xlab = 'repeat family'
			if (poi == 'KRAB'){
				ncol = 5
			} else {
				ncol = 4
			}
			xtype = 'factor'
			
		} else if(class=='lad'){
			xlab = 'LAD-state'
			text_size = 20
			ncol = 2
			xtype = 'factor'
		} else if(class=='chrom'){
			xlab = 'Chromatin state'
			text_size = 20
			ncol = 3
			xtype = 'factor'
		}else if(class=='cpg_distance'){
			xlab = 'distance to nearest cpg'
			text_size = 20
			ncol = 3
			xtype = 'value'
		}else if(class=='rep_class'){
			xlab = 'repeat class'
			text_size = 20
			ncol = 3
			xtype = 'factor'
		}
		# the column name of the fold-change on day 2
		condition_2 = paste0(poi, '_GPvsG_day2')
		# filter the data and get table with x and y values and a statistics table
		# x and y value table:
		# 	x: a factor of the classification
		# 	y: the log2 transformed fold changes
		# statistics table:
		#	for each selection criteria how many barcodes were kicked out and how many were left.
		fc_2 = filter_data(fc_table, class,  condition_2, poi, 2, xtype, 'log2')
		xy_table_2= fc_2[[1]]
		# create a grob for plotting the statistics and a plot for the fold-changes
		stats_2[[class]][[poi]] = tableGrob(fc_2[[2]], rows=NULL)
		plots_2[[class]][[poi]] = plot_values(xy_table_2, xlab, 'log2(fold change)', 
			                                  text_size=text_size)
		fc_9 = list()
		condition_9 = list()
		# do the same for later day
		stats_9[[class]][[poi]] = list()
		plots_9[[class]][[poi]] = list()
		for (day in days[[poi]]){
			d = as.character(day)
			condition_9[[d]] = sprintf('%s_GPvsG_day%i',poi,day)

			fc_9[[d]] = filter_data(fc_table, class,  condition_9[[d]], poi, day, xtype, 'log2')
			xy_table_9 = fc_9[[d]][[1]]
			stats_9[[class]][[poi]][[d]] = tableGrob(fc_9[[d]][[2]], rows=NULL)
			plots_9[[class]][[poi]][[d]] = plot_values(xy_table_9, xlab, 
				   									   'log2(fold change)', 
				                                       text_size=text_size)

		}
		
		# and for the expression at day 2
		condition_exp = paste0(poi,'_GAL4_exp_2')
		fc = filter_data(fc_table, class, condition_exp, poi, 2, xtype, 'log2')
		xy_table = fc[[1]]
		
		stats_exp[[class]][[poi]] = tableGrob(fc[[2]], rows=NULL)
		plots_exp[[class]][[poi]] = plot_values(xy_table, xlab, 'log2(expression)', 
			                                      text_size=text_size)

		if (class %in% categorical){
			plots_wilcox_2[[class]][[poi]] = list()
			plots_wilcox_9[[class]][[poi]] = list()
			# and now the wilcoxon tests for each state in a class
			for (state in unique(fc_table[,class])){
				wc=plot_wilcox(fc_2[[1]], state, poi, 2,  xlab, sprintf("fold change day %i\n%s",2,poi), 'log2(fold-change)')
				if (any(!is.na(wc))){
					plots_wilcox_2[[class]][[poi]][[state]] = wc
				}
			}
			for (day in days[[poi]]){
				d = as.character(day)
				plots_wilcox_9[[class]][[poi]][[d]] = list()
				for (state in unique(fc_table[,class])){
					wc=plot_wilcox(fc_9[[d]][[1]], state, poi, day,  xlab, sprintf("fold change day%i\n%s",day,poi), 'log2(fold-change)')
					if (any(!is.na(wc))){
						plots_wilcox_9[[class]][[poi]][[d]][[state]] = wc
					}
				}
			}
		}
		

		stats_expVs9[[class]][[poi]] = list()
		plots_expVs9[[class]][[poi]] = list()
		corr_expVs9[[class]][[poi]] = list()
		stats_2vs9[[class]][[poi]] = list()
		plots_2vs9[[class]][[poi]] = list()
		corr_2vs9[[class]][[poi]] = list()
		lim_exp = xlim(min(xy_table$y), max(xy_table$y))
		lim_2 = xlim(min(xy_table_2$y), max(xy_table_2$y))
		for (day in days[[poi]]){
			d = as.character(day)
			# and plots of base expression vs fold change on day 9
			fc_expVs9 = filter_data(fc_table, condition_exp,  condition_9[[d]], poi, day, 'log2', 'log2', class)
			xy_table = fc_expVs9[[1]]
			stats_expVs9[[class]][[poi]][[d]] = tableGrob(fc_expVs9[[2]], rows=NULL)
			p = plot_2vs9(xy_table, 'log2(expression)', 'log2(fold change)', 2,
				          day, x_lim=lim_exp, nCol=ncol, text_size=text_size)
			plots_expVs9[[class]][[poi]][[d]] = p[[1]]
			
			corr_expVs9[[class]][[poi]][[d]] = p[[2]]

			# and plots of fold change on day 2 vs day 9
			fc_2vs9 = filter_data(fc_table, condition_2,  condition_9[[d]], poi, day, 'log2', 'log2', class)
			xy_table = fc_2vs9[[1]]
			stats_2vs9[[class]][[poi]][[d]] = tableGrob(fc_2vs9[[2]], rows=NULL)
			p = plot_2vs9(xy_table, 'log2(fold change)', 'log2(fold change)', 2,
				          day,x_lim = lim_2, nCol=ncol, text_size=text_size)
			plots_2vs9[[class]][[poi]][[d]] = p[[1]]
			
			corr_2vs9[[class]][[poi]][[d]] = p[[2]]
		}

	}
	
}

```
```{r save_plots, echo=F}

save(plots_2, plots_9, plots_exp, plots_wilcox_2, plots_wilcox_9, plots_2vs9, plots_expVs9, stats_2, stats_9, stats_exp, corr_2vs9, stats_2vs9, corr_expVs9, stats_expVs9, file = 'plots.rData')

```
```{r}
xy_table = filter_data(fc_table, 'KRAB_GAL4_exp_2', 'KRAB_GAL4.POI_exp_2','KRAB', 2, 'log2', 'log2')
svg('KRAB_day2.svg')
ggplot(xy_table[[1]], aes(x=x, y=y)) + theme(panel.background = element_rect(fill="lavender")) + geom_point( shape=19, size=1, colour='RED') + geom_abline() + xlab('log2(expr GAL4 + 0.1)') + ylab('log2(expr GAL4-KRAB + 0.1)') + ggtitle('GAL4-KRAB vs GAL4 day 2') + theme(plot.title = element_text(size=24),text = element_text(size=20))
dev.off()

xy_table = filter_data(fc_table, 'G9a_GAL4_exp_2', 'G9a_GAL4.POI_exp_2','G9a', 2, 'log2', 'log2')
svg('G9a_day2.svg')
ggplot(xy_table[[1]], aes(x=x, y=y)) + theme(panel.background = element_rect(fill="lavender")) + geom_point( shape=19, size=1, colour='RED') + geom_abline() + xlab('log2(expr G9a + 0.1)') + ylab('log2(expr GAL4-G9a + 0.1)') + ggtitle('GAL4-G9a vs GAL4 day 2') + theme(plot.title = element_text(size=24),text = element_text(size=20))
dev.off()


```
```{r}
xy_table = filter_data(fc_table, 'KRAB_GAL4.POI_exp_2', 'KRAB_GAL4_exp_2','KRAB', 2, 'value', 'value', state='lad')
xy_lad_2 = xy_table[[1]]
xy_lad_2$lad = '-'
xy_lad_2[which(xy_lad_2$state=='ciLAD'|xy_lad_2$state=='fiLAD'),'lad'] = 'interLAD'
xy_lad_2[which(xy_lad_2$state=='cLAD'|xy_lad_2$state=='fLAD'),'lad'] = 'LAD'

xy_lad_2$fc = log2(xy_lad_2$x / xy_lad_2$y)

wc = wilcox.test(xy_lad_2[xy_lad_2$lad=='LAD','fc'], xy_lad_2[xy_lad_2$lad=='interLAD','fc'], conf.int=T)
t = textGrob(sprintf('        Wilcoxon rank sum test with continuity correction\ndata: KRAB day 2 LAD vs interLAD\nW = %i, p-value = %0.5g\n alternative hypothesis: true location shift is not equal to 0\n95 percent confidence interval:\n %0.8f %0.8f\nsample estimates:\ndifference in location\n%0.7f', wc$statistic, wc$p.value, wc$conf.int[1], wc$conf.int[2], wc$estimate))
p = ggplot(xy_lad_2,aes(x = factor(lad, levels=c('LAD', 'interLAD')), y=fc, colour=lad)) +
				theme(panel.background = element_rect(fill = "lavender")) +
				theme(strip.text.x = element_text(size = 28)) +
				geom_point(shape=19, position=position_jitter(width=.9)) + geom_violin(alpha=0.3) + ggtitle('KRAB day 2 fold-change by lamina association') +
				stat_summary(fun.y=median, aes(ymin=..y.., ymax=..y..), geom='errorbar', width=0.3, color='black', size=1.25) +
				theme(legend.position="none") +
				theme(axis.title = element_text(size = 28)) +
				theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
				theme(axis.text = element_text(size = text_size)) +
				geom_hline(yintercept=0, colour = "grey30") +
				theme(plot.title = element_text(size=24)) +
				ylab('log2(fold-change)') + xlab('lamina association')
svg('KRAB_lad_day2.svg', width=8, height=15)
grid.arrange(t,p, heights = c(1,3))
dev.off()


xy_table = filter_data(fc_table, 'KRAB_GAL4.POI_exp_11', 'KRAB_GAL4_exp_11','KRAB', 11, 'value', 'value', state='lad')
xy_lad_11 = xy_table[[1]]
xy_lad_11$lad = '-'
xy_lad_11[which(xy_lad_11$state=='ciLAD'|xy_lad_11$state=='fiLAD'),'lad'] = 'interLAD'
xy_lad_11[which(xy_lad_11$state=='cLAD'|xy_lad_11$state=='fLAD'),'lad'] = 'LAD'

xy_lad_11$fc = log2(xy_lad_11$x / xy_lad_11$y)

wc = wilcox.test(xy_lad_11[xy_lad_11$lad=='LAD','fc'], xy_lad_11[xy_lad_11$lad=='interLAD','fc'], conf.int=T)
t = textGrob(sprintf('        Wilcoxon rank sum test with continuity correction\ndata: KRAB day 11 LAD vs interLAD\nW = %i, p-value = %0.5g\n alternative hypothesis: true location shift is not equal to 0\n95 percent confidence interval:\n %0.8f %0.8f\nsample estimates:\ndifference in location\n%0.7f', wc$statistic, wc$p.value, wc$conf.int[1], wc$conf.int[2], wc$estimate))
p = ggplot(xy_lad_11,aes(x = factor(lad, levels=c('LAD', 'interLAD')), y=fc, colour=lad)) +
				theme(panel.background = element_rect(fill = "lavender")) +
				theme(strip.text.x = element_text(size = 28)) +
				geom_point(shape=19, position=position_jitter(width=.9)) + geom_violin(alpha=0.3) + ggtitle('KRAB day 11 fold-change by lamina association') +
				stat_summary(fun.y=median, aes(ymin=..y.., ymax=..y..), geom='errorbar', width=0.3, color='black', size=1.25) +
				theme(legend.position="none") +
				theme(axis.title = element_text(size = 28)) +
				theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
				theme(axis.text = element_text(size = text_size)) +
				geom_hline(yintercept=0, colour = "grey30") +
				theme(plot.title = element_text(size=24)) +
				ylab('log2(fold-change)') + xlab('lamina association')
svg('KRAB_lad_day11.svg', width=8, height=15)
grid.arrange(t,p, heights = c(1,3))
dev.off()



xy_rep = filter_data(fc_table, 'KRAB_GAL4.POI_exp_2', 'KRAB_GAL4_exp_2','KRAB', 2, 'value', 'value', state='rep_class')[[1]]
xy_rep$fc = log2(xy_rep$x / xy_rep$y)
svg('KRAB_repeat_day2.svg', width=8, height=10)
ggplot(xy_rep,aes(x = state, y=fc, colour=state)) +
				theme(panel.background = element_rect(fill = "lavender")) +
				theme(strip.text.x = element_text(size = 28)) +
				geom_point(shape=19, position=position_jitter(width=.9)) + geom_boxplot(alpha=0.3) + ggtitle('KRAB day 2 fold-change for integrations in repetitive elements') +
				stat_summary(fun.y=median, aes(ymin=..y.., ymax=..y..), geom='errorbar', width=0.3, color='black', size=1.25) +
				theme(legend.position="none") +
				theme(axis.title = element_text(size = 28)) +
				theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
				theme(axis.text = element_text(size = text_size)) +
				geom_hline(yintercept=0, colour = "grey30") +
				theme(plot.title = element_text(size=24)) +
				ylab('log2(fold-change)') + xlab('repeat class')
dev.off()


svg('G9a_repeat_day2.svg', width=8, height=10)
xy_rep = filter_data(fc_table, 'G9a_GAL4.POI_exp_2', 'G9a_GAL4_exp_2','G9a', 2, 'value', 'value', state='rep_class')[[1]]
xy_rep$fc = log2(xy_rep$x / xy_rep$y)
ggplot(xy_rep,aes(x = state, y=fc, colour=state)) +
				theme(panel.background = element_rect(fill = "lavender")) +
				theme(strip.text.x = element_text(size = 28)) +
				geom_point(shape=19, position=position_jitter(width=.9)) + geom_boxplot(alpha=0.3) + ggtitle('G9a day 2 fold-change for integrations in repetitive elements') +
				stat_summary(fun.y=median, aes(ymin=..y.., ymax=..y..), geom='errorbar', width=0.3, color='black', size=1.25) +
				theme(legend.position="none") +
				theme(axis.title = element_text(size = 28)) +
				theme(axis.text.x = element_text(hjust = 1, angle = 90)) +
				theme(axis.text = element_text(size = text_size)) +
				geom_hline(yintercept=0, colour = "grey30") +
				theme(plot.title = element_text(size=24)) +
				ylab('log2(fold-change)') + xlab('repeat class')
dev.off()
```

```{r}
sessionInfo()
getwd()
date()
paste("Run time: ",format(Sys.time()-StartTime))
```
