# r document van Steensel lab

# Gene repression in LADs
## Christ Leemans, 03-11-2016 - 26-07-2017

## Introduction
Generally speaking, genes inside lamina associated domains are not or very lowly expressed. These genes can either be actively repressed by their DNA context (e.g. heterochromatin, lamina association), or simply be inactive (because essential factors for expression are missing?). Yet another group of genes seem to evade gene repression in the context of lamina associated domains. In this report I would like to give an overview of what we have found so far.

## Different promoter classes of lamina associated transcripts.

Comparing autonomous promoter activity measured by SuRE and endogenous promoter activity


## library and data loading
```{r, fig.width=10, fig.height=10, echo=FALSE, fig.width=10, fig.height=10}
library(reshape2)
library(rtracklayer)
library(ggplot2)
library(gridExtra)
library(plyr)
library(scales)
library(grid)
library(gtable)
library(affy)
library(limma)
library(biomaRt)
## FROM STACKOVERFLOW:
## https://stackoverflow.com/questions/12539348/ggplot-separate-legend-and-plot
g_legend<-function(a.gplot){
    tmp <- ggplot_gtable(ggplot_build(a.gplot))
    leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
    legend <- tmp$grobs[[leg]]
    return(legend)
}

pseudo_log10 <- function(val_vec){
    Pseud=min(val_vec[val_vec > 0], na.rm=TRUE)/2
    val_vec = val_vec + Pseud
    return(log10(val_vec))
}

## get a table with matching sets
## table = complete table to take matching sets from
## class_col = column name of class of interest
## class = name of class to match the set on
## order_on = column name to order on
matchSet <- function(table, class_col, class, order_on){
  o_vec = order(table[,order_on])
  o_table = table[o_vec, ]
  setA = which(o_table[,class_col]==class)
  setB = c(setA + 1, setA -1)
  ## check if setB is all within the possible indexes
  setB = setB[setB %in% 1:length(o_vec)]
  ## can also return o_table[unique(c(setA, setB)), ]
  ## but this way order is perserved.
  i_vec = o_vec[unique(c(setA, setB))]
  return(table[i_vec[order(i_vec)], ])
}


COLi<-"#00BBFF11" #dot color for iLAD promoters
COL_lad<-c("#FF0000", "#0077FF")
names(COL_lad)<-c('LAD', 'iLAD')

#color vector for plotting:
COL_class<-c("#A020F0", "#FFA500", "#006400", "#7e7e7e", "#0077FF")
names(COL_class)<-c("repressed", "escaper", "inactive", 'boundary', 'iLAD')

COL<-c("#A020F0", "#FFA500", "#006400")
names(COL)<-c("repressed", "escaper", "inactive")



id_table = read.table('../raw_data/transcript.table', stringsAsFactors=F,
                      row.names=1, col.names=c('transcript_id', 'gene_id',
                                               'symbol'))
load("~joris/mydata/git/SuRE/Joris//analysis_postNBT/Gencode_DF_generation_170707/gencode.sure.170712.rda")


P<-gencode.sure.170712[,c('chr', 'strand', 'txStart', 'txEnd', 'name', 'name2',
                          'tss', 'distance.to.previous.tss',
                          'k562.combined.45.55.sense', 'HEPG2.sense',
                          'HT1080.sense', 'gro.cap.1kb.sense',
                          'encode.cage.reprocessed.1kb.sense',
                          'nr.of.tissues.in.which.expressed.max')]

names(P)[9:14]<-c("SuRE_K562", "SuRE_HepG2", "SuRE_HT1080", "GROcap_K562",
                  "CAGE_K562_encode", 'tissues_expressed')

rownames(P) = P$name


for (col in c('SuRE_K562', 'SuRE_HepG2', 'SuRE_HT1080', 'GROcap_K562')){
    P[,col] = pseudo_log10(P[,col])
}

P$gene_id = id_table[P$name,'gene_id']

most_active <- function(P){
  result = ddply(P, .(gene_id), function(x){
      if (nrow(x)==1){
          result = x[1,]
      } else {
          result = x[order(x$SuRE_K562, decreasing=T)[1],]
      }
      return(result)
  })
  rownames(result) = result$name
  return(result)
}

p_most_active = most_active(P)

p_other = P[!rownames(P)%in%rownames(p_most_active), ]
p_new_names = rownames(p_most_active)
while (nrow(p_other) > 0){
  p_new = P[p_new_names, ]
  active_gr = makeGRangesFromDataFrame(data.frame(seqnames=p_new$chr,
                                                 start=p_new$tss,
                                                 end=p_new$tss,
                                                 strand=p_new$strand),
                                                 keep.extra.columns=TRUE)
  other_gr = makeGRangesFromDataFrame(data.frame(seqnames=p_other$chr,
                                                 start=p_other$tss,
                                                 end=p_other$tss,
                                                 strand=p_other$strand),
                                                 keep.extra.columns=TRUE)
  o = findOverlaps(active_gr,other_gr, maxgap=500, ignore.strand=FALSE)
  sub_o = o[p_new[queryHits(o), 'gene_id'] == p_other[subjectHits(o), 'gene_id']]
  p_other = p_other[-subjectHits(sub_o), ]
  p_active = most_active(p_other)
  p_other = p_other[!rownames(p_other)%in%rownames(p_active), ]
  p_new_names = c(p_new_names, rownames(p_active))
}

p_complete = rownames(P)

P = P[rownames(P)%in%p_new_names, ]

gene_gr <-makeGRangesFromDataFrame(data.frame(seqnames=P$chr,
                                            start=P$txStart,
                                            end=P$txEnd,
                                            strand=P$strand,
                                            name=P$name,
                                            tss=P$tss),
                                            keep.extra.columns=TRUE)
names(gene_gr) = P$name
tss_gr = gene_gr
ranges(tss_gr) = IRanges(gene_gr$tss,
                       gene_gr$tss)
names(tss_gr) = P$name
export.bed(tss_gr, '../raw_data/tss.bed')

## get LAD data for K562
LAD_K562 = import.bed('~c.leemans/mydata/data/tracks/hg19/cl20161019_LAD_continuous_2state_K562.bed')
## to keep with Joris's previous analysis, let's assign a state to every promoter
o = findOverlaps(tss_gr, LAD_K562[LAD_K562$name=='LAD'])
P$LAD_K562 = 0
P$LAD_K562[queryHits(o)] = 1

LAD_K562 = import.bed('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170821_GCF4577_K562_p13_LMNB1/results/HMM/K562_lamina_10kb_LADs.bed')
o = findOverlaps(tss_gr, LAD_K562)
P$LAD_K562_seq = 0
P$LAD_K562_seq[queryHits(o)] = 1

## now repeat for HT1080
LAD_HT1080 = import.bed('~c.leemans/mydata/data/tracks/hg19/cl20170713_HT1080_LAD_continuous_2state.bed')
## to keep with Joris's previous analysis, let's assign a state to every promoter
o = findOverlaps(tss_gr, LAD_HT1080[LAD_HT1080$name=='LAD'])
P$LAD_HT1080 = 0
P$LAD_HT1080[queryHits(o)] = 1


LAD_HepG2 = import.bed('~t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/HMM/HEPG2_lamina_LADs.bed')
o = findOverlaps(tss_gr, LAD_HepG2)
P$LAD_HepG2 = 0
P$LAD_HepG2[queryHits(o)] = 1

```

## CAGE calculation
I am using different data from Joris, since HT1080 was not in the ENCODE data, only in the Fantom data. There are some mayor differences between these datasets, whereas previous data was poly-A selected and PCR-amplified, Fantom did not select, nor did it PCR-amplify


```bash
bedtools intersect -c -s -a <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) -b ~c.leemans/mydata/data/tracks/hg19/K562_ENCODE_biol_rep1.CNhs12334.10824-111C5.hg19.nobarcode.bam > results/tss_CAGE_K562_rep1.bed

bedtools intersect -c -s -a <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) -b ~c.leemans/mydata/data/tracks/hg19/K562_ENCODE_biol_rep2.CNhs12335.10825-111C6.hg19.nobarcode.bam > results/tss_CAGE_K562_rep2.bed

bedtools intersect -c -s -a <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) -b ~c.leemans/mydata/data/tracks/hg19/HT-1080.CNhs11860.10758-110E2.hg19.nobarcode.bam  > results/tss_CAGE_HT1080.bed

bedtools intersect -c -s -a <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) -b ~c.leemans/mydata/data/tracks/hg19/HepG2_ENCODE_biol_rep1.CNhs12328.10818-111B8.hg19.nobarcode.bam > results/tss_CAGE_HepG2_rep1.bed

bedtools intersect -c -s -a <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) -b ~c.leemans/mydata/data/tracks/hg19/HepG2_ENCODE_biol_rep2.CNhs12329.10819-111B9.hg19.nobarcode.bam > results/tss_CAGE_HepG2_rep2.bed

bedtools intersect -c -s -a <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) -b ~c.leemans/mydata/data/tracks/hg19/HepG2_ENCODE_biol_rep3.CNhs12330.10820-111C1.hg19.nobarcode.bam > results/tss_CAGE_HepG2_rep3.bed

```

```{r}
K562_CAGE_rep1 = read.table('../results/tss_CAGE_K562_rep1.bed')
K562_CAGE_rep2 = read.table('../results/tss_CAGE_K562_rep2.bed')
HT1080_CAGE = read.table('../results/tss_CAGE_HT1080.bed')
HepG2_CAGE_rep1 = read.table('../results/tss_CAGE_HepG2_rep1.bed')
HepG2_CAGE_rep2 = read.table('../results/tss_CAGE_HepG2_rep2.bed')
HepG2_CAGE_rep3 = read.table('../results/tss_CAGE_HepG2_rep3.bed')


K562_CAGE = rowSums(cbind(K562_CAGE_rep1[,7], K562_CAGE_rep2[,7]))
HepG2_CAGE = rowSums(cbind(HepG2_CAGE_rep1[,7], HepG2_CAGE_rep2[,7],
                           HepG2_CAGE_rep3[,7]))

P$CAGE_K562 = K562_CAGE[p_complete%in%rownames(P)]
P$CAGE_HT1080 = HT1080_CAGE[p_complete%in%rownames(P),7]
P$CAGE_HepG2 = HepG2_CAGE[p_complete%in%rownames(P)]

pseudo_log10 <- function(val_vec){
    Pseud=min(val_vec[val_vec > 0], na.rm=TRUE)/2
    val_vec = val_vec + Pseud
    return(log10(val_vec))
}
for (col in c('CAGE_K562', 'CAGE_HT1080', 'CAGE_HepG2')){
    P[,col] = pseudo_log10(P[,col])
}


```


## DNAse accessibility
Maybe DNAse signal can be used as an extra proxy so we can build a multi-dimensional classifier with CAGE and DNAse.

```bash
bwtool summary -header <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) ~/mydata/data/tracks/hg19/DNase-seq/K562_rep1_ENCFF001BSF.bigWig /dev/stdout > results/tss_DNAse_K562_rep1.txt

bwtool summary -header <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) ~/mydata/data/tracks/hg19/DNase-seq/K562_rep2_ENCFF001DPE.bigWig /dev/stdout > results/tss_DNAse_K562_rep2.txt

bwtool summary -header <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) ~/mydata/data/tracks/hg19/DNase-seq/HT1080_rep1_ENCFF930ZWL.bigWig /dev/stdout > results/tss_DNAse_HT1080_rep1.txt

bwtool summary -header <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) ~/mydata/data/tracks/hg19/DNase-seq/HT1080_rep2_ENCFF760VXZ.bigWig /dev/stdout > results/tss_DNAse_HT1080_rep2.txt

bwtool summary -header <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, $5, $6}' workspace/tss.bed) ~/mydata/data/tracks/hg19/DNase-seq/GSM2400287_ENCFF646UXG_signal_of_unique_reads_hg19.bigWig /dev/stdout > results/tss_DNAse_HepG2.txt



```

```{r, fig.width=10, fig.height=10}
K5_DNAse_rep1 = read.table('../results/tss_DNAse_K562_rep1.txt')
K5_DNAse_rep2 = read.table('../results/tss_DNAse_K562_rep2.txt')
K5_DNAse_rep1[,7][is.na(K5_DNAse_rep1[,7])] = K5_DNAse_rep2[,7][is.na(K5_DNAse_rep2[,7])] = 0
cor = cor(K5_DNAse_rep1[,7], K5_DNAse_rep2[,7])
ggplot(data.frame(rep1=K5_DNAse_rep1[,7], rep2=K5_DNAse_rep2[,7]),
       aes(x=log10(rep1), y=log10(rep2))) +
    ggtitle(paste('Max DNAse signal K562 rep1 vs rep2\ncor=', cor)) +
    geom_point(size=0.1, alpha=0.1)

DNAse_K562 = rowMeans(cbind(K5_DNAse_rep1[,7], K5_DNAse_rep2[,7]))
P$DNAse_K562 = DNAse_K562[p_complete%in%rownames(P)]

cor = cor(P$GROcap_K562, P$DNAse_K562)
ggplot(P, aes(x=log10(GROcap_K562), y=log10(DNAse_K562))) +
    ggtitle(paste('K562 DNAse vs K562 GROcap\ncor=', cor)) +
    geom_point(size=0.1, alpha=0.1)

cor = cor(P$CAGE_K562, P$DNAse_K562)
ggplot(P, aes(x=log10(CAGE_K562), y=log10(DNAse_K562))) +
    ggtitle(paste('K562 DNAse vs K562 Fantom CAGE\ncor=', cor)) +
    geom_point(size=0.1, alpha=0.1)

HT_DNAse_rep1 = read.table('../results/tss_DNAse_HT1080_rep1.txt')
HT_DNAse_rep2 = read.table('../results/tss_DNAse_HT1080_rep2.txt')
HT_DNAse_rep1[,7][is.na(HT_DNAse_rep1[,7])] = HT_DNAse_rep2[,7][is.na(HT_DNAse_rep2[,7])] = 0
cor = cor(HT_DNAse_rep1[,7], HT_DNAse_rep2[,7])
ggplot(data.frame(rep1=HT_DNAse_rep1[,7], rep2=HT_DNAse_rep2[,7]),
       aes(x=log10(rep1), y=log10(rep2))) +
    ggtitle(paste('Max DNAse signal HT1080 rep1 vs rep2\ncor=', cor)) +
    geom_point(size=0.1, alpha=0.1)


DNAse_HT1080 = rowMeans(cbind(HT_DNAse_rep1[,7], HT_DNAse_rep2[,7]))
P$DNAse_HT1080 = DNAse_HT1080[p_complete%in%rownames(P)]



cor = cor(P$GROcap_K562, P$DNAse_HT1080)
ggplot(P, aes(x=log10(GROcap_K562), y=log10(DNAse_HT1080))) +
    ggtitle(paste('HT1080 DNAse vs K562 GROcap\ncor=', cor)) +
    geom_point(size=0.1, alpha=0.1)

cor = cor(P$CAGE_HT1080, P$DNAse_HT1080)
ggplot(P, aes(x=log10(CAGE_HT1080), y=log10(DNAse_HT1080))) +
    ggtitle(paste('HT1080 DNAse vs HT1080 Fantom CAGE\ncor=', cor)) +
    geom_point(size=0.1, alpha=0.1)

HepG2_DNAse = read.table('../results/tss_DNAse_HepG2.txt')
P$DNAse_HepG2 = HepG2_DNAse[p_complete%in%rownames(P), 7]

for (col in c('DNAse_K562', 'DNAse_HT1080', 'DNAse_HepG2')){
    P[,col] = pseudo_log10(P[,col])
}

```

**conclusion**
There is not a lot of correlation, but there is some, hopefully we can combine DNAse and CAGE.


```{r}

create_RM <-function(data, x, y, lad){
    #then calculate running mean for iLAD promoters:
    #sort by SuRE and then random for ties
    o = order(data[,x],sample(c(1:nrow(data))))

    x_sorted = data[o,x]
    y_sorted = data[o,y]
    lad_sorted = data[o,lad]

    n<-60 #number of windows
    w<-501 #window width (number of datapoints); if n*w > nrow(P) then windows overlap
    s<-round(seq(from=w/2+0.0001, to=nrow(data)-w/2, length.out=n))
    RM<-data.frame(x.low=rep(NA,n), x.mean=rep(NA,n), x.hi=rep(NA,n), y.lad=rep(NA,n), y.ilad=rep(NA,n))
    RM$x.low=x_sorted[s-floor(w/2)]
    for(i in 1:n){RM$x.mean[i]=mean(x_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))], na.rm=TRUE)}
    RM$x.hi=x_sorted[s+floor(w/2)]
    for(i in 1:n)
      {t<-data.frame(LAD=lad_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))],
                     y=y_sorted[(s[i]-floor(w/2)):(s[i]+floor(w/2))])
       RM$y.lad[i]<-mean(t$y[t$LAD==1], na.rm=TRUE)
       RM$y.ilad[i]<-mean(t$y[t$LAD==0], na.rm=TRUE)
      }
    #add first datapoint (SuRE equals pseudocount)
    RM1<-RM[0,] #empty df
    RM1[1,]<-c(rep(min(x_sorted),3), mean(y_sorted[x_sorted==min(x_sorted) & lad_sorted==1]), mean(y_sorted[x_sorted==min(x_sorted) & lad_sorted==0]))
    RM<-rbind(RM1, RM)
    rm(RM1)
    return(RM)
}

RM_GRO = create_RM(P, 'SuRE_K562', 'GROcap_K562', lad='LAD_K562')
RM_GRO_seq = create_RM(P, 'SuRE_K562', 'GROcap_K562', lad='LAD_K562_seq')
RM_K562 = create_RM(P, 'SuRE_K562', 'CAGE_K562', lad='LAD_K562')
RM_HT1080 = create_RM(P, 'SuRE_HT1080', 'CAGE_HT1080', lad='LAD_HT1080')
RM_HepG2 = create_RM(P, 'SuRE_HepG2', 'CAGE_HepG2', lad='LAD_HepG2')
P$LRS_GROcap <- P$GROcap_K562 - approx(x=RM_GRO$x.mean, y=RM_GRO$y.ilad, xout=P$SuRE_K562, rule=2)$y

P$LRS_GROcap_seq <- P$GROcap_K562 - approx(x=RM_GRO_seq$x.mean, y=RM_GRO_seq$y.ilad, xout=P$SuRE_K562, rule=2)$y

P$LRS_K562 <- P$CAGE_K562 - approx(x=RM_K562$x.mean, y=RM_K562$y.ilad, xout=P$SuRE_K562, rule=2)$y

P$LRS_HT1080 <- P$CAGE_HT1080 - approx(x=RM_HT1080$x.mean, y=RM_HT1080$y.ilad, xout=P$SuRE_HT1080, rule=2)$y

P$LRS_HepG2 <- P$CAGE_HepG2 - approx(x=RM_HepG2$x.mean, y=RM_HepG2$y.ilad, xout=P$SuRE_HepG2, rule=2)$y

lad_names_K562 = c(LAD=paste0('LAD; n=', table(P$LAD_K562)['1']),
                   iLAD=paste0('iLAD; n=', table(P$LAD_K562)['0']))
P$lad_K562_n = factor(ifelse(P$LAD_K562==1, lad_names_K562['LAD'], lad_names_K562['iLAD']))
COL_lad_K562_n = COL_lad
names(COL_lad_K562_n) = lad_names_K562

lad_names_K562 = c(LAD=paste0('LAD; n=', table(P$LAD_K562_seq)['1']),
                   iLAD=paste0('iLAD; n=', table(P$LAD_K562_seq)['0']))
P$lad_K562_seq_n = factor(ifelse(P$LAD_K562_seq==1, lad_names_K562['LAD'], lad_names_K562['iLAD']))
COL_lad_K562_seq_n = COL_lad
names(COL_lad_K562_seq_n) = lad_names_K562



lad_names_HT1080 = c(LAD=paste0('LAD; n=', table(P$LAD_HT1080)['1']),
                     iLAD=paste0('iLAD; n=', table(P$LAD_HT1080)['0']))
P$lad_HT1080_n = factor(ifelse(P$LAD_HT1080==1, lad_names_HT1080['LAD'], lad_names_HT1080['iLAD']))
COL_lad_HT1080_n = COL_lad
names(COL_lad_HT1080_n) = lad_names_HT1080

lad_names_HepG2 = c(LAD=paste0('LAD; n=', table(P$LAD_HepG2)['1']),
                     iLAD=paste0('iLAD; n=', table(P$LAD_HepG2)['0']))
P$lad_HepG2_n = factor(ifelse(P$LAD_HepG2==1, lad_names_HepG2['LAD'], lad_names_HepG2['iLAD']))
COL_lad_HepG2_n = COL_lad
names(COL_lad_HepG2_n) = lad_names_HepG2


RM_melt = melt(RM_GRO, measure.vars=c('y.ilad', 'y.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='y.lad', lad_names_K562['LAD'], lad_names_K562['iLAD'])
ggplot(P, aes(x=SuRE_K562, y=GROcap_K562, color=lad_K562_n)) +
    geom_point(data=P[P$LAD_K562==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_K562==1, ], size=0.5, alpha=0.2) +
    ggtitle('GROcap vs SuRE K562') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=x.mean, y=value, color=variable), size=1) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_K562_n)

RM_melt = melt(RM_GRO_seq, measure.vars=c('y.ilad', 'y.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='y.lad', lad_names_K562['LAD'], lad_names_K562['iLAD'])
ggplot(P, aes(x=SuRE_K562, y=GROcap_K562, color=lad_K562_seq_n)) +
    geom_point(data=P[P$LAD_K562==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_K562==1, ], size=0.5, alpha=0.2) +
    ggtitle('GROcap vs SuRE K562') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=x.mean, y=value, color=variable), size=1) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_K562_seq_n)



RM_melt = melt(RM_K562, measure.vars=c('y.ilad', 'y.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='y.lad', lad_names_K562['LAD'], lad_names_K562['iLAD'])
ggplot(P, aes(x=SuRE_K562, y=CAGE_K562, color=lad_K562_n)) +
    geom_point(data=P[P$LAD_K562==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_K562==1, ], size=0.5, alpha=0.2) +
    ggtitle('CAGE vs SuRE K562') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=x.mean, y=value, color=variable), size=1) +
    labs(y='log10(CAGE)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_K562_n)

ggplot(P, aes(x=SuRE_K562, y=DNAse_K562, color=lad_K562_n)) +
    geom_point(data=P[P$LAD_K562==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_K562==1, ], size=0.5, alpha=0.2) +
    ggtitle('DNAse vs SuRE K562') +
    theme_bw() +
    labs(y='log10(DNAse)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_K562_n)

ggplot(P, aes(x=CAGE_K562, y=DNAse_K562, color=lad_K562_n)) +
    geom_point(data=P[P$LAD_K562==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_K562==1, ], size=0.5, alpha=0.2) +
    ggtitle('CAGE vs DNAse K562') +
    theme_bw() +
    labs(y='log10(DNAse)', x='log10(CAGE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_K562_n)




RM_melt = melt(RM_HepG2, measure.vars=c('y.ilad', 'y.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='y.lad', lad_names_HepG2['LAD'], lad_names_HepG2['iLAD'])
ggplot(P, aes(x=SuRE_HepG2, y=CAGE_HepG2, color=lad_HepG2_n)) +
    geom_point(data=P[P$LAD_HepG2==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_HepG2==1, ], size=0.5, alpha=0.2) +
    ggtitle('CAGE vs SuRE HepG2') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=x.mean, y=value, color=variable), size=1) +
    labs(y='log10(CAGE)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_HepG2_n)

ggplot(P, aes(x=SuRE_HepG2, y=DNAse_HepG2, color=lad_HepG2_n)) +
    geom_point(data=P[P$LAD_HepG2==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_HepG2==1, ], size=0.5, alpha=0.2) +
    ggtitle('DNAse vs SuRE HepG2') +
    theme_bw() +
    labs(y='log10(DNAse)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_HepG2_n)

ggplot(P, aes(x=CAGE_HepG2, y=DNAse_HepG2, color=lad_HepG2_n)) +
    geom_point(data=P[P$LAD_HepG2==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P[P$LAD_HepG2==1, ], size=0.5, alpha=0.2) +
    ggtitle('CAGE vs DNAse HepG2') +
    theme_bw() +
    labs(y='log10(DNAse)', x='log10(CAGE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_HepG2_n)

```


**conclusions:**

Although the spread of the data is different between cell-types, this seems to have worked quite nicely.

```{r}

classify <- function(sure, exp, lrs, lad, exp_cut){
    INACT<- sure< -0.3 & lad & exp< exp_cut #inactive
    NREP<- sure> 0 & lrs > -0.5 & lad & exp> exp_cut #not repressed
    REP<- sure> 0.3 & lrs < -1 & lad  & exp< exp_cut #repressed
    Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
    names(Pcnts)<-c("repressed", "escaper", "inactive")
    BND <- lad & !INACT & !NREP & !REP
    class = rep(NA, length(sure))
    class[lad==0] = 'iLAD'
    class[INACT]<-"inactive"
    class[NREP]<-"escaper"
    class[REP]<-"repressed"
    class[BND] <- "boundary"
    return(factor(class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary')))
}

P$class_GROcap = classify(P$SuRE_K562, P$GROcap_K562, P$LRS_GROcap, P$LAD_K562, -2)
P$class_GROcap_seq = classify(P$SuRE_K562, P$GROcap_K562, P$LRS_GROcap_seq, P$LAD_K562_seq, -2)

length(which(P$class_GROcap=='escaper'&P$class_GROcap_seq=='escaper'))


class_names = paste0(levels(P$class_GROcap), '; n=',table(P$class_GROcap))
names(class_names) = levels(P$class_GROcap)
P$class_GROcap_n = P$class_GROcap
levels(P$class_GROcap_n) = class_names
COL_class_GROcap_n = COL_class[names(class_names)]
names(COL_class_GROcap_n) = class_names

p_classes = P[which(P$class_GROcap %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P, aes(x=SuRE_K562, y=GROcap_K562)) +
    geom_point(size=0.1,color=COLi) +
    geom_point(data=p_classes, aes(color=class_GROcap_n), size=0.6) +
    labs(y='log10(GROcap)', x='log10(SuRE)') +
    theme_bw() +
    ggtitle('GROcap vs SuRE K562') +
    geom_line(data=RM_GRO, aes(x=x.mean, y=y.ilad), color=COL_lad['iLAD']) +
    theme(legend.title=element_blank()) +
    scale_colour_manual(values=COL_class_GROcap_n)

```

## classify with 3 dimensions

```{r}
classify_3d <- function(sure, exp_y, exp_z, lrs_y, lad, exp_cut_y, exp_cut_z){
    INACT<- sure< -0.3 & lad & exp_y < exp_cut_y  & exp_z < exp_cut_z #inactive
    NREP<- sure> 0 & lrs_y > -0.5 & lad & exp_y > exp_cut_y  & exp_z > exp_cut_z #not repressed
    REP<- sure> 0.3 & lrs_y < -0.75  & lad  & exp_y < exp_cut_y  & exp_z < exp_cut_z #repressed
    Pcnts<-c(length(which(REP)), length(which(NREP)), length(which(INACT)))
    names(Pcnts)<-c("repressed", "escaper", "inactive")
    BND <- lad & !INACT & !NREP & !REP
    class = rep(NA, length(sure))
    class[lad==0] = 'iLAD'
    class[INACT]<-"inactive"
    class[NREP]<-"escaper"
    class[REP]<-"repressed"
    class[BND] <- "boundary"
    return(factor(class, levels=c('iLAD', 'escaper', 'repressed', 'inactive', 'boundary')))
}


P$class_K562_3D = classify_3d(P$SuRE_K562,
                              P$CAGE_K562, P$DNAse_K562,
                              P$LRS_K562, P$LAD_K562, 1.25, 0.75)

P$class_HT1080_3D = classify_3d(P$SuRE_HT1080,
                                P$CAGE_HT1080, P$DNAse_HT1080,
                                P$LRS_HT1080, P$LAD_HT1080, 1.25, 1.75)

P$class_HepG2_3D = classify_3d(P$SuRE_HepG2,
                                P$CAGE_HepG2, P$DNAse_HepG2,
                                P$LRS_HepG2, P$LAD_HepG2, 2.5, 1.75)

p_class = P[P$class_HepG2_3D%in%c('escaper', 'repressed', 'inactive'), ]
class_gr = gene_gr[gene_gr$name %in% p_class$name]


```

```{r}
p_classes = P[which(P$class_HepG2_3D %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P, aes(x=SuRE_HepG2, y=CAGE_HepG2)) +
    geom_point(size=0.1,color=COLi) +
    geom_point(data=p_classes, aes(color=class_HepG2_3D), size=0.6) +
    labs(y='log10(CAGE)', x='log10(SuRE)') +
    theme_bw() +
    ggtitle('CAGE vs SuRE K562') +
    geom_line(data=RM_HepG2, aes(x=x.mean, y=y.ilad), color=COL_lad['iLAD']) +
    theme(legend.title=element_blank()) +
    scale_colour_manual(values=COL_class)
```

```{r}

lmnb1_count_K562 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170821_GCF4577_K562_p13_LMNB1/results/counts/LMNB1-gatc.counts.txt.gz')
dam_count_K562 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170821_GCF4577_K562_p13_LMNB1/results/counts/Dam16-gatc.counts.txt.gz')



lmnb1_count_rep1 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_1_LMNB1-gatc.counts.txt.gz')


lmnb1_count_rep2 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_2_LMNB1-gatc.counts.txt.gz')

dam10_count_rep1 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_1_Dam10-gatc.counts.txt.gz')

dam10_count_rep2 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_2_Dam10-gatc.counts.txt.gz')

dam1_count_rep1 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_1_Dam1-gatc.counts.txt.gz')
dam1_count_rep2 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_2_Dam1-gatc.counts.txt.gz')

dam100_count_rep1 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_1_Dam100-gatc.counts.txt.gz')
dam100_count_rep2 = read.table('/home/t.v.schaik/mydata/proj/3D_nucleus/results/ts170828_GCF4579_2xHEPG2_RPE_HFF/results/counts/HEPG2_2_Dam100-gatc.counts.txt.gz')

colnames(lmnb1_count_K562) = colnames(dam_count_K562) =
    colnames(lmnb1_count_rep1) = colnames(lmnb1_count_rep2) =
    colnames(dam1_count_rep1) = colnames(dam1_count_rep2) =
    colnames(dam10_count_rep1) = colnames(dam10_count_rep2) =
    colnames(dam100_count_rep1) = colnames(dam100_count_rep2) =
    c('seqnames', 'start', 'end', 'count')



lmnb1_count_HepG2 = dam1_count_HepG2 =
    dam10_count_HepG2 = dam100_count_HepG2 =
    lmnb1_count_rep1[,c('seqnames', 'start', 'end')]

lmnb1_count_HepG2$count = rowSums(cbind(lmnb1_count_rep1[, 'count'],
                                        lmnb1_count_rep2[, 'count']))

dam1_count_HepG2$count = rowSums(cbind(dam1_count_rep1[, 'count'],
                                      dam1_count_rep2[, 'count']))
dam10_count_HepG2$count = rowSums(cbind(dam10_count_rep1[, 'count'],
                                        dam10_count_rep2[, 'count']))
dam100_count_HepG2$count = rowSums(cbind(dam100_count_rep1[, 'count'],
                                        dam100_count_rep2[, 'count']))


lmnb1_gr = makeGRangesFromDataFrame(lmnb1_count_K562)

lad_overlap = findOverlaps(LAD_K562[LAD_K562$name=='LAD'], lmnb1_gr)
ilad_overlap = findOverlaps(LAD_K562[LAD_K562$name=='interLAD'], lmnb1_gr)

lad_log2_K562 = log2(sum(lmnb1_count_K562[to(lad_overlap),'count'])/
                     sum(dam_count_K562[to(lad_overlap),'count']))
ilad_log2_K562 = log2(sum(lmnb1_count_K562[to(ilad_overlap),'count'])/
                      sum(dam_count_K562[to(ilad_overlap),'count']))

lad_overlap = findOverlaps(LAD_HepG2, lmnb1_gr)


lad_log2_dam1_HepG2 = log2(sum(lmnb1_count_HepG2[to(lad_overlap),'count'])/
                           sum(dam1_count_HepG2[to(lad_overlap),'count']))
ilad_log2_dam1_HepG2 = log2(sum(lmnb1_count_HepG2[-to(lad_overlap),'count'])/
                            sum(dam1_count_HepG2[-to(lad_overlap),'count']))

lad_log2_dam10_HepG2 = log2(sum(lmnb1_count_HepG2[to(lad_overlap),'count'])/
                            sum(dam10_count_HepG2[to(lad_overlap),'count']))
ilad_log2_dam10_HepG2 = log2(sum(lmnb1_count_HepG2[-to(lad_overlap),'count'])/
                             sum(dam10_count_HepG2[-to(lad_overlap),'count']))

lad_log2_dam100_HepG2 = log2(sum(lmnb1_count_HepG2[to(lad_overlap),'count'])/
                         sum(dam100_count_HepG2[to(lad_overlap),'count']))
ilad_log2_dam100_HepG2 = log2(sum(lmnb1_count_HepG2[-to(lad_overlap),'count'])/
                          sum(dam100_count_HepG2[-to(lad_overlap),'count']))


h<-findOverlaps(tss_gr, lmnb1_gr, maxgap=22000)

oENST<-tss_gr[from(h)]$name
oPOS<-ifelse(strand(tss_gr[from(h)])=='+',
            (start(lmnb1_gr[to(h)])+end(lmnb1_gr[to(h)]))/2 - start(tss_gr[from(h)]),
            end(tss_gr[from(h)]) - (start(lmnb1_gr[to(h)])+end(lmnb1_gr[to(h)]))/2)
        #coordinates of all overlapping probes relative to the gene starts


get_run_log2 <- function(class, class_col, P, oENST, oPOS, lmnb1_vec, dam_vec){
   s<-unique(P$name[which(P[, class_col]==class) ])
   s<-s[!is.na(s)]
   w<-oENST %in% s #which rows in oENST correspond to genes in s
   subPOS<-oPOS[w]
   subLMNB1 <-lmnb1_vec[w]
   subDam <-dam_vec[w]
   o<-order(subPOS) #need to order all values by distance to gene start for running median plot:
   subPOS<-subPOS[o]
   subLMNB1<-subLMNB1[o]
   subDam<-subDam[o]
   #determine runmed k:
   wsize<-floor(length(subPOS)/20) #2% of all datapoints in the set
   if(!wsize %% 2) {wsize<-wsize+1} #must be odd
   #plot:
   run_log2 = log2(runsum(Rle(subLMNB1), k=wsize, endrule='constant')/
                   runsum(Rle(subDam), k=wsize, endrule='constanraw_datat'))
   return(data.frame(log2=as.numeric(run_log2), pos=subPOS, class=class))
}

lmnb1_vec <- lmnb1_count_K562[to(h), 'count']
dam_vec <- dam_count_K562[to(h), 'count']

log2_list = lapply(names(COL), get_run_log2, 'class_GROcap', P, oENST, oPOS,
                   lmnb1_vec, dam_vec)

log2_data_K562 = do.call(rbind.data.frame, log2_list)



log2_list_seq = lapply(names(COL), get_run_log2, 'class_GROcap_seq', P, oENST, oPOS,
                   lmnb1_vec, dam_vec)

log2_data_K562_seq = do.call(rbind.data.frame, log2_list_seq)



lmnb1_vec <- lmnb1_count_HepG2[to(h), 'count']
dam1_vec <- dam1_count_HepG2[to(h), 'count']
dam10_vec <- dam10_count_HepG2[to(h), 'count']
dam100_vec <- dam100_count_HepG2[to(h), 'count']

log2_list = lapply(names(COL), get_run_log2, 'class_HepG2_3D', P, oENST, oPOS,
                   lmnb1_vec, dam1_vec)

log2_data_HepG2_dam1 = do.call(rbind.data.frame, log2_list)


log2_list = lapply(names(COL), get_run_log2, 'class_HepG2_3D', P, oENST, oPOS,
                   lmnb1_vec, dam10_vec)

log2_data_HepG2_dam10 = do.call(rbind.data.frame, log2_list)

log2_list = lapply(names(COL), get_run_log2, 'class_HepG2_3D', P, oENST, oPOS,
                   lmnb1_vec, dam100_vec)

log2_data_HepG2_dam100 = do.call(rbind.data.frame, log2_list)


p1 = ggplot(log2_data_K562, aes(x=pos, y=log2, color=class)) +
    geom_line() +
    ggtitle('lamin dip using HMM from microarray for LAD definition') +
    scale_color_manual(values=COL) +
    ylim(-3,3.5) +
    geom_hline(yintercept = lad_log2_K562, color=COL_lad['LAD']) +
    geom_hline(yintercept = ilad_log2_K562, color=COL_lad['iLAD']) +
    geom_hline(yintercept = 0, color='black') +
    geom_vline(xintercept = 0, linetype='dotdash')

p2 = ggplot(log2_data_K562_seq, aes(x=pos, y=log2, color=class)) +
    geom_line() +
    ggtitle('lamin dip using HMM from 10kb sequencing bins for LAD definition') +
    scale_color_manual(values=COL) +
    ylim(-3,3.5) +
    geom_hline(yintercept = lad_log2_K562, color=COL_lad['LAD']) +
    geom_hline(yintercept = ilad_log2_K562, color=COL_lad['iLAD']) +
    geom_hline(yintercept = 0, color='black') +
    geom_vline(xintercept = 0, linetype='dotdash')

grid.arrange(p1,p2, nrow=1)


p1 = ggplot(log2_data_HepG2_dam1, aes(x=pos, y=log2, color=class)) +
    geom_line() +
    scale_color_manual(values=COL) +
    ylim(-3,3.5) +
    geom_hline(yintercept = lad_log2_dam1_HepG2, color=COL_lad['LAD']) +
    geom_hline(yintercept = ilad_log2_dam1_HepG2, color=COL_lad['iLAD']) +
    geom_hline(yintercept = 0, color='black') +
    geom_vline(xintercept = 0, linetype='dotdash')
p2 = ggplot(log2_data_HepG2_dam10, aes(x=pos, y=log2, color=class)) +
    geom_line() +
    scale_color_manual(values=COL) +
    ylim(-3,3.5) +
    geom_hline(yintercept = lad_log2_dam10_HepG2, color=COL_lad['LAD']) +
    geom_hline(yintercept = ilad_log2_dam10_HepG2, color=COL_lad['iLAD']) +
    geom_hline(yintercept = 0, color='black') +
    geom_vline(xintercept = 0, linetype='dotdash')
p3 = ggplot(log2_data_HepG2_dam100, aes(x=pos, y=log2, color=class)) +
    geom_line() +
    scale_color_manual(values=COL) +
    ylim(-3,3.5) +
    geom_hline(yintercept = lad_log2_dam100_HepG2, color=COL_lad['LAD']) +
    geom_hline(yintercept = ilad_log2_dam100_HepG2, color=COL_lad['iLAD']) +
    geom_hline(yintercept = 0, color='black') +
    geom_vline(xintercept = 0, linetype='dotdash')

grid.arrange(p1,p2,p3, nrow=1)

full_lamin = data.frame(pos=oPOS,
                        log2=log2((lmnb1_vec+1) /(dam_vec+1)),
                        class=P[oENST, 'class_HepG2_3D'])

ggplot(full_lamin[full_lamin$class %in% c('escaper', 'repressed', 'inactive'), ],
       aes(x=pos, y=log2, color=class)) +
    geom_point(position=position_jitter(height=0.1), alpha=0.4, size=0.3) +
    scale_color_manual(values=COL) +
    geom_hline(yintercept = lad_log2_HepG2, color=COL_lad['LAD']) +
    geom_hline(yintercept = ilad_log2_HepG2, color=COL_lad['iLAD']) +
    geom_hline(yintercept = 0, color='black') +
    geom_vline(xintercept = 0, linetype='dotdash') +
    facet_wrap(~class)


get_run_mean <- function(class, class_col, P, oENST, oPOS, lamin_vec){
   s<-unique(P$name[which(P[, class_col]==class) ])
   s<-s[!is.na(s)]
   w<-oENST %in% s #which rows in oENST correspond to genes in s
   subPOS<-oPOS[w]
   subLam <-lamin_vec[w]
   o<-order(subPOS) #need to order all values by distance to gene start for running median plot:
   subPOS<-subPOS[o]
   subLam<-subLam[o]
   #determine runmed k:
   wsize<-floor(length(subPOS)/20) #2% of all datapoints in the set
   if(!wsize %% 2) {wsize<-wsize+1} #must be odd
   #plot:
   run_sum = runsum(Rle(subLam), k=wsize, endrule='constant')
   return(data.frame(sum=as.numeric(run_sum/length(s)), pos=subPOS, class=class))
}


lmnb1_K562_list = lapply(names(COL), get_run_mean, 'class_GROcap', P, oENST, oPOS,
                        lmnb1_count_K562[to(h), 'count'])

lmnb1_data_K562 = do.call(rbind.data.frame, lmnb1_K562_list)

dam_K562_list = lapply(names(COL), get_run_mean, 'class_GROcap', P, oENST, oPOS,
                        dam_count_K562[to(h), 'count'])

dam_data_K562 = do.call(rbind.data.frame, dam_K562_list)


p1 = ggplot(lmnb1_data_K562, aes(x=pos, y=log2(sum), color=class)) +
        geom_line() +
        ggtitle('LMNB1 running mean K562') +
        scale_color_manual(values=COL) +
        geom_hline(yintercept = 0, color='black') +
        geom_vline(xintercept = 0, linetype='dotdash')

p2 = ggplot(dam_data_K562, aes(x=pos, y=log2(sum), color=class)) +
        geom_line() +
        ggtitle('DAM running mean K562') +
        scale_color_manual(values=COL) +
        geom_hline(yintercept = 0, color='black') +
        geom_vline(xintercept = 0, linetype='dotdash')



lmnb1_rep1_list = lapply(names(COL), get_run_mean, 'class_HepG2_3D', P, oENST, oPOS,
                        lmnb1_count_rep1[to(h), 'count'])

lmnb1_data_rep1 = do.call(rbind.data.frame, lmnb1_rep1_list)

lmnb1_rep2_list = lapply(names(COL), get_run_mean, 'class_HepG2_3D', P, oENST, oPOS,
                        lmnb1_count_rep2[to(h), 'count'])

lmnb1_data_rep2 = do.call(rbind.data.frame, lmnb1_rep2_list)

dam_rep1_list = lapply(names(COL), get_run_mean, 'class_HepG2_3D', P, oENST, oPOS,
                        dam_count_rep1[to(h), 'count'])

dam_data_rep1 = do.call(rbind.data.frame, dam_rep1_list)

dam_rep2_list = lapply(names(COL), get_run_mean, 'class_HepG2_3D', P, oENST, oPOS,
                        dam_count_rep2[to(h), 'count'])

dam_data_rep2 = do.call(rbind.data.frame, dam_rep2_list)


lad_lmnb1 = mean(lmnb1_count_HepG2[to(lad_overlap),'count'])
ilad_lmnb1 = mean(lmnb1_count_HepG2[-to(lad_overlap),'count'])

lad_dam = mean(dam_count_HepG2[to(lad_overlap),'count'])
ilad_dam = mean(dam_count_HepG2[-to(lad_overlap),'count'])

p1 = ggplot(lmnb1_data_rep1, aes(x=pos, y=log2(sum), color=class)) +
        geom_line() +
        ggtitle('LMNB1 running sum rep1') +
        scale_color_manual(values=COL) +
        geom_hline(yintercept = 0, color='black') +
        geom_vline(xintercept = 0, linetype='dotdash')

p2 = ggplot(lmnb1_data_rep2, aes(x=pos, y=log2(sum), color=class)) +
        geom_line() +
        ggtitle('LMNB1 running sum rep2') +
        scale_color_manual(values=COL) +
        geom_hline(yintercept = 0, color='black') +
        geom_vline(xintercept = 0, linetype='dotdash')

p3 = ggplot(dam_data_rep1, aes(x=pos, y=log2(sum), color=class)) +
        geom_line() +
        ggtitle('DAM-only running sum rep1') +
        scale_color_manual(values=COL) +
        geom_hline(yintercept = 0, color='black') +
        geom_vline(xintercept = 0, linetype='dotdash')
p4 = ggplot(dam_data_rep2, aes(x=pos, y=log2(sum), color=class)) +
        geom_line() +
        ggtitle('DAM-only running sum rep2') +
        scale_color_manual(values=COL) +
        geom_hline(yintercept = 0, color='black') +
        geom_vline(xintercept = 0, linetype='dotdash')
grid.arrange(p1,p2,p3,p4, nrow=1)
```

```
bwtool summary -header \
               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
               <(awk -vOFS='\t' '{
                   if ($6=="+"){
                       start = $2 - 50
                       end = $3 + 300
                   } else {
                       start = $2-300
                       end = $3 + 50
                   }
                   start = start > 0 ? start : 0 ;
                   end = end > 0 ? end : 0 ;
                   print $1, start, end, $4, 0, $6}' raw_data/tss_hg38.bed) \
               ~/mydata/data/tracks/hg38/POL2A_HepG2_hg38_ENCFF378NKG.bigWig \
               /dev/stdout | gzip -c > raw_data/tssr_POL2A_HepG2.txt.gz &


bwtool summary -header \
               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
               <(awk -vOFS='\t' '{
                   if ($6=="+"){
                       start = $2 + 300
                       end = $3 + 3000
                   } else {
                       start = $2 - 3000
                       end = $3 - 300
                   }
                   start = start > 0 ? start : 0 ;
                   end = end > 0 ? end : 0 ;
                   print $1, start, end, $4, 0, $6}' raw_data/tss_hg38.bed) \
               ~/mydata/data/tracks/hg38/POL2A_HepG2_hg38_ENCFF378NKG.bigWig \
               /dev/stdout | gzip -c > raw_data/body_POL2A_HepG2.txt.gz &


bwtool summary -header \
#               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
              <(awk -vOFS='\t' '{
                  if ($6=="+"){
                      start = $2 - 50
                      end = $3 + 300
                  } else {
                      start = $2-300
                      end = $3 + 50
                  }
                  start = start > 0 ? start : 0 ;
                  end = end > 0 ? end : 0 ;
                  print $1, start, end, $4, 0, $6}' raw_data/tss_hg38.bed) \
              ~/mydata/data/tracks/hg38/POL2A_K562_ENCFF042CRO.bigWig \
              /dev/stdout | gzip -c > raw_data/tssr_POL2A_K562.txt.gz &


bwtool summary -header \
#               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
              <(awk -vOFS='\t' '{
                  if ($6=="+"){
                      start = $2 + 300
                      end = $3 + 3000
                  } else {
                      start = $2 - 3000
                      end = $3 - 300
                  }
                  start = start > 0 ? start : 0 ;
                  end = end > 0 ? end : 0 ;
                  print $1, start, end, $4, 0, $6}' raw_data/tss_hg38.bed) \
              ~/mydata/data/tracks/hg38/POL2A_K562_ENCFF042CRO.bigWig \
              /dev/stdout | gzip -c > raw_data/body_POL2A_K562.txt.gz &

```



```{r}
pol2_tssr = read.table('../raw_data/tssr_POL2A_K562.txt.gz', row.names=4,
                       col.names=c('chrom', 'start', 'end', 'name', 'score',
                                   'strand', 'size', 'num_data', 'min', 'max',
                                   'mean', 'sum'), stringsAsFactors=F)

pol2_body = read.table('../raw_data/body_POL2A_K562.txt.gz', row.names=4, skip=1,
                       col.names=c('chrom', 'start', 'end', 'name', 'score',
                                  'strand', 'size', 'num_data', 'min', 'max',
                                  'mean', 'sum'), stringsAsFactors=F)
pol2_tssr$mean = pol2_tssr$sum / pol2_tssr$num_data
pol2_body$mean = pol2_body$sum / pol2_body$num_data

P$tssr_POL2A_K562 = pol2_tssr[P$name, 'mean']
P$body_POL2A_K562 = pol2_body[P$name, 'mean']

p_matched = matchSet(P[P$class_K562_3D%in%c('escaper', 'iLAD'), ],
                     'class_K562_3D', 'escaper', 'CAGE_K562')

ggplot(p_matched, aes(x=log10(tssr_POL2A_K562 + body_POL2A_K562),
                      y=log10(tssr_POL2A_K562 / body_POL2A_K562),
                      color=class_K562_3D)) +
    geom_point(alpha=0.5, size=0.5) +
    ggtitle('POL2 sum of mean signal vs TSS POL2 / gene body POL2\nK562 Classification based on CAGE and DNAse\niLAD promoters matched on CAGE') +
    geom_smooth(method='lm') +
    scale_color_manual(values=COL_class)

p_pol2 = melt(p_matched[,c('class_K562_3D', 'tssr_POL2A_K562',
                           'body_POL2A_K562')])
p_pol2$variable = factor(p_pol2$variable)
levels(p_pol2) = c("TSS region (-50bp:+300bp from TSS)",
                   "Gene body (+300bp:3000bp from TSS)")
ggplot(p_pol2, aes(x=class_K562_3D, y=log10(value), color=class_K562_3D)) +
    geom_violin() +
    ggtitle('POL2 mean signal on TSS and gene body\nK562 Classification based on CAGE and DNAse\niLAD promoters matched on CAGE') +
    geom_point(position=position_jitter(width=0.1), alpha=0.5, size=0.5) +
    scale_color_manual(values=COL_class) +
    xlab('promoter class') +
    facet_wrap('variable')

p_matched = matchSet(P[P$class_GROcap%in%c('escaper', 'iLAD'), ],
                     'class_GROcap', 'escaper', 'GROcap_K562')

ggplot(p_matched, aes(x=log2(tssr_POL2A_K562 + body_POL2A_K562),
                      y=log10(tssr_POL2A_K562 / body_POL2A_K562),
                      color=class_GROcap)) +
    geom_point(alpha=0.5, size=0.5) +
    ggtitle('POL2 sum of mean signal vs TSS POL2 / gene body POL2\nK562 Classification based on GROcap\niLAD promoters matched on GROcap') +
    geom_smooth(method='lm') +
    scale_color_manual(values=COL_class)

p_pol2 = melt(p_matched[,c('class_GROcap', 'tssr_POL2A_K562',
                           'body_POL2A_K562')])
p_pol2$variable = factor(p_pol2$variable)
levels(p_pol2) = c("TSS region (-50bp:+300bp from TSS)",
                   "Gene body (+300bp:3000bp from TSS)")
ggplot(p_pol2, aes(x=class_GROcap, y=log10(value), color=class_GROcap)) +
    geom_violin() +
    geom_point(position=position_jitter(width=0.1), alpha=0.5, size=0.5) +
    scale_color_manual(values=COL_class) +
    ggtitle('POL2 mean signal on TSS and gene body\nK562 Classification based on GROcap\niLAD promoters matched on GROcap') +
    xlab('promoter class') +
    facet_wrap('variable')



pol2_tssr = read.table('../raw_data/tssr_POL2A_HepG2.txt.gz', row.names=4,
                       col.names=c('chrom', 'start', 'end', 'name', 'score',
                                   'strand', 'size', 'num_data', 'min', 'max',
                                   'mean', 'sum'), stringsAsFactors=F)

pol2_body = read.table('../raw_data/body_POL2A_HepG2.txt.gz', row.names=4,
                       col.names=c('chrom', 'start', 'end', 'name', 'score',
                                  'strand', 'size', 'num_data', 'min', 'max',
                                  'mean', 'sum'), stringsAsFactors=F)
pol2_tssr$mean = pol2_tssr$sum / pol2_tssr$num_data
pol2_body$mean = pol2_body$sum / pol2_body$num_data
P$tssr_POL2A_HepG2 = pol2_tssr[P$name, 'mean']
P$body_POL2A_HepG2 = pol2_body[P$name, 'mean']

p_matched = matchSet(P[P$class_HepG2_3D%in%c('escaper', 'iLAD'), ],
                     'class_HepG2_3D', 'escaper', 'CAGE_HepG2')

ggplot(p_matched, aes(x=log2(tssr_POL2A_HepG2 + body_POL2A_HepG2),
                      y=log10(tssr_POL2A_HepG2 / body_POL2A_HepG2),
                      color=class_HepG2_3D)) +
    geom_point(alpha=0.5, size=0.5) +
    ggtitle('POL2 sum of mean signal vs TSS POL2 / gene body POL2\nHepG2 Classification based on CAGE and DNAse\niLAD promoters matched on CAGE') +
    geom_smooth(method='lm') +
    scale_color_manual(values=COL_class)

p_pol2 = melt(p_matched[,c('class_HepG2_3D', 'tssr_POL2A_HepG2',
                           'body_POL2A_HepG2')])
p_pol2$variable = factor(p_pol2$variable)
levels(p_pol2) = c("TSS region (-50bp:+300bp from TSS)",
                   "Gene body (+300bp:3000bp from TSS)")
ggplot(p_pol2, aes(x=class_HepG2_3D, y=log10(value), color=class_HepG2_3D)) +
    geom_violin() +
    ggtitle('POL2 mean signal on TSS and gene body\nHepG2 Classification based on CAGE and DNAse\niLAD promoters matched on CAGE') +
    geom_point(position=position_jitter(width=0.1), alpha=0.5, size=0.5) +
    scale_color_manual(values=COL_class) +
    xlab('promoter class') +
    facet_wrap('variable')


p_matched = matchSet(P[P$class_HepG2_3D%in%c('escaper', 'iLAD'), ],
                     'class_HepG2_3D', 'escaper', 'tssr_POL2A_HepG2')

ggplot(p_matched, aes(x=log2(tssr_POL2A_HepG2 + body_POL2A_HepG2),
                      y=log10(tssr_POL2A_HepG2 / body_POL2A_HepG2),
                      color=class_HepG2_3D)) +
    ggtitle('POL2 sum of mean signal vs TSS POL2 / gene body POL2\nHepG2 Classification based on CAGE and DNAse\niLAD promoters matched on tss POL2') +
    geom_point(alpha=0.5, size=0.5) +
    geom_smooth(method='lm') +
    scale_color_manual(values=COL_class)

p_pol2 = melt(p_matched[,c('class_HepG2_3D', 'tssr_POL2A_HepG2',
                           'body_POL2A_HepG2')])
p_pol2$variable = factor(p_pol2$variable)
levels(p_pol2) = c("TSS region (-50bp:+300bp from TSS)",
                   "Gene body (+300bp:3000bp from TSS)")
ggplot(p_pol2, aes(x=class_HepG2_3D, y=log10(value), color=class_HepG2_3D)) +
    geom_violin() +
    ggtitle('POL2 mean signal on TSS and gene body\nHepG2 Classification based on CAGE and DNAse\niLAD promoters matched on tss POL2') +
    geom_point(position=position_jitter(width=0.1), alpha=0.5, size=0.5) +
    scale_color_manual(values=COL_class) +
    xlab('promoter class') +
    facet_wrap('variable')
```

**conclusions:**
This new way of analyses where I use the bigwig track instead of the bam files
shows a bit less clear differences between body Pol2 compared to tssr pol2 for
escapers and iLADs.


```
nice -19 \
bwtool summary -header \
               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
               raw_data/enhancer_regions.bed \
               /DATA/usr/joris/git_data/SuRE_bigwig/SuRE23_HEPG2/sure23.hepg2.plasmid.norm.HEPG2_norm.minus.170707.bw \
               /dev/stdout | gzip -c > raw_data/enhancer_regions_hepg2_sure_minus.txt.gz &
nice -19 \
bwtool summary -header \
               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
               raw_data/enhancer_regions.bed \
               /DATA/usr/joris/git_data/SuRE_bigwig/SuRE23_HEPG2/sure23.hepg2.plasmid.norm.HEPG2_norm.plus.170707.bw \
               /dev/stdout | gzip -c > raw_data/enhancer_regions_hepg2_sure_plus.txt.gz &

bedtools coverage -c -a raw_data/enhancer_regions.bed -b ~c.leemans/mydata/data/tracks/hg19/HepG2_ENCODE_biol_rep1.CNhs12328.10818-111B8.hg19.nobarcode.bam | gzip -c > raw_data/enhancer_CAGE_HepG2_rep1.txt.gz &

bedtools coverage -c -a raw_data/enhancer_regions.bed -b ~c.leemans/mydata/data/tracks/hg19/HepG2_ENCODE_biol_rep2.CNhs12329.10819-111B9.hg19.nobarcode.bam | gzip -c > raw_data/enhancer_CAGE_HepG2_rep2.txt.gz &

bedtools coverage -c -s -a raw_data/enhancer_regions.bed -b ~c.leemans/mydata/data/tracks/hg19/HepG2_ENCODE_biol_rep3.CNhs12330.10820-111C1.hg19.nobarcode.bam | gzip -c > raw_data/enhancer_CAGE_HepG2_rep3.txt.gz &



bwtool summary -header \
               -fill=0 \
               -with-sum \
               -skip-median \
               -keep-bed \
               <(awk -vOFS='\t' '{print $1, $2 - 500, $3 + 500, $4, 0, "."}' raw_data/enhancer_regions.bed)  /home/c.leemans/mydata/data/tracks/hg19/DNase-seq/GSM2400287_ENCFF646UXG_signal_of_unique_reads_hg19.bigWig \
               /dev/stdout | gzip -c > raw_data/enhancer_DNAse_HepG2.txt.gz &

```

```{r}
sure_plus_K562 = read.table('../raw_data/enhancer_regions_sure_plus.txt.gz',
                       col.names=c('chrom', 'start', 'end', 'name',
                                   'size', 'num_data', 'min', 'max',
                                   'mean', 'sum'),
                       row.names=4, stringsAsFactors=F)
sure_minus_K562 = read.table('../raw_data/enhancer_regions_sure_minus.txt.gz',
                        col.names=c('chrom', 'start', 'end', 'name',
                                    'size', 'num_data', 'min', 'max',
                                    'mean', 'sum'),
                        row.names=4, stringsAsFactors=F)

grocap_plus_K562 = read.table('../raw_data/enhancer_regions_grocap_plus.txt.gz',
                              col.names=c('chrom', 'start', 'end', 'name',
                                          'size', 'num_data', 'min', 'max',
                                          'mean', 'sum'),
                              row.names=4, stringsAsFactors=F)
grocap_minus_K562 = read.table('../raw_data/enhancer_regions_grocap_minus.txt.gz',
                               col.names=c('chrom', 'start', 'end', 'name',
                                           'size', 'num_data', 'min', 'max',
                                           'mean', 'sum'),
                               row.names=4, stringsAsFactors=F)


sure_plus_hepg2 = read.table('../raw_data/enhancer_regions_hepg2_sure_plus.txt.gz',
                       col.names=c('chrom', 'start', 'end', 'name',
                                   'size', 'num_data', 'min', 'max',
                                   'mean', 'sum'),
                       row.names=4, stringsAsFactors=F)
sure_minus_hepg2 = read.table('../raw_data/enhancer_regions_hepg2_sure_minus.txt.gz',
                        col.names=c('chrom', 'start', 'end', 'name',
                                    'size', 'num_data', 'min', 'max',
                                    'mean', 'sum'),
                        row.names=4, stringsAsFactors=F)

cage_rep1_hepg2 = read.table('../raw_data/enhancer_CAGE_HepG2_rep1.txt.gz',
                       col.names=c('chrom', 'start', 'end', 'name', 'mean'),
                       row.names=4, stringsAsFactors=F)
cage_rep2_hepg2 = read.table('../raw_data/enhancer_CAGE_HepG2_rep2.txt.gz',
                       col.names=c('chrom', 'start', 'end', 'name', 'mean'),
                       row.names=4, stringsAsFactors=F)
cage_rep3_hepg2 = read.table('../raw_data/enhancer_CAGE_HepG2_rep3.txt.gz',
                       col.names=c('chrom', 'start', 'end', 'name', 'mean'),
                       row.names=4, stringsAsFactors=F)

dnase_hepg2 = read.table('../raw_data/enhancer_DNAse_HepG2.txt.gz',
                   col.names=c('chrom', 'start', 'end', 'name', 'score',
                               'strand', 'size', 'num_data', 'min', 'max',
                               'mean', 'sum'),
                   row.names=4, stringsAsFactors=F)



P_enh = data.frame(SuRE_K562=rowMeans(cbind(sure_plus_K562$mean,
                                            sure_minus_K562$mean)),
                   GROcap_K562=rowMeans(cbind(grocap_plus_K562$mean,
                                              abs(grocap_minus_K562$mean))),
                   SuRE_HepG2=rowMeans(cbind(sure_plus_hepg2$mean,
                                             sure_minus_hepg2$mean)),
                   CAGE_HepG2=rowMeans(cbind(cage_rep1_hepg2$mean,
                                             cage_rep2_hepg2$mean,
                                             cage_rep3_hepg2$mean)),
                   DNAse_HepG2=dnase_hepg2$mean,
                   row.names=rownames(sure_plus))

for (name in colnames(P_enh)){
    P_enh[,name] = pseudo_log10(P_enh[,name])
}
chr_y = grep('chrY', rownames(P_enh))
P_enh = P_enh[-chr_y,]
enh_gr = makeGRangesFromDataFrame(sure_plus[-chr_y, ])

o = findOverlaps(enh_gr, LAD_K562)
P_enh$LAD_K562 = 0
P_enh$LAD_K562[queryHits(o)] = 1


o = findOverlaps(enh_gr, LAD_HepG2)
P_enh$LAD_HepG2 = 0
P_enh$LAD_HepG2[queryHits(o)] = 1

RM_enh_K562 = create_RM(P_enh, 'SuRE_K562', 'GROcap_K562', lad='LAD_K562')


P_enh$LRS_K562<- P_enh$GROcap_K562 - approx(x=RM_enh_K562$x.mean,
                                            y=RM_enh_K562$y.ilad,
                                            xout=P_enh$SuRE_K562, rule=2)$y

RM_enh_HepG2 = create_RM(P_enh, 'SuRE_HepG2', 'CAGE_HepG2', lad='LAD_HepG2')


P_enh$LRS_HepG2 <- P_enh$CAGE_HepG2 - approx(x=RM_enh_HepG2$x.mean,
                                             y=RM_enh_HepG2$y.ilad,
                                             xout=P_enh$SuRE_HepG2, rule=2)$y


P_enh$class_GROcap = classify(P_enh$SuRE_K562, P_enh$GROcap_K562,
                              P_enh$LRS_K562, P_enh$LAD_K562, -2)


P_enh$class_HepG2_3D = classify_3d(P_enh$SuRE_HepG2,
                                P_enh$CAGE_HepG2, P_enh$DNAse_HepG2,
                                P_enh$LRS_HepG2, P_enh$LAD_HepG2, 1, 0.5)


lad_names_HepG2 = c(LAD=paste0('LAD; n=', table(P_enh$LAD_HepG2)['1']),
                     iLAD=paste0('iLAD; n=', table(P_enh$LAD_HepG2)['0']))
P_enh$lad_HepG2_n = factor(ifelse(P_enh$LAD_HepG2==1, lad_names_HepG2['LAD'], lad_names_HepG2['iLAD']))
COL_lad_HepG2_enh = COL_lad
names(COL_lad_HepG2_enh) = lad_names_HepG2

RM_melt = melt(RM_enh_HepG2, measure.vars=c('y.ilad', 'y.lad'))
RM_melt$variable = ifelse(RM_melt$variable=='y.lad', lad_names_HepG2['LAD'], lad_names_HepG2['iLAD'])
ggplot(P_enh, aes(x=SuRE_HepG2, y=CAGE_HepG2, color=lad_HepG2_n)) +
    geom_point(data=P_enh[P_enh$LAD_HepG2==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P_enh[P_enh$LAD_HepG2==1, ], size=0.5, alpha=0.2) +
    ggtitle('CAGE vs SuRE HepG2') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=x.mean, y=value, color=variable), size=1) +
    labs(y='log10(CAGE)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_HepG2_enh)

ggplot(P_enh, aes(x=SuRE_HepG2, y=DNAse_HepG2, color=lad_HepG2_n)) +
    geom_point(data=P_enh[P_enh$LAD_HepG2==0, ], size=0.5, alpha=0.05) +
    geom_point(data=P_enh[P_enh$LAD_HepG2==1, ], size=0.5, alpha=0.2) +
    ggtitle('CAGE vs SuRE HepG2') +
    theme_bw() +
    geom_line(data=RM_melt, aes(x=x.mean, y=value, color=variable), size=1) +
    labs(y='log10(CAGE)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_lad_HepG2_enh)


class_names = paste0(levels(P_enh$class_HepG2_3D), '; n=',table(P_enh$class_HepG2_3D))
names(class_names) = levels(P_enh$class_HepG2_3D)
P_enh$class_HepG2_3D_n = P_enh$class_HepG2_3D
levels(P_enh$class_HepG2_3D_n) = class_names
COL_class_HepG2_enh = COL_class[names(class_names)]
names(COL_class_HepG2_enh) = class_names

p_classes = P_enh[which(P_enh$class_HepG2_3D %in% c('inactive', 'escaper', 'repressed')),]
ggplot(P_enh, aes(x=SuRE_HepG2, y=DNAse_HepG2, color=class_HepG2_3D_n)) +
    geom_point(size=0.1,color=COLi) +
    geom_point(data=p_classes, size=0.5, alpha=0.5) +
    ggtitle('CAGE vs SuRE HepG2') +
    theme_bw() +
    geom_line(data=RM_enh_HepG2, aes(x=x.mean, y=y.ilad), color=COL_lad['iLAD']) +
    geom_line(data=RM_enh_HepG2, aes(x=x.mean, y=y.lad), color=COL_lad['LAD']) +
    labs(y='log10(CAGE)', x='log10(SuRE)') +
    theme(legend.title=element_blank()) +
    scale_color_manual(values=COL_class_HepG2_enh)
```


```{r}

wilcox_affinity <- function(x_affinity, y_affinity, groups, tf_table, id_vec=NULL){
  if (is.null(id_vec)){
    id_vec = colnames(x_affinity)
  }  
  fit = mclapply(id_vec, function(id){
    x = x_affinity[,id]
    y = y_affinity[,id]
    r = rank(c(x,y))
    r_x = r[1:length(x)]
    r_y = r[-(1:length(x))]
    mean_r = c(mean(r_x), mean(r_y))
    direction = groups[which(mean_r==max(mean_r))]
    if (length(direction) == 2){
      direction = 'unchanged'
    }
    median_fc = median(x) / median(y)
    mean_fc = mean(x) / mean(y)
    rank_sum = sum(r_x)
    w = wilcox.test(x, y)
    return(list(w,direction, median_fc, mean_fc, rank_fc))
  })
  p_vec = unlist(lapply(fit, function(x){ x[[1]]$p.value}))
  p_adjust = p.adjust(p_vec, method='fdr')
  direction = lapply(fit, function(x){ x[[2]]})
  median_fc =  lapply(fit, function(x){ x[[3]]})
  mean_fc =  lapply(fit, function(x){ x[[4]]})
  rank_fc =  lapply(fit, function(x){ x[[5]]})
  result_table = cbind(id=id_vec,
                       tf_table[id_vec, ],
                       direction=unlist(direction),
                       p_adjust = p_adjust,
                       median_fc = unlist(median_fc),
                       mean_fc = unlist(mean_fc),
                       rank_fc = unlist(rank_fc), stringsAsFactors=F)
  return(result_table)
}

apply_wilcox <- function(affinity, prom_table, group_vec, class_name,
                         tf_table, id_vec=NULL){
    if (is.null(id_vec)){
        id_vec = colnames(affinity)
    }
    id_vec = id_vec[id_vec%in%colnames(affinity)]
    fit = mclapply(id_vec, function(id){
        aff_vec = affinity[rownames(prom_table),id]
        group1 = which(prom_table[, class_name]==group_vec[1])
        rank_aff = rank(aff_vec)
        x = aff_vec[group1]
        y = aff_vec[-group1]
        rank_mean = c(mean(rank_aff[group1]), mean(rank_aff[-group1]))
        names(rank_mean) = paste0('mean_rank_', group_vec)
        rank_fc = mean(rank_aff[group1]) /mean(rank_aff[-group1])
        direction = ifelse(rank_fc > 1, group_vec[1], group_vec[2])
        median_fc = median(x) / median(y)
        mean_fc = mean(x) / mean(y)
        w = wilcox.test(x, y)
        return(list(w,direction, median_fc, mean_fc, rank_fc, rank_mean))
  })
  p_vec = unlist(lapply(fit, function(x){ x[[1]]$p.value}))
  p_adjust = p.adjust(p_vec, method='fdr')
  direction = lapply(fit, function(x){ x[[2]]})
  median_fc =  lapply(fit, function(x){ x[[3]]})
  mean_fc =  lapply(fit, function(x){ x[[4]]})
  rank_fc =  lapply(fit, function(x){ x[[5]]})
  rank_mean = do.call(rbind, lapply(fit, function(x){ x[[6]]}))
  result_table = data.frame(id=id_vec,
                            tf_table[id_vec, ],
                            direction=unlist(direction),
                            p_adjust = p_adjust,
                            mean_fc = unlist(mean_fc),
                            median_fc = unlist(median_fc),
                            rank_fc = unlist(rank_fc),
                            rank_mean,
                            stringsAsFactors=F)
  return(result_table)
}

apply_kruskal <- function(affinity, prom_table, class_name,
                         tf_table, id_vec=NULL){
    if (is.null(id_vec)){
        id_vec = colnames(affinity)
    }
    id_vec = id_vec[id_vec%in%colnames(affinity)]
    fit = mclapply(id_vec, function(id){
        aff_vec = affinity[rownames(prom_table),id]
        data = data.frame(affinity=aff_vec,
                          class=factor(prom_table[,class_name]))
        data$rank = rank(aff_vec)
        rank_mean_list = lapply(levels(data$class),
                                function(class, data){
                                        group_vec = which(data$class==class)
                                        mean(data$rank[group_vec])
                                    }, data)
        rank_mean_vec = unlist(rank_mean_list)
        names(rank_mean_vec) = paste0(levels(data$class),
                                      '_mean_rank')
        k = kruskal.test(affinity ~ class, data=data)
        min_rank = levels(data$class)[rank_mean_vec==min(rank_mean_vec)]
        max_rank = levels(data$class)[rank_mean_vec==max(rank_mean_vec)]
        return(list(k, rank_mean_vec, min_rank, max_rank))
  }, mc.cores=10)
  p_vec = unlist(lapply(fit, function(x){ x[[1]]$p.value}))
  p_adjust = p.adjust(p_vec, method='fdr')
  rank_mean =  do.call(rbind, lapply(fit, function(x){ x[[2]]}))
  rank_min =  lapply(fit, function(x){ x[[3]]})
  rank_max =  lapply(fit, function(x){ x[[4]]})
  result_table = data.frame(id=id_vec,
                            tf_table[id_vec, ],
                            p_adjust = p_adjust,
                            rank_mean,
                            min_rank = unlist(rank_min),
                            max_rank = unlist(rank_max),
                            stringsAsFactors=F)
  return(result_table)
}


aff_table_jaspar = read.table('../raw_data/jaspar_affinity_pseudo/seq_psam.dat',
                              stringsAsFactors=F)
tf_table = read.table('../raw_data/tf_table.txt', sep='\t', row.names=1,
                      stringsAsFactors=F)
colnames(tf_table) = c('name', 'species', 'class', 'family')

tf_translation = ddply(tf_table[,c('species','name')], .(name),
                       function(x){
                         symbol = gsub('[(]var.[0-9][)]','', x[,2])
                         cbind(x,symbol=unlist(strsplit(symbol, '::')))
                       })
tf_translation$symbol = as.character(tf_translation$symbol)

find_symbol <- function(symbol, id_table){
    result = id_table[which(id_table$symbol==toupper(symbol)),'gene_id']
    if (length(result) == 0){
        result = NA
    } else {
        table = table(result)
        result = names(which.max(table))
    }
    return(result)
}

id_table = read.table('../raw_data/transcript.table', stringsAsFactors=F,
                    row.names=1, col.names=c('transcript_id', 'gene_id',
                                             'symbol'))
id_table = unique(id_table)
rownames(id_table) = id_table$gene_id

tf_translation$gene_id = unlist(lapply(tf_translation$symbol, find_symbol,
                                       id_table))


rnaseq_rep1 = read.table('../raw_data/K562_rna_rep1_ENCFF004LGY.tsv',
                         header=T, row.names=1, stringsAsFactors=F)
rnaseq_rep2 = read.table('../raw_data/K562_rna_rep2_ENCFF222NCB.tsv',
                         header=T, row.names=1, stringsAsFactors=F)


fpkm_rep1 = rnaseq_rep1[tf_translation$gene_id, 'pme_FPKM']
fpkm_rep2 = rnaseq_rep2[tf_translation$gene_id, 'pme_FPKM']



tf_translation$K562_fpkm = pseudo_log10(rowMeans(cbind(fpkm_rep1, fpkm_rep2)))


rnaseq_rep1 = read.table('../raw_data/HepG2_rna_rep1_ENCFF974MUO.tsv',
                         header=T, row.names=1, stringsAsFactors=F)
rnaseq_rep2 = read.table('../raw_data/HepG2_rna_rep2_ENCFF649AHO.tsv',
                         header=T, row.names=1, stringsAsFactors=F)


fpkm_rep1 = rnaseq_rep1[tf_translation$gene_id, 'pme_FPKM']
fpkm_rep2 = rnaseq_rep2[tf_translation$gene_id, 'pme_FPKM']



tf_translation$HepG2_fpkm = pseudo_log10(rowMeans(cbind(fpkm_rep1, fpkm_rep2)))


KBM7_essential = read.table('../raw_data/KBM7_essentialome_aac7557_SM_Table_S1.csv',
                            stringsAsFactors=F, skip=1, header=T, sep='\t',
                            row.names=2)
HAP1_essential = read.table('../raw_data/HAP1_essentialome_aac7557_SM_Table_S2.csv',
                            stringsAsFactors=F, skip=1, header=T, sep='\t',
                            row.names=2)
ens_vec = gsub('[.].*','',tf_translation$gene_id)
tf_translation$KBM7_essential = KBM7_essential[ens_vec, 'selected']=='YES'
tf_translation$HAP1_essential = HAP1_essential[ens_vec, 'selected']=='YES'

tf_expression = ddply(tf_translation,.(name),
                      function(x){
                          c(K562_fpkm=min(x$K562_fpkm),
                            HepG2_fpkm=min(x$HepG2_fpkm),
                            KBM7_essential=any(x$KBM7_essential),
                            HAP1_essential=any(x$HAP1_essential))
                      })


tf_table$K562_fpkm = NaN
tf_table$HepG2_fpkm = NaN
tf_match = match(tf_expression$name, tf_table$name)
tf_table$K562_fpkm[tf_match] = tf_expression$K562_fpkm
tf_table$HepG2_fpkm[tf_match] = tf_expression$HepG2_fpkm
tf_table$KBM7_essential = NA
tf_table$KBM7_essential[tf_match] = tf_expression$KBM7_essential
tf_table$HAP1_essential = NA
tf_table$HAP1_essential[tf_match] = tf_expression$HAP1_essential


tf_type = read.table('../raw_data/12859_2016_1349_MOESM2_ESM.csv',
                     sep='\t', header=T, stringsAsFactors=F)
match_vec = match(toupper(tf_table$name), tf_type$Symbol)
tf_table$tf_type = NA
not_na = !is.na(match_vec)
tf_table[not_na, 'tf_type'] = tf_type[match_vec[not_na],
                                      'Chromatin.Opening.Type']

colnames(aff_table_jaspar) = gsub('.xml','', colnames(aff_table_jaspar))
rownames(aff_table_jaspar) = gsub('::.*','', rownames(aff_table_jaspar))


load('../raw_data/cl20170814_jaspar_gencode_cage_cor.rda')
cor_data = t(cor_data)
colnames(cor_data) = rownames(tf_table)[match(colnames(cor_data), tf_table$name)]


aff_table_enhancer = read.table('../raw_data/jaspar_affinity_enhancers/seq_psam.dat',
                                stringsAsFactors=F)
colnames(aff_table_enhancer) = gsub('.xml','', colnames(aff_table_enhancer))  
rownames(aff_table_enhancer) = gsub('::.*','', rownames(aff_table_enhancer))                      

aff_table_enhancer = aff_table_enhancer[, colnames(cor_data)]


id_vec = colnames(cor_data)

matched_evsr_K562 = matchSet(P[P$class_GROcap%in%c('repressed', 'escaper'), ],
                             'class_GROcap', 'escaper', 'SuRE_K562')

evsr_jaspar_K562 = apply_wilcox(aff_table_jaspar, matched_evsr_K562,
                                c('repressed', 'escaper'), 'class_GROcap',
                                tf_table, id_vec)


evsr_jaspar_exp = apply_wilcox(cor_data, matched_evsr_K562,
                               c('escaper', 'repressed'), 'class_GROcap',
                               tf_table)
evsr_jaspar_K562$p_fantom = evsr_jaspar_exp[evsr_jaspar_K562$id, 'p_adjust']
evsr_jaspar_K562$mean_rank_fantom_escaper = evsr_jaspar_exp[evsr_jaspar_K562$id,
                                                          'mean_rank_escaper']
evsr_jaspar_K562$mean_rank_fantom_repressed = evsr_jaspar_exp[evsr_jaspar_K562$id,
                                                            'mean_rank_repressed']


matched_enhancer = matchSet(P_enh[P_enh$class_GROcap%in%c('repressed', 'escaper'), ],
                            'class_GROcap', 'escaper', 'SuRE_K562')


evsr_jaspar_enh = apply_wilcox(aff_table_enhancer, matched_enhancer,
                               c('escaper', 'repressed'), 'class_GROcap',
                               tf_table, id_vec)

evsr_jaspar_K562$p_enh = evsr_jaspar_enh[evsr_jaspar_K562$id, 'p_adjust']
evsr_jaspar_K562$mean_rank_enh_escaper = evsr_jaspar_enh[evsr_jaspar_K562$id,
                                                         'mean_rank_escaper']
evsr_jaspar_K562$mean_rank_enh_repressed = evsr_jaspar_enh[evsr_jaspar_K562$id,
                                                           'mean_rank_repressed']





matched_evsr_HepG2 = matchSet(P[P$class_GROcap%in%c('repressed', 'escaper'), ],
                            'class_HepG2_3D', 'escaper', 'SuRE_HepG2')

evsr_jaspar_HepG2 = apply_wilcox(aff_table_jaspar, matched_evsr_HepG2,
                               c('repressed', 'escaper'), 'class_HepG2_3D',
                               tf_table, id_vec)


evsr_jaspar_exp = apply_wilcox(cor_data, matched_evsr_HepG2,
                              c('escaper', 'repressed'), 'class_HepG2_3D',
                              tf_table)
evsr_jaspar_HepG2$p_fantom = evsr_jaspar_exp[evsr_jaspar_HepG2$id, 'p_adjust']
evsr_jaspar_HepG2$mean_rank_fantom_escaper = evsr_jaspar_exp[evsr_jaspar_HepG2$id,
                                                         'mean_rank_escaper']
evsr_jaspar_HepG2$mean_rank_fantom_repressed = evsr_jaspar_exp[evsr_jaspar_HepG2$id,
                                                           'mean_rank_repressed']


matched_enhancer = matchSet(P_enh[P_enh$class_HepG2_3D%in%
                                  c('repressed', 'escaper'), ],
                           'class_HepG2_3D', 'escaper', 'SuRE_HepG2')


evsr_jaspar_enh = apply_wilcox(aff_table_enhancer, matched_enhancer,
                              c('escaper', 'repressed'), 'class_HepG2_3D',
                              tf_table, id_vec)

evsr_jaspar_HepG2$p_cage = evsr_jaspar_enh[evsr_jaspar_HepG2$id, 'p_adjust']
evsr_jaspar_HepG2$mean_rank_enh_escaper = evsr_jaspar_enh[evsr_jaspar_HepG2$id,
                                                        'mean_rank_escaper']
evsr_jaspar_HepG2$mean_rank_enh_repressed = evsr_jaspar_enh[evsr_jaspar_HepG2$id,
                                                          'mean_rank_repressed']

```

```{r}
evsr_jaspar_K562$x = tolower(evsr_jaspar_K562$tf_type)
evsr_jaspar_K562$x[is.na(evsr_jaspar_K562$x)] = ''

evsr_jaspar_HepG2$x = tolower(evsr_jaspar_HepG2$tf_type)
evsr_jaspar_HepG2$x[is.na(evsr_jaspar_HepG2$x)] = ''

ggplot(evsr_jaspar_K562, aes(x=x, y=mean_rank_escaper)) +
    geom_violin(color='black') +
    ggtitle('mean rank enhancer promoter affinity K562') +
    geom_point(position=position_jitter(width=0.3)) +
    scale_color_manual(values= COL_class)

ggplot(evsr_jaspar_HepG2, aes(x=x, y=mean_rank_escaper)) +
    geom_violin(color='black') +
    ggtitle('mean rank enhancer promoter affinity HepG2') +
    geom_point(position=position_jitter(width=0.3)) +
    scale_color_manual(values= COL_class)

```

```{r}

evsr_jaspar_K562$distance_evsr = sqrt(rank(evsr_jaspar_K562$mean_rank_escaper)^2 +
                                   rank(evsr_jaspar_K562$mean_rank_enh_escaper)^2)

evsr_jaspar_HepG2$distance_evsr = sqrt(rank(evsr_jaspar_HepG2$mean_rank_escaper)^2 +
                                       rank(evsr_jaspar_HepG2$mean_rank_enh_escaper)^2)


dist_order = order(evsr_jaspar_K562$distance_evsr, decreasing=T)
kable(head(evsr_jaspar_K562[dist_order, c('name', 'class', 'tf_type', 'K562_fpkm',
                                          'HepG2_fpkm', 'distance_evsr')], 20),
      row.names=F)



dist_order = order(evsr_jaspar_HepG2$distance_evsr, decreasing=T)
kable(head(evsr_jaspar_HepG2[dist_order, c('name', 'class', 'tf_type', 'K562_fpkm',
                                           'HepG2_fpkm', 'distance_evsr')], 20),
    row.names=F)

```

```{r}

evsr_jaspar_K562$distance_evsr = sqrt(rank(evsr_jaspar_K562$mean_rank_escaper)^2 +
                                   rank(evsr_jaspar_K562$mean_rank_enh_escaper)^2 +
                                   rank(evsr_jaspar_K562$mean_rank_fantom_escaper)^2)

evsr_jaspar_HepG2$distance_evsr = sqrt(rank(evsr_jaspar_HepG2$mean_rank_escaper)^2 +
                                       rank(evsr_jaspar_HepG2$mean_rank_enh_escaper)^2 +
                                       rank(evsr_jaspar_HepG2$mean_rank_fantom_escaper)^2)


dist_order = order(evsr_jaspar_K562$distance_evsr, decreasing=T)
kable(head(evsr_jaspar_K562[dist_order, c('name', 'class', 'tf_type', 'K562_fpkm',
                                          'HepG2_fpkm', 'distance_evsr')], 20),
      row.names=F)



dist_order = order(evsr_jaspar_HepG2$distance_evsr, decreasing=T)
kable(head(evsr_jaspar_HepG2[dist_order, c('name', 'class', 'tf_type', 'K562_fpkm',
                                           'HepG2_fpkm', 'distance_evsr')], 20),
    row.names=F)

```
