# knitr document van Steensel lab

# Modes and mechanisms of gene regulation in lamina-associated domains
## Christ Leemans, 03-11-2016 - 26-07-2017

## Introduction
In this report I would like to catalog the code used to supply the figures for
the manuscript.

## promoter selection
First we have to select our promoters. Previously Joris used gencode dataset and
selected promoters for which there was also expression in CAGE in at least one
cell type. I don't see any reason not to keep using this method.

```

## select all transcript start sites
zcat ~/mydata/data/tracks/hg38/gencode.v24.annotation.gff3.gz | \
    sed -rn "s/ID=([^;]+)/\1\t/p" | sed -rn "s/gene_id=([^;]+)/\t\1\t/p" | \
    awk -vOFS='\t' '{
            if ($3=="transcript"){
                if ($7 == "+"){
                    tss=$4 - 1
                } else {
                    tss=$5 - 1
                }
                print $1 , tss , tss + 1, $9 , "." , $7 , $11
            }
        }' | gzip -c > raw_data/gencode_v24_tss.bed.gz


## For multiple transcripts coming from the same gene, we want to select transcription start
## sites at least 500bp apart.

## select unique transcript start sites which overlap with a cage peak.
## CAGE peaks have at least 1 transcript in one of the tissues.
## (multiple transcripts of same gene can start at same position we don't want those).

bedtools intersect -wb -a raw_data/gencode_v24_tss.bed.gz \
                   -b ~/mydata/data/tracks/hg38/hg38_fair+new_CAGE_peaks_phase1and2.bed.gz | \
    awk -vOFS='\t' '{
        if ($8 > 0){
            name=$1 FS $3 FS $6 FS $7 FS $11
            if (!(name in seen)){
                location=$1 FS $3 FS $6 FS $7
                if (!(location in loc)){
                    print $1, $3, $4, $6, $7 > "raw_data/gencode_v24_fantom_tss.txt"
                }
                loc[location] = 1
                print $4, $11 > "raw_data/gencode_v24_to_fantom.txt"
            }
            seen[name]=1
        }}'
gzip raw_data/gencode_v24_fantom_tss.txt
gzip raw_data/gencode_v24_to_fantom.txt


## If TSS's are closer together, let's take the TSS that is generally most highly transcribed.
## for this we will need to have some information on transcription rates across fantom5
## dataset.
## might as well also count the number of tissues expressed, since this will be used later.

awk -vOFS='\t' '{
        if (NR==FNR){
            arr[$2][$1] = 1
        } else {
            if ($1=="01STAT:MAPPED"){
                for (i=2;i<NF;i++){
                    total[i] = $i
                }
            } else if ($1 in arr) {
                sum=0
                number=0
                for (i=2;i<NF;i++){
                    sum+=$i/total[i]
                    if ($i > 0){
                        number++
                    }
                }
                for (name in arr[$1]){
                    print name, sum, number
                }
            }
        }
    }' <(zcat raw_data/gencode_v24_to_fantom.txt.gz) \
    <(zcat ~/mydata/data/tracks/hg38/hg38_fair+new_CAGE_peaks_phase1and2_counts.osc.txt.gz) | \
    gzip -c > raw_data/gencode_v24_fantom_expr.txt.gz



```



```{r}
library(rtracklayer)
library(data.table)

P = read.table('../raw_data/gencode_v24_fantom_tss.txt.gz',
               col.names=c('seqnames', 'tss', 'transcript_id', 'strand', 'gene_id'),
               stringsAsFactors=F)
rownames(P) = P$transcript_id
P = P[!P$seqnames %in% c('chrM', 'chrY'),  ]


fantom_count = read.table('../raw_data/gencode_v24_fantom_expr.txt.gz',
                          col.names=c('transcript_id', 'sum', 'tissues_expressed'))
fantom_table = data.table(fantom_count)
setkey(fantom_table, transcript_id)
fantom_high = fantom_table[, list(sum=max(sum),
                                  tissues_expressed=max(tissues_expressed)),
                           by=key(fantom_table)]

P$sum_fantom = unlist(fantom_high[P$transcript_id, 'sum'])
P$tissues_expressed = unlist(fantom_high[P$transcript_id, 'tissues_expressed'])

most_active <- function(P){
    p_table = as.data.table(P)
    max_sum = p_table[, list(transcript_id=transcript_id[which.max(sum_fantom)]),
                      by=gene_id]
    return(as.vector(unlist(max_sum$transcript_id)))
}

p_active = most_active(P)
p_other = P[!rownames(P)%in%p_active, ]
p_new_names = p_active

while (nrow(p_other) > 0){
    p_new = P[p_new_names, ]
    active_gr = makeGRangesFromDataFrame(p_new, start.field = "tss",
                                         end.field="tss", keep.extra.columns=TRUE)
    other_gr = makeGRangesFromDataFrame(p_other, start.field = "tss",
                                         end.field="tss", keep.extra.columns=TRUE)
    o = findOverlaps(active_gr,other_gr, maxgap=500, ignore.strand=FALSE)
    sub_o = o[p_new[queryHits(o), 'gene_id'] == p_other[subjectHits(o), 'gene_id']]
    p_other = p_other[-subjectHits(sub_o), ]
    p_active = most_active(p_other)
    p_other = p_other[!rownames(p_other)%in%p_active, ]
    p_new_names = c(p_new_names, p_active)
}

p_complete = P

P = P[rownames(P)%in%p_new_names, ]


write.table(P, '../raw_data/gencode_v24_maxcage_tss.txt',
            quote=F, sep='\t', col.names=F, row.names=F)

gencode_gr = import.gff('~/mydata/data/tracks/hg38/gencode.v24.annotation.gff3.gz')

p_gr = gencode_gr[which(gencode_gr$transcript_id %in% p_new_names)]
export.gff(p_gr, "../raw_data/transcript_selection_hg38.gff")

```

```
paste <(awk -vOFS='\t' '{print $1, $2 - 501, $2 + 500, $3, "0", $4}' \
          raw_data/gencode_v24_maxcage_tss.txt | \
          bwtool summary -skip-median -with-sum -keep-bed /dev/stdin \
                         ~/mydata/data/grocap/grocap_wTAP_plus_K562_hg38_SRR1552480.bw \
                         /dev/stdout ) \
      <(awk -vOFS='\t' '{print $1, $2 - 501, $2 + 500, $3, "0", $4}' \
         raw_data/gencode_v24_maxcage_tss.txt | \
         bwtool summary -skip-median -with-sum -keep-bed /dev/stdin \
                              ~/mydata/data/grocap/grocap_wTAP_min_K562_hg38_SRR1552480.bw \
                              /dev/stdout ) | \
    awk -vOFS='\t' 'BEGIN{
                        print "name", "plus", "min", "sense"
                    }{
                        print $4 , $12 , $24 , $6=="+"?$12:$24
                    }' | \
    gzip -c > raw_data/tss_v24_grocap_wTAP_SRR1552480.txt.gz


```

```{r}
grocap_v24 = read.table('../raw_data/tss_v24_grocap_wTAP_SRR1552480.txt.gz',
                        stringsAsFactors=F, row.names=1, header=T)

grocap_v24$sense_norm = grocap_v24$sense / sum(grocap_v24$sense) * 1000000 / 1001

load("~joris/mydata/git/SuRE/Joris//analysis_postNBT/Gencode_DF_generation_170707/gencode.sure.170712.rda")


P<-gencode.sure.170712[,c('chr', 'strand', 'txStart', 'txEnd', 'name', 'name2',
                          'tss', 'distance.to.previous.tss',
                          'k562.combined.45.55.sense',
                          'HT1080.sense', 'gro.cap.1kb.sense',
                          'encode.cage.reprocessed.1kb.sense',
                          'nr.of.tissues.in.which.expressed.max')]
names(P)[9:13]<-c("SuRE_K562", "SuRE_HT1080", "GROcap_K562", "CAGE_K562_encode", 'tissues_expressed')

tid_v24 = gsub('[.][0-9]+', '', rownames(grocap_v24))
tid_v19 = gsub('[.][0-9]+', '', P$name)

match_vec = match(tid_v24, tid_v19)



grocap_data = data.frame(v24=grocap_v24$sense_norm[!is.na(match_vec)],
                         v19=P$GROcap_K562[match_vec[!is.na(match_vec)]],
                         row.names=tid_v24[!is.na(match_vec)])

ggplot(grocap_data, aes(x=log10(v24), y=log10(v19))) +
    geom_point(size=0.3, alpha=0.3)
```

**conclusion:**
In general grocap voor v24 and v19 are similar, however there are a bunch of
transcripts with higher expression estimates for v24. This could be multiple
things: a difference annotation, a difference in the GROcap analysis, or a
difference in how the mean was calculated.

```
awk -vOFS='\t' '{print $1, $2, $2, $3, "0", $4}' \
   raw_data/gencode_v24_maxcage_tss.txt > raw_data/gencode_v24_maxcage_tss.bed

```

```{r}
library(CoverageView)
grocap_plus_gr = CoverageBigWigFile('~/mydata/data/grocap/grocap_wTAP_plus_K562_hg38_SRR1552480.bw')
grocap_min_gr = CoverageBigWigFile('~/mydata/data/grocap/grocap_wTAP_min_K562_hg38_SRR1552480.bw')


cov_vec_plus = cov.matrix(grocap_plus_gr,
                          coordfile='../raw_data/gencode_v24_maxcage_tss.bed',
                          extend = 500, bin_width=1000, num_cores=20)

cov_vec_min = cov.matrix(grocap_min_gr,
                         coordfile='../raw_data/gencode_v24_maxcage_tss.bed',
                         extend = 500, bin_width=1000, num_cores=20)


tss_gr = import.bed('../raw_data/gencode_v24_maxcage_tss.bed')

cov_vec = ifelse(strand(tss_gr)=='+', cov_vec_plus, cov_vec_min)

grocap_data = data.frame(v24_cview=cov_vec,
                         v24_bwtool=grocap_v24$sense_norm,
                         row.names=tid_v24)
ggplot(grocap_data, aes(x=log10(v24_cview), y=log10(v24_bwtool))) +
    geom_point(size=0.3, alpha=0.3)
```
**conclusion:**
Difference is not due to the method used to obtain the expression for each promoter.
The mapping was used in previous analysis is more stringent and might have included
contigs/alternative assemblies of hg38.

```
bedtools intersect -c -a raw_data/gencode_v24_maxcage_tss.bed \
    -b /home/t.v.schaik/mydata/proj/tests/results/ts171030_PipelineImprovements/results_alldata/HMM/K562_r1_LMNB1-5kb_AD.bed.gz | \
    awk -vOFS='\t' '{print $4, $7}' | \
    gzip -c > raw_data/lad_hmm_tss_v24.txt.gz




```



```{r}




lmnb1_gatc_gr = import.bw('/home/t.v.schaik/mydata/proj/tests/results/ts171030_PipelineImprovements/results_alldata/tracks/normalized/K562_r1_LMNB1-norm.bw')


lad_gr = import.bed('/home/t.v.schaik/mydata/proj/tests/results/ts171030_PipelineImprovements/results_alldata/HMM/K562_r1_LMNB1-5kb_AD.bed.gz')




```
